{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2e0e043e-1fe1-48e2-86ad-493abc1588b9",
   "metadata": {},
   "source": [
    "# Creating a gold-standard dataset from pf8 and GenRE MEKONG data\n",
    "\n",
    "The aim of this notebook is to create a gold-standard dataset for malaria genomics pipelines that predict drug resistance.\n",
    "\n",
    "The following resources will be used:\n",
    "1. [MalariaGEN Pf8](https://www.malariagen.net/data_package/open-dataset-plasmodium-falciparum-v80/)\n",
    "2. [GenRE Mekong](https://www.malariagen.net/resource/29/)\n",
    "\n",
    "## Rationale for selection of resources\n",
    "Pipelines are usually tailored to a particular sequences technology, such as whole-genome or amplicon sequencing. While Pf8 is based on (selective) Whole Genome Sequencing (sWGS) data, GenRE Mekong utilised the SPotMalaria panel for amplicon sequencing (download available at [https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx](https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx).\n",
    "Some GenRE Mekong samples have been re-analysed with sWGS for Pf8, thus providing data for the same sample from two different technologies and two independent resistance phenotype calls.\n",
    "We will use the intersect between Pf8 and GenRE Mekong as the gold-standard dataset.\n",
    "\n",
    "In Pf8, drug-resistance phenotypes are inferred based on criteria described in [this PDF](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf)\n",
    "\n",
    "## Goal\n",
    "THe notebook creates a csv file of sample IDs and metadata. The sample IDs can be used to retrieve the raw read data from public databases.  \n",
    "This can be done manually or by using a tool such as [nf-core fetchngs](https://nf-co.re/fetchngs/1.12.0).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7e72b18-623d-48c5-b585-c15c9524b4e4",
   "metadata": {},
   "source": [
    "## Set up environment\n",
    "All packages imported here must be installed locally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "947cd6a8-8192-4831-a821-4c1072c19799",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from warnings import simplefilter \n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "pd.options.mode.copy_on_write = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b76cd977-c763-4e6e-b8cd-dad98e7898b3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load Pf8 data\n",
    "Obtain sample metadata and drug resisance inference data from Pf8.\n",
    "Details can be found [here](https://www.malariagen.net/data_package/open-dataset-plasmodium-falciparum-v80/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee77eaca-480f-4054-92f4-034d436dc1ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_infer_resistance_df=pd.read_csv(\"https://pf8-release.cog.sanger.ac.uk/Pf8_inferred_resistance_status_classification.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6645dee-44ba-4cba-a84d-fa60aec1c5c6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_samples_df=pd.read_csv(\"https://pf8-release.cog.sanger.ac.uk/metadata/Pf8_samples.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96193a2d-f629-4438-9237-5d3716f4f153",
   "metadata": {},
   "source": [
    "## Merge Pf8 resistance and sample metadata\n",
    "Also rename the sample column with lower case for consistency across data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "07200b79-bc54-4207-b6b9-f42506098629",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_samples_df.rename(columns={'Sample': 'sample'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "af63fb33-857b-47ce-96c4-931143ad50d3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_df = pd.merge(pf8_infer_resistance_df, pf8_samples_df, on=[\"sample\"], how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "74cbe906-aa62-4e06-8ba5-e7a47f90d90c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>SP (uncomplicated)</th>\n",
       "      <th>SP (IPTp)</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>...</th>\n",
       "      <th>Admin level 1 longitude</th>\n",
       "      <th>Year</th>\n",
       "      <th>ENA</th>\n",
       "      <th>All samples same case</th>\n",
       "      <th>Population</th>\n",
       "      <th>% callable</th>\n",
       "      <th>QC pass</th>\n",
       "      <th>Exclusion reason</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>Sample was in Pf7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081237</td>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>82.48</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081238</td>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>88.95</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889621</td>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>87.01</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FP0011-CW</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889624</td>\n",
       "      <td>FP0011-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>86.95</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FP0012-CW</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889627</td>\n",
       "      <td>FP0012-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>89.86</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33320</th>\n",
       "      <td>SPT92268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>ERR10940681</td>\n",
       "      <td>SPT92268</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>0.10</td>\n",
       "      <td>False</td>\n",
       "      <td>Low_coverage</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33321</th>\n",
       "      <td>SPT92269</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>ERR11009733</td>\n",
       "      <td>SPT92269</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>0.12</td>\n",
       "      <td>False</td>\n",
       "      <td>Low_coverage</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33322</th>\n",
       "      <td>SPT92270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>ERR11009737</td>\n",
       "      <td>SPT92270</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>0.01</td>\n",
       "      <td>False</td>\n",
       "      <td>Low_coverage</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33323</th>\n",
       "      <td>SPT94772</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.401559</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>ERR10789456</td>\n",
       "      <td>SPT94772</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>64.96</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33324</th>\n",
       "      <td>SPT94773</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.401559</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>ERR10789457</td>\n",
       "      <td>SPT94773</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>90.53</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33325 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample   Chloroquine Pyrimethamine   Sulfadoxine    Mefloquine  \\\n",
       "0       FP0008-C  Undetermined  Undetermined  Undetermined     Sensitive   \n",
       "1       FP0009-C     Resistant     Resistant     Sensitive     Sensitive   \n",
       "2      FP0010-CW  Undetermined     Resistant     Resistant     Sensitive   \n",
       "3      FP0011-CW  Undetermined     Resistant  Undetermined     Sensitive   \n",
       "4      FP0012-CW     Resistant     Resistant     Sensitive     Sensitive   \n",
       "...          ...           ...           ...           ...           ...   \n",
       "33320   SPT92268           NaN           NaN           NaN           NaN   \n",
       "33321   SPT92269           NaN           NaN           NaN           NaN   \n",
       "33322   SPT92270           NaN           NaN           NaN           NaN   \n",
       "33323   SPT94772     Resistant     Resistant     Resistant  Undetermined   \n",
       "33324   SPT94773     Resistant     Resistant     Resistant     Sensitive   \n",
       "\n",
       "        Artemisinin   Piperaquine SP (uncomplicated)  SP (IPTp)         AS-MQ  \\\n",
       "0         Sensitive     Sensitive          Sensitive  Sensitive     Sensitive   \n",
       "1         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "2         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "3         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "4         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "...             ...           ...                ...        ...           ...   \n",
       "33320           NaN           NaN                NaN        NaN           NaN   \n",
       "33321           NaN           NaN                NaN        NaN           NaN   \n",
       "33322           NaN           NaN                NaN        NaN           NaN   \n",
       "33323  Undetermined  Undetermined          Resistant  Sensitive  Undetermined   \n",
       "33324     Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "\n",
       "       ... Admin level 1 longitude    Year          ENA All samples same case  \\\n",
       "0      ...               -9.832345  2014.0   ERR1081237              FP0008-C   \n",
       "1      ...               -9.832345  2014.0   ERR1081238              FP0009-C   \n",
       "2      ...               -9.832345  2014.0   ERR2889621             FP0010-CW   \n",
       "3      ...               -9.832345  2014.0   ERR2889624             FP0011-CW   \n",
       "4      ...               -9.832345  2014.0   ERR2889627             FP0012-CW   \n",
       "...    ...                     ...     ...          ...                   ...   \n",
       "33320  ...                3.618846  2000.0  ERR10940681              SPT92268   \n",
       "33321  ...                3.618846  2000.0  ERR11009733              SPT92269   \n",
       "33322  ...                3.618846  2000.0  ERR11009737              SPT92270   \n",
       "33323  ...              -16.401559  2017.0  ERR10789456              SPT94772   \n",
       "33324  ...              -16.401559  2017.0  ERR10789457              SPT94773   \n",
       "\n",
       "       Population  % callable  QC pass  Exclusion reason  Sample type  \\\n",
       "0            AF-W       82.48     True      Analysis_set         gDNA   \n",
       "1            AF-W       88.95     True      Analysis_set         gDNA   \n",
       "2            AF-W       87.01     True      Analysis_set         sWGA   \n",
       "3            AF-W       86.95     True      Analysis_set         sWGA   \n",
       "4            AF-W       89.86     True      Analysis_set         sWGA   \n",
       "...           ...         ...      ...               ...          ...   \n",
       "33320        AF-W        0.10    False      Low_coverage         sWGA   \n",
       "33321        AF-W        0.12    False      Low_coverage         sWGA   \n",
       "33322        AF-W        0.01    False      Low_coverage         sWGA   \n",
       "33323        AF-W       64.96     True      Analysis_set         sWGA   \n",
       "33324        AF-W       90.53     True      Analysis_set         sWGA   \n",
       "\n",
       "      Sample was in Pf7  \n",
       "0                  True  \n",
       "1                  True  \n",
       "2                  True  \n",
       "3                  True  \n",
       "4                  True  \n",
       "...                 ...  \n",
       "33320             False  \n",
       "33321             False  \n",
       "33322             False  \n",
       "33323             False  \n",
       "33324             False  \n",
       "\n",
       "[33325 rows x 27 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pf8_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f0df980-9e83-491f-b072-c5f6666861c8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load GenRe Mekong data\n",
    "Using this file from the GenRe Mekong webstie, release v1.4 14/05/24 (latest release):  \n",
    "\n",
    "https://assets.super.so/abd63e6d-92df-4a9f-b19c-3a65d59b8a34/files/4accef63-0436-4893-a00a-6ea0e8a55249/Pf-GenRe-GRCv1.4-Full-PublicRelease-20240514.xlsx  \n",
    "Tab: GRC\n",
    "\n",
    "Sample IDs for samples that were re-analysed in Pf8 are the same in both projects, i.e. it is possible to merge GenRe Mekong and Pf8 data on sample ID.  \n",
    "\n",
    "For more information on the data in this spreadsheet, consult the [GenRe Mekong data dictionary](https://github.com/GenRe-Mekong/Documents/blob/e331501b8e93752221ebd7ae72e5eddd3e6e924d/Pf-GenRe-GRCv1.4-DataDictionary.xlsx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4a01fb15-45b6-4f21-ac00-e052a5fb6b1d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GenRe_df = pd.read_excel(\n",
    "    \"https://assets.super.so/abd63e6d-92df-4a9f-b19c-3a65d59b8a34/files/4accef63-0436-4893-a00a-6ea0e8a55249/Pf-GenRe-GRCv1.4-Full-PublicRelease-20240514.xlsx\",\n",
    "    sheet_name=\"GRC\"\n",
    ")\n",
    "\n",
    "# make sample ID column name match Pf8\n",
    "GenRe_df.rename(columns={'SampleId': 'sample'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d6fd8d9-03ef-49c5-9d2a-b270dd1affff",
   "metadata": {},
   "source": [
    "## Merge GenRe Mekong and Pf8 data\n",
    "Sample IDs for samples that were analysed in Pf8 are not changed, i.e. we can merge GenRe and Pf8 on sample ID.  \n",
    "Retain only rows that have a sample ID that is found in GenRe and Pf8."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1c560ee2-5090-43f2-a6a9-687791c6a8cb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_df = pd.merge(GenRe_df, pf8_df, on=[\"sample\"], suffixes=('_GRMK','_pf8'), how=\"inner\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a989ef4-07e8-46ad-90a8-7e3db4d6e656",
   "metadata": {},
   "source": [
    "## Remove GenRe Agena data\n",
    "In the GenRe Mekong project, three techniques were used:\n",
    "- Agena\n",
    "- AmpSeqV1\n",
    "- AmpSeqV2\n",
    "\n",
    "These are recorded in the \"Process\" column of the GenRe Mekong data. Only AmpSeq data is relevant here because we are building a dataset for genotyping by sequencing. \"Agena\" refers to [\"Agena Bioscience MassARRAY\"](https://link.springer.com/protocol/10.1007/978-1-4939-6442-0_5), a technique based on PCR amplification of SNP regions that does not use sequencing and should thus not be used in this dataset.  \n",
    "\n",
    "Count samples by process used in GenRe Mekong and \"Sample type\" in Pf8.  The Pf8 sample type, according to the [Pf8 data release README](https://pf8-release.cog.sanger.ac.uk/Pf8_README.txt), refers to\n",
    "\"Amplification technology used on the sample (MDA, gDNA or sWGA)\", where \"gDNA\" means non-selective WGS, as opposed to selective sWGS. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d5ad7bbe-6480-4c5d-9132-3fc867449438",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Process</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Agena</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>3690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AmpSeqV1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AmpSeqV1</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>2730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Process Sample type  count\n",
       "0     Agena        sWGA   3690\n",
       "1  AmpSeqV1        gDNA     64\n",
       "2  AmpSeqV1        sWGA    558\n",
       "3  AmpSeqV2        sWGA   2730"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_df.groupby(['Process','Sample type']).size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d89019a1-b2e7-4c38-b3c1-71d336823831",
   "metadata": {},
   "source": [
    "-> 3690 samples in GenRe Mekong are based on Agena array, not on sequencing.  \n",
    "Remove those samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7fe5360f-3aa2-4955-b861-9cc31464fbf7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_df = genre_pf8_merged_df[ genre_pf8_merged_df['Process'] != \"Agena\" ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "57b156cd-8f7e-4681-be7b-2510e167daaa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Process</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AmpSeqV1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AmpSeqV1</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>2730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Process Sample type  count\n",
       "0  AmpSeqV1        gDNA     64\n",
       "1  AmpSeqV1        sWGA    558\n",
       "2  AmpSeqV2        sWGA   2730"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_df.groupby(['Process','Sample type']).size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b393a83-e1cc-47d7-9fdd-96c342253024",
   "metadata": {},
   "source": [
    "## Identify concordant drug-resistance calls and high-quality data\n",
    "Both projects, GenRe Mekong and Pf8, provide drug-resistance phenotype predictions.  the final gold-standard dataset should contain only samples that have the same phenotype predictions in both projects and are of high quality.  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc76211-4591-4fbe-a0d7-c0306e6c7b89",
   "metadata": {},
   "source": [
    "Identify columns that are found in both GenRe and Pf8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1650eb93-5e19-4b2c-a8f0-857a0cc7a145",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sample',\n",
       " 'Study',\n",
       " 'Year',\n",
       " 'Country',\n",
       " 'Artemisinin',\n",
       " 'Piperaquine',\n",
       " 'Mefloquine',\n",
       " 'Chloroquine',\n",
       " 'Pyrimethamine',\n",
       " 'Sulfadoxine',\n",
       " 'DHA-PPQ',\n",
       " 'AS-MQ']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(GenRe_df.columns.intersection( pf8_df.columns).values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bde1e446-1821-40fc-95e7-d48182751c4b",
   "metadata": {},
   "source": [
    "-> all except 'sample', 'Study', 'Year' and 'Country'are drug-resistance phenotypes.  \n",
    "Create a list of field names to compare."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9c016c3c-4027-48d6-8b5b-aae89aa361ae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "phenotypes = [\n",
    " 'Artemisinin',\n",
    " 'Piperaquine',\n",
    " 'Mefloquine',\n",
    " 'Chloroquine',\n",
    " 'Pyrimethamine',\n",
    " 'Sulfadoxine',\n",
    " 'DHA-PPQ',\n",
    " 'AS-MQ']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2627f9f7-1b25-4b65-998b-da6fccf513b6",
   "metadata": {},
   "source": [
    "Get counts for concordant vs. discordant phenotype calls in Pf8 and GenRe Mekong"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "30d9612e-16cd-46a4-8af2-8a3b1fb8b602",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>discordant_count</th>\n",
       "      <th>concordant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Artemisinin</td>\n",
       "      <td>832</td>\n",
       "      <td>2520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Piperaquine</td>\n",
       "      <td>1807</td>\n",
       "      <td>1545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mefloquine</td>\n",
       "      <td>1853</td>\n",
       "      <td>1499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chloroquine</td>\n",
       "      <td>413</td>\n",
       "      <td>2939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>480</td>\n",
       "      <td>2872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Sulfadoxine</td>\n",
       "      <td>478</td>\n",
       "      <td>2874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DHA-PPQ</td>\n",
       "      <td>1784</td>\n",
       "      <td>1568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AS-MQ</td>\n",
       "      <td>1834</td>\n",
       "      <td>1518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            drug  discordant_count  concordant_count\n",
       "0    Artemisinin               832              2520\n",
       "1    Piperaquine              1807              1545\n",
       "2     Mefloquine              1853              1499\n",
       "3    Chloroquine               413              2939\n",
       "4  Pyrimethamine               480              2872\n",
       "5    Sulfadoxine               478              2874\n",
       "6        DHA-PPQ              1784              1568\n",
       "7          AS-MQ              1834              1518"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "discordant_counts = []\n",
    "for p in phenotypes:\n",
    "    discordant_counts.append(\n",
    "        {\n",
    "            'drug': p, \n",
    "            'discordant_count': len(\n",
    "                genre_pf8_merged_seq_df[\n",
    "                    genre_pf8_merged_seq_df[ p + '_GRMK'] != genre_pf8_merged_seq_df[ p + '_pf8']\n",
    "                ]\n",
    "            ),\n",
    "            'concordant_count': len(\n",
    "                genre_pf8_merged_seq_df[\n",
    "                    genre_pf8_merged_seq_df[ p + '_GRMK'] == genre_pf8_merged_seq_df[ p + '_pf8']\n",
    "                ]\n",
    "            )\n",
    "    })\n",
    "pd.DataFrame(discordant_counts)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a510d0dc-98d5-480e-8e8d-00360044cf1d",
   "metadata": {},
   "source": [
    "-> the data shows that there are discordant as well as concordant calls for all drug resistance phenotypes, thus it is important to filter for samples that are concordant."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef297e01-b702-498f-acbb-8748b5a0ee30",
   "metadata": {},
   "source": [
    "Filter for rows that are concordant on all 8 phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0f9c59ff-27e6-4d92-af44-7d1d7e539606",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_conc_df = genre_pf8_merged_seq_df\n",
    "for p in phenotypes:\n",
    "    cols = [ p + '_GRMK', p + '_pf8' ]\n",
    "    genre_pf8_conc_df = genre_pf8_conc_df[\n",
    "        ( genre_pf8_conc_df[cols].notnull().all(axis='columns') ) &\n",
    "        ( genre_pf8_conc_df[cols[0]] == genre_pf8_conc_df[cols[1]] )\n",
    "    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17c4e3e5-fca7-4ec2-a792-436a527b945a",
   "metadata": {},
   "source": [
    "Quality filter: percent of callable bases > 85 and 'QC pass' is True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7eb29de4-2b9e-4518-be2e-d11bd2e4a242",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_conc_df = genre_pf8_conc_df[ \n",
    "    (genre_pf8_conc_df['% callable'] >= 85 ) &\n",
    "    (genre_pf8_conc_df['QC pass'] == True )\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0a811c9-30f2-4495-89db-2f9045f0c5bd",
   "metadata": {},
   "source": [
    "Number of rows after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0e44f560-527a-4029-a3c8-ff70b83325d6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "968"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genre_pf8_conc_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f4b7452-1d78-4aa7-b8b2-c99dad5a338a",
   "metadata": {},
   "source": [
    "Reduce the dataset to a smaller number of relevant columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "79566537-e06b-44b2-9522-df88cf7fb40f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "keep_cols = [\n",
    "        'sample', 'Artemisinin_GRMK', 'Piperaquine_GRMK',\n",
    "       'Mefloquine_GRMK', 'Chloroquine_GRMK', 'Pyrimethamine_GRMK',\n",
    "       'Sulfadoxine_GRMK', 'DHA-PPQ_GRMK', 'AS-MQ_GRMK', 'Species', \n",
    "       'Chloroquine_pf8', 'Pyrimethamine_pf8',\n",
    "       'Sulfadoxine_pf8', 'Mefloquine_pf8', 'Artemisinin_pf8',\n",
    "       'Piperaquine_pf8', 'AS-MQ_pf8',\n",
    "       'DHA-PPQ_pf8', 'ENA', '% callable'\n",
    "]\n",
    "\n",
    "genre_pf8_conc_s_df = genre_pf8_conc_df[keep_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bed55f8a-654b-4dd8-be1d-cd24e58369ff",
   "metadata": {
    "tags": []
   },
   "source": [
    "Combine the phenotype calls into single column each (they are all concordant due to the filter used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8c740df8-51eb-47d6-9b65-ffff74b80fc9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for p in phenotypes:\n",
    "    # just to be safe\n",
    "    if not genre_pf8_conc_s_df[p + '_GRMK'].equals(genre_pf8_conc_s_df[p + '_pf8']):\n",
    "        raise Exception('phenotypes are not concordant but they should be if filtered correctly')\n",
    "    genre_pf8_conc_s_df = genre_pf8_conc_s_df.rename(columns={p + '_GRMK': p}, errors=\"raise\")\n",
    "    genre_pf8_conc_s_df = genre_pf8_conc_s_df.drop([ p + '_pf8' ], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e236a175-ed1b-4e3f-bfc4-475da4752f0e",
   "metadata": {},
   "source": [
    "Overview of final data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cae41e8d-44a6-4af0-bced-975395447c78",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>DHA-PPQ</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>Species</th>\n",
       "      <th>ENA</th>\n",
       "      <th>% callable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3730</th>\n",
       "      <td>RCN12025</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409305</td>\n",
       "      <td>89.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3731</th>\n",
       "      <td>RCN12026</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409312</td>\n",
       "      <td>85.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3733</th>\n",
       "      <td>RCN12028</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409373</td>\n",
       "      <td>87.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3734</th>\n",
       "      <td>RCN12031</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409314</td>\n",
       "      <td>87.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3735</th>\n",
       "      <td>RCN12032</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409311</td>\n",
       "      <td>88.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6975</th>\n",
       "      <td>RCN26767</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221085</td>\n",
       "      <td>87.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6976</th>\n",
       "      <td>RCN26775</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221086</td>\n",
       "      <td>87.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6987</th>\n",
       "      <td>RCN26820</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221102</td>\n",
       "      <td>89.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7021</th>\n",
       "      <td>RCN26932</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221141</td>\n",
       "      <td>86.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7025</th>\n",
       "      <td>RCN26938</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221132</td>\n",
       "      <td>89.21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>968 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample Artemisinin Piperaquine Mefloquine Chloroquine Pyrimethamine  \\\n",
       "3730  RCN12025   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "3731  RCN12026   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3733  RCN12028   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3734  RCN12031   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3735  RCN12032   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "...        ...         ...         ...        ...         ...           ...   \n",
       "6975  RCN26767   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "6976  RCN26775   Sensitive   Sensitive  Sensitive   Resistant     Resistant   \n",
       "6987  RCN26820   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "7021  RCN26932   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "7025  RCN26938   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "\n",
       "     Sulfadoxine    DHA-PPQ      AS-MQ Species         ENA  % callable  \n",
       "3730   Resistant  Sensitive  Sensitive      Pf  ERR3409305       89.14  \n",
       "3731   Resistant  Resistant  Sensitive      Pf  ERR3409312       85.83  \n",
       "3733   Resistant  Resistant  Sensitive      Pf  ERR3409373       87.23  \n",
       "3734   Resistant  Resistant  Sensitive      Pf  ERR3409314       87.04  \n",
       "3735   Resistant  Sensitive  Sensitive      Pf  ERR3409311       88.65  \n",
       "...          ...        ...        ...     ...         ...         ...  \n",
       "6975   Resistant  Sensitive  Sensitive      Pf  ERR7221085       87.92  \n",
       "6976   Sensitive  Sensitive  Sensitive      Pf  ERR7221086       87.30  \n",
       "6987   Resistant  Sensitive  Sensitive      Pf  ERR7221102       89.35  \n",
       "7021   Resistant  Sensitive  Sensitive      Pf  ERR7221141       86.03  \n",
       "7025   Resistant  Sensitive  Sensitive      Pf  ERR7221132       89.21  \n",
       "\n",
       "[968 rows x 12 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_conc_s_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "258ef86e-45a3-4ba8-b46a-3487883d7610",
   "metadata": {},
   "source": [
    "## Save results\n",
    "Save to csv file\n",
    "```Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv```  \n",
    "\n",
    "In summary: this table of 968 rows represents samples that have been sequenced by amplicon sequencing for GenRe Mekong and WGS for Pf8. All samples are marked as high-quality in Pf8 and they have conconrdant phenotype calls, inferred from genotype, in both projects.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e4fd520c-6075-443c-bddf-b6830283427d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_conc_s_df.to_csv('Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffc5fe1c-db61-42eb-a62d-11e99a53c153",
   "metadata": {},
   "source": [
    "## Add genotypes\n",
    "In the above steps, concordance was defined on high-level drug resistance phenotype calls.  \n",
    "Adding the underlying genotype calls will provide a lower-level truth dataset that can be used to validate genotype calls of a novel pipeline.  \n",
    "___Note___ that \"genotype\" calls are provided at amino acid level - not nucelotide level - in both studies. \n",
    "\n",
    "### Genes and haplotypes\n",
    "For the purpose of this gold-standard dataset, the intention is to identify samples that are fully concordant on their haplotype calls for important genes between Pf8 and GenRe Mekong, i.e. two different sequencing techinques and two pipelines have resulted in the same haplotype call.  \n",
    "\n",
    "This dataset will focus on 8 key drug resistance genes where drug resistance is determined by non-synonymous point mutations:\n",
    "- dhfr\n",
    "- crt\n",
    "- dhps\n",
    "- kelch13\n",
    "- mdr1\n",
    "- mdr2\n",
    "- fd\n",
    "- arps10\n",
    "\n",
    "This list of genes and \"core\" amino acid positions was compiled according to Table SM1 (Markers associated with drug resistance for P. falciparum) in the [SpotMalaria Platform Technical Notes and Methods](https://ngs.sanger.ac.uk/production/malaria/Resource/29/20200705-GenRe-04a-SpotMalaria-0.39.pdf). For Pf8, details about drug resistance classification can be found in in [this document](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf).  \n",
    "\n",
    "This results in the following list of drug-resistance haplotypes to be considered for the gold- standard dataset:\n",
    "\n",
    "| Gene | Amino Acid position |\n",
    "|---|---|\n",
    "| dhfr | 51, 59, 108, 164 |\n",
    "| crt | 72, ~~73~~\\*, 74, 75, 76, 326, 356 |\n",
    "| dhps | 436, 437, 540, 581, 613 |\n",
    "| kelch13 | 349-726 |\n",
    "| mdr1 | 86, 184, 1246 |\n",
    "| mdr2 | 484 |\n",
    "| fd | 193 |\n",
    "| arps10 | 127, 128 |\n",
    "\\* Position crt:73 is listed as a core position in the reference table but it is not provided in GenRe Mekong data and it is always WT in Pf8 data, hence this position will be ignored in the following code.  \n",
    "\n",
    "For more details on gene haplotypes in P. falciparum see the [Pf-HaploAtlas](https://www.malariagen.net/article/introducing-pf-haploatlas-a-new-app-to-track-malaria-parasite-mutations/). \n",
    "       "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f97089cf-3736-4dfb-9687-2a316d4eb3b0",
   "metadata": {},
   "source": [
    "### Load Pf8 genotype calls from data release\n",
    "The GenRe data loaded earlier does contain genotype data already."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "44396456-e86e-49ad-8ffa-3000ffea4dd3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_gt_df = pd.read_csv(\"https://pf8-release.cog.sanger.ac.uk/Pf8_drug_resistance_marker_genotypes.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01e86e1c-3b44-443c-963d-9ed75428895f",
   "metadata": {},
   "source": [
    "### Merge Pf8 genotypes \n",
    "Merge into into existing dataframe  \n",
    "```genre_pf8_merged_seq_df```  \n",
    "created above, which is the merged data for Pf8 and GenRe Mekong, with array data removed. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b15e53af-a105-4c49-acc4-86cd5ffbd7b4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_gt_df.rename(columns={'Sample': 'sample'}, inplace=True)\n",
    "\n",
    "genre_pf8_merged_seq_gt_df = pd.merge(\n",
    "    genre_pf8_merged_seq_df, \n",
    "    pf8_gt_df, \n",
    "    on=[\"sample\"], \n",
    "    suffixes=('_GRMK','_pf8'), \n",
    "    how=\"inner\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6fee5a6-340e-4d1d-82a0-47187820625f",
   "metadata": {},
   "source": [
    "The dataframe now contains Pf8 and GenRe Mekong genotypes for all samples that have been sequenced (excluding Array data) in both projects and can be linked via sample ID."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba3e1534-bf4d-42d8-b5dd-51202e785cfd",
   "metadata": {},
   "source": [
    "### Create new dataframe with amino acid genotype calls\n",
    "Create a list of gene/aa-positions from the above table and use it to comile a new data frame with amino acid calls from Pf8 and GenRe Mekong for these core positions. The kelch13 list needs to be built separately because the data is provided in separate column in GenRe Mekong and as a sigle column of comma-separated lists in Pf8.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2664f9c-256b-4f14-985a-423a07ab9df3",
   "metadata": {},
   "source": [
    "The new columns will have a unified naming scheme:  \n",
    "\"core_mutation_{GENE}_{AAPos}_\\[pf8/GMRK\\]\"\n",
    "\n",
    "The majority of genotypes can be extracted from the Pf8 and GenRe data by a simple pattern.  \n",
    "The respective pattern of genotype calls by gene and AA position is as follows:  \n",
    "__Pf8__: {GENE}\\_{AAPos} (example \"crt_75\\[N\\]\")  \n",
    "__GenRe Mekong__: Pf{GENE}:{AAPos} (example: \"PfCRT:75\")  \n",
    "\n",
    "Create a set of new column names and respective Pf8 and GenRe column names accordingly.  \n",
    "There are ___two excpetions___: \n",
    "1. kelch13: in both datasets, kelch13 data is provided as a column of lists of AA positions that need to be converted separately\n",
    "2. arps10: provided as AA pos 127-128 in Pf8 and just pos 127 in GenRe. Needs to be converted separately\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b2a0b271-8de9-444e-9bbb-4693bd29f66d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genes_pos = {\n",
    "    'dhfr': [51, 59, 108, 164],\n",
    "    'crt': [72, 74, 75, 76, 326, 356], # omitting pos 73, see above for explanation\n",
    "    'dhps': [436, 437, 540, 581, 613],\n",
    "    'mdr1': [86, 184, 1246],\n",
    "    'mdr2': [484],\n",
    "    'fd': [193]\n",
    "}\n",
    "\n",
    "gt_cols = {}\n",
    "for gene, pos_list in genes_pos.items():\n",
    "    gt_cols[ gene ] = {}\n",
    "    for pos in pos_list:\n",
    "        gt_cols[ gene ][ pos ] = {\n",
    "            'pf8_pattern': gene + '_' + str(pos) + '\\[[A-Z]\\]',\n",
    "            'GRMK_name': 'Pf' + gene.upper() + ':' + str(pos),\n",
    "            'new_col_basename': 'core_mutation_' + gene + '_' + str(pos)\n",
    "        }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aded4aa5-d782-488b-9734-c32cbf10ac32",
   "metadata": {},
   "source": [
    "### Add new genotype columns \n",
    "Add new genotype columns for the core mutation positions to dataframe   \n",
    "```genre_pf8_merged_seq_gt_df```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "95c121c3-12e2-4945-b4b3-fe5c68f01a92",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in gt_cols:\n",
    "    for pos in gt_cols[ gene ]:\n",
    "        \n",
    "        # genotype from Pf8\n",
    "        pf8_pattern = gt_cols[gene][pos]['pf8_pattern']\n",
    "        new_col_pf8 = gt_cols[gene][pos]['new_col_basename'] + '_pf8'\n",
    "        pf8_col = [ c for c in genre_pf8_merged_seq_gt_df if re.match(pf8_pattern, c) ]\n",
    "        if not pf8_col:\n",
    "            raise ValueError(f'could not find Pf8 column starting with \"{pf8_pattern}\"')\n",
    "        elif len(pf8_col) > 1:\n",
    "            raise ValueError(f'found more than one Pf8 column starting with \"{pf8_pattern}\":{\", \".join(pf8_col)}')\n",
    "        else:\n",
    "            genre_pf8_merged_seq_gt_df[new_col_pf8] = genre_pf8_merged_seq_gt_df[pf8_col[0]].copy()\n",
    "            \n",
    "        # genotype from GenRe\n",
    "        genre_pattern = gt_cols[gene][pos]['GRMK_name']\n",
    "        new_col_genre = gt_cols[gene][pos]['new_col_basename'] + '_GRMK'\n",
    "        if genre_pattern in genre_pf8_merged_seq_gt_df:\n",
    "            genre_pf8_merged_seq_gt_df[new_col_genre] = genre_pf8_merged_seq_gt_df[genre_pattern].copy()\n",
    "        else:\n",
    "            raise ValueError(f'could not find GenRe column \"{genre_pattern}\"')\n",
    "                                                                         "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c812509a-8c40-4953-acda-99731ea726cf",
   "metadata": {
    "tags": []
   },
   "source": [
    "Add columns for arps10: use only position arsp10:127, which is provided as a column inf GenRe and needs to be extracted from column 'arps10_127-128\\[VD\\]' in Pf8."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4031cbdb-274d-40a7-baf1-73ff4e7ddd3d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "arps10_basename = 'core_mutation_arps10_127'\n",
    "genre_pf8_merged_seq_gt_df[ arps10_basename + '_GRMK'] = genre_pf8_merged_seq_gt_df['PfARPS10:127'].copy()\n",
    "genre_pf8_merged_seq_gt_df[ arps10_basename + '_pf8'] = genre_pf8_merged_seq_gt_df['arps10_127-128[VD]'].astype(str).str[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "184395e7-5bfb-48aa-b948-c612874257e7",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Add columns for kelch13\n",
    "Add columns for all kelch13 AA positions for which data exists in at least one sample in both projects.  \n",
    "__Pf8__:  \n",
    "    _column_: 'kelch13_349-726_ns_changes'  \n",
    "    _content_: comma separated list of haplotypes at Kelch13 positions 349-726. Each haplotype contains one or more non-synonymous variations, separateed by \"/\". Homozygous mutations are shown in upper case and heterozygous in lower case.   Details are [here (heading \"Pf8_drug_resistance_marker_genotypes.tsv\"](https://pf8-release.cog.sanger.ac.uk/Pf8_README.txt)  \n",
    "    \n",
    "__GenRe__:  \n",
    "    _column_: 'Pfkelch13'  \n",
    "    _content_: Comma-separated list of amino acid mutations. ___NOTE___ this differs from Pf8. In GenRe, the comma delimits mutations in the same haplotype, while in Pf8 (see above) a comma delimts haplotypes. There is no disctiontion between homo- and heterozygous mutations.  For details see the [GenRe Mekong data dictionary](https://github.com/GenRe-Mekong/Documents/blob/e331501b8e93752221ebd7ae72e5eddd3e6e924d/Pf-GenRe-GRCv1.4-DataDictionary.xlsx)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15d08078-852a-4f15-89b7-5cb66871c089",
   "metadata": {},
   "source": [
    "Show the unique contents of the respective kelch13 columns in Pf8 and GenRe Mekong: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7909f47f-a8f1-4ef8-9b53-25a83702d37c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['C580Y', nan, 'c580y,p553l*', '-', 'c580y,v445i/c580y', 'c580y',\n",
       "       'R561H', 'F446I', 'c580y,r561h/c580y', '!*', 'c580y,a578d/c580y',\n",
       "       'c580y,p419s/c580y', 'P553L', 'R539T', 'Y493H', 'e668k', 'g625e',\n",
       "       'g718s', 'e362k', 'A578S', 'e567k', '!', 'd353n', 'd399n', 'r515k',\n",
       "       'c580y,t573i/c580y', 'p570l', 'c580y,f506y/c580y', 'p615l',\n",
       "       'G453S', 'd584n', 'd397n', 'c580y,p574l/c580y',\n",
       "       'e401k/c580y,c580y', 'G357D', 'C580Y/E705K', 'G449D', 'r404k',\n",
       "       'c580y,a564t/c580y', 'c580y,c580y/q654h', 'c580y,c580y/i723v',\n",
       "       'c580y,c580y/p615t', 'c580y,y456f/c580y', 't474i',\n",
       "       'c580y,d512n/c580y', 'r539t', 'r539t,r539t/c580f'], dtype=object)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_gt_df['kelch13_349-726_ns_changes'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5765825e-8a47-4934-88ca-e1b3203420e4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['C580Y', 'WT,A432S,C580Y', 'WT', 'WT,P553L,C580Y', '-', 'R561H',\n",
       "       'WT,C469.,C580Y', 'WT,P419S,C580Y', 'P553L', 'WT,C580Y', 'R539T',\n",
       "       'Y493H', 'WT,G453S', 'WT,G545R', 'WT,D464N', 'WT,M351I', 'A578S',\n",
       "       'WT,E567K', 'WT,F662Y', 'WT,G548V', 'WT,K658R', 'WT,D353N',\n",
       "       'WT,D399N', 'WT,R515K', 'WT,E431K', 'WT,P615L', 'WT,G453S,W518.',\n",
       "       'WT,D397N', 'WT,P574L,C580Y', 'WT,G538D,C580Y', 'WT,G358V,C580Y',\n",
       "       'WT,C580Y,N599D', 'WT,C580Y,Q654R', 'WT,S549P,C580Y', 'WT,D373G',\n",
       "       'WT,S485G,C580Y', 'WT,G496D,C580Y', 'WT,C580Y,R622K',\n",
       "       'WT,I540V,C580Y', 'WT,C580Y,D648G', 'WT,C580Y,F656L',\n",
       "       'WT,E401K,C580Y', 'WT,D547G', 'WT,F442S,C580Y', 'WT,R471S,C580Y',\n",
       "       'WT,G357D,C580Y', 'WT,C542R', 'WT,L444S,C580Y', 'WT,C580Y,G595D',\n",
       "       'WT,I376M,C580Y', 'WT,C580Y,E612.', 'WT,W565.,C580Y',\n",
       "       'WT,H384R,C580Y', 'WT,G449D,C580Y', 'WT,A557V', 'WT,R404K',\n",
       "       'WT,C542Y,C580Y', 'WT,C580Y,S649P', 'WT,F434S,C580Y',\n",
       "       'WT,Y493H,S550P,C580Y', 'WT,K378E,C580Y', 'WT,A481P',\n",
       "       'WT,C580Y,Q654H', 'WT,Y456F,C580Y', 'P419S,C580Y', 'T363K,C580Y',\n",
       "       'WT,W470R', 'Y502H,C580Y', 'WT,S364P', 'C580Y,R597I',\n",
       "       'E401K,C580Y', 'C580Y,S600F', 'WT,N609D', 'WT,T350A,L488F',\n",
       "       'WT,V566I', 'WT,T474I,E605K', 'WT,Q371R', 'C580Y,D680G',\n",
       "       'WT,T363K', 'WT,R471C', 'WT,Y541F', 'WT,D512N,C580Y', 'WT,R539T',\n",
       "       'WT,A504S', 'WT,G639D', 'R539T,P615T'], dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_gt_df['Pfkelch13'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f3acdf7-cf41-48ed-aa18-a12cc32e7b0d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create individual kelch13 mutation columns\n",
    "For each individual mutation found in any sample in any project, create a column with a name following this schema:  \n",
    "```core_mutation_kelch13:{MUTATION}_{pf8/GRMK}```  \n",
    "MUTATION is of the format ```{WT allele}{Pos}{mutant allele}```   \n",
    "The column contains boolean values where True indicates that this mutation is present in a given sample. Due to the differences in data provided between the two projects, no resolution at haplotype level is provided and no information about zygosity is recorded. Consequently, the \"WT\" allele recorded in some samples in GenRe Mekong is ignored. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "36079b97-30b2-4743-8ddf-3558675d629f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "kelch13_mutations = set()\n",
    "\n",
    "# Pf8\n",
    "for mut_list in genre_pf8_merged_seq_gt_df['kelch13_349-726_ns_changes'].unique():\n",
    "    mut_list = str(mut_list)\n",
    "    if mut_list and re.search(r'[1-9]', mut_list):\n",
    "        mut_list = mut_list.replace('/',',') # treat all mutations individually, regardless of haplotype grouping\n",
    "        mut_list = mut_list.replace('*','')\n",
    "        for mut in mut_list.split(','):\n",
    "            kelch13_mutations.add(mut.upper())\n",
    "            \n",
    "# GenRe Mekong\n",
    "for mut_list in genre_pf8_merged_seq_gt_df['Pfkelch13'].unique():\n",
    "    mut_list = str(mut_list)\n",
    "    if mut_list and re.search(r'[1-9]', mut_list):\n",
    "        for mut in mut_list.split(','):\n",
    "            if mut != 'WT':\n",
    "                kelch13_mutations.add(mut.upper())\n",
    "\n",
    "# add the new columns\n",
    "genre_pf8_merged_seq_gt_df = genre_pf8_merged_seq_gt_df.copy()\n",
    "for mut in kelch13_mutations:\n",
    "    base_name = 'core_mutation_kelch13:' + mut\n",
    "    genre_pf8_merged_seq_gt_df[base_name + '_pf8'] = genre_pf8_merged_seq_gt_df['kelch13_349-726_ns_changes'].str.contains(mut, case=False)\n",
    "    genre_pf8_merged_seq_gt_df[base_name + '_GRMK'] = genre_pf8_merged_seq_gt_df['Pfkelch13'].str.contains(mut, case=False)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e70bfe4-9425-4b16-a32c-721f272288a7",
   "metadata": {},
   "source": [
    "### Identify concordant samples\n",
    "Filter for samples that are fully concordant on all core mutation genotype calls, i.e. all columns added with the \"core_mutation_\" prefix above.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6339e484-22cc-46f4-b1a0-34c37efe3d13",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def is_concordant(row, *core_mut_col_basenames):\n",
    "    \"\"\"\n",
    "    Return True if all 'core_mutation' columns match between \n",
    "    Pf8 and GenRe Mekong\n",
    "    \"\"\"\n",
    "    for col in core_mut_col_basenames:\n",
    "        if row[col + '_pf8'] != row[col + '_GRMK']:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "regex = re.compile(r'^(core_mutation_.+)_(pf8|GRMK)$')\n",
    "core_mut_col_basenames = set(\n",
    "    [ regex.sub(r'\\1', x) for x in genre_pf8_merged_seq_gt_df.columns.values if regex.match(x) ]\n",
    ")\n",
    "genre_pf8_merged_seq_gt_df = genre_pf8_merged_seq_gt_df.copy()\n",
    "genre_pf8_merged_seq_gt_df['fully_concordant_gt'] = genre_pf8_merged_seq_gt_df.apply(is_concordant, axis=1, args=core_mut_col_basenames)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "927fa9b9-5931-45b0-8cc5-ec0b6fee3309",
   "metadata": {},
   "source": [
    "Count fully concordant (on all genotypes reported) vs. not fully concordant samples across the two projects:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9dd4475b-991f-4e50-a64d-aee4b65434bf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fully_concordant_gt</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>1572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>1394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   fully_concordant_gt  count\n",
       "0                False   1572\n",
       "1                 True   1394"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_gt_df.groupby( ['fully_concordant_gt']).size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44ca7bb6-23ad-4764-abaf-807d295cdff6",
   "metadata": {},
   "source": [
    "47% of samples are fully concordant on all drug-reistance \"core mutation\" genotypes between Pf8 (WGS) and GenRe Mekong (AmpSeq)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f680f155-accf-44ea-8516-cbc787cdb588",
   "metadata": {},
   "source": [
    "### Create a filtered dataset containing only fully concordant samples\n",
    "In addition, apply filters on Pf8 quality, as done above for phenotypes (%callable >=85 and QC pass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "63a5407c-9d79-4376-8362-51753a6e39f7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_df = genre_pf8_merged_seq_gt_df[\n",
    "    (genre_pf8_merged_seq_gt_df['fully_concordant_gt'] == True ) &\n",
    "    (genre_pf8_merged_seq_gt_df['% callable'] >= 85 ) &\n",
    "    (genre_pf8_merged_seq_gt_df['QC pass'] == True )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a41d5774-56ab-4ef6-aff9-cb1acab44759",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "976"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genre_pf8_merged_seq_gt_concord_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab193a8e-b2d7-4e68-b323-547400a2072b",
   "metadata": {},
   "source": [
    "There are 976 high-quality samples where all reported drug-resitance genotype positions are fully concordant between Pf8 and GenRe Mekong."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2782f187-6467-475e-9583-48a9d32f9a1c",
   "metadata": {},
   "source": [
    "### Save file of fully concordant samples\n",
    "Keep only relevant columns. As all samples are fully concordant, keep only one column per core mutation and drop the pf8/GRMK suffix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "79fb2e2d-98a4-41c2-8e30-91acb68e1016",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "keep_cols = ['sample', 'Source', 'Process', \n",
    "    'TimePoint', 'Country_GRMK', 'Species',\n",
    "    'fully_concordant_gt']\n",
    "\n",
    "core_mut_cols = [x for x in genre_pf8_merged_seq_gt_concord_df.columns.values if x.startswith('core_mutation')]\n",
    "keep_cols.extend( core_mut_cols )\n",
    "genre_pf8_merged_seq_gt_concord_small_df = genre_pf8_merged_seq_gt_concord_df[keep_cols]\n",
    "genre_pf8_merged_seq_gt_concord_small_df.rename({'Country_GRMK':'Country'})\n",
    "\n",
    "regex = re.compile(r'^(core_mutation_.+)_(pf8|GRMK)$')\n",
    "for col in core_mut_cols:\n",
    "    col_basename = regex.sub(r'\\1', col)\n",
    "    genre_pf8_merged_seq_gt_concord_small_df[col_basename] = genre_pf8_merged_seq_gt_concord_small_df[col].copy()\n",
    "    genre_pf8_merged_seq_gt_concord_small_df.drop(col, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0de529f7-1eb7-418c-b30a-078d8b541b0c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_small_df.to_csv('Pf8-GenReMekong_fully_concordant_genotype_high_quality_samples.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dce7e08-fc81-46a0-bd4a-dc112d7543b2",
   "metadata": {},
   "source": [
    "### Extract distinct genotype patterns and representative samples\n",
    "First, get a count of different patterns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a03c1d4b-bd88-4075-9047-2e3fff1785e9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>core_mutation_dhfr_51</th>\n",
       "      <th>core_mutation_dhfr_59</th>\n",
       "      <th>core_mutation_dhfr_108</th>\n",
       "      <th>core_mutation_dhfr_164</th>\n",
       "      <th>core_mutation_crt_72</th>\n",
       "      <th>core_mutation_crt_74</th>\n",
       "      <th>core_mutation_crt_75</th>\n",
       "      <th>core_mutation_crt_76</th>\n",
       "      <th>core_mutation_crt_326</th>\n",
       "      <th>core_mutation_crt_356</th>\n",
       "      <th>...</th>\n",
       "      <th>core_mutation_kelch13:S550P</th>\n",
       "      <th>core_mutation_kelch13:D464N</th>\n",
       "      <th>core_mutation_kelch13:G358V</th>\n",
       "      <th>core_mutation_kelch13:E362K</th>\n",
       "      <th>core_mutation_kelch13:E431K</th>\n",
       "      <th>core_mutation_kelch13:P615L</th>\n",
       "      <th>core_mutation_kelch13:R471S</th>\n",
       "      <th>core_mutation_kelch13:G625E</th>\n",
       "      <th>core_mutation_kelch13:N599D</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I,L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I,L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>N</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>M</td>\n",
       "      <td>N</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41 rows × 117 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   core_mutation_dhfr_51 core_mutation_dhfr_59 core_mutation_dhfr_108  \\\n",
       "0                      I                     R                      N   \n",
       "1                      I                     R                      N   \n",
       "2                      I                     R                      N   \n",
       "3                      I                     R                      N   \n",
       "4                      I                     R                      N   \n",
       "5                      I                     R                      N   \n",
       "6                      I                     R                      N   \n",
       "7                      I                     R                      N   \n",
       "8                      I                     R                      N   \n",
       "9                      I                     R                      N   \n",
       "10                     I                     R                      N   \n",
       "11                     I                     R                      N   \n",
       "12                     I                     R                      N   \n",
       "13                     I                     R                      N   \n",
       "14                     I                     R                      N   \n",
       "15                     I                     R                      N   \n",
       "16                     I                     R                      N   \n",
       "17                     I                     R                      N   \n",
       "18                     I                     R                      N   \n",
       "19                     I                     R                      N   \n",
       "20                     I                     R                      N   \n",
       "21                     I                     R                      N   \n",
       "22                     I                     R                      N   \n",
       "23                     I                     R                      N   \n",
       "24                     I                     R                      N   \n",
       "25                     I                     R                      N   \n",
       "26                     I                     R                      N   \n",
       "27                     I                     R                      N   \n",
       "28                     I                     R                      N   \n",
       "29                     I                     R                      N   \n",
       "30                     I                     R                      N   \n",
       "31                     I                     R                      N   \n",
       "32                     I                     R                      N   \n",
       "33                     I                     R                      N   \n",
       "34                     I                     R                      N   \n",
       "35                     I                     R                      N   \n",
       "36                     I                     R                      N   \n",
       "37                     I                     R                      N   \n",
       "38                     I                     R                      N   \n",
       "39                     I                     R                      N   \n",
       "40                     N                     R                      N   \n",
       "\n",
       "   core_mutation_dhfr_164 core_mutation_crt_72 core_mutation_crt_74  \\\n",
       "0                       I                    C                    I   \n",
       "1                       I                    C                    I   \n",
       "2                       I                    C                    I   \n",
       "3                       I                    C                    I   \n",
       "4                       I                    C                    I   \n",
       "5                       I                    C                    I   \n",
       "6                       I                    C                    I   \n",
       "7                       I                    C                    I   \n",
       "8                       I                    C                    I   \n",
       "9                       I                    C                    I   \n",
       "10                      I                    C                    I   \n",
       "11                      I                    C                    I   \n",
       "12                      I                    C                    I   \n",
       "13                      I                    C                    I   \n",
       "14                      I                    C                    I   \n",
       "15                      I                    C                    I   \n",
       "16                      I                    C                    I   \n",
       "17                      I                    C                    I   \n",
       "18                      I                    C                    I   \n",
       "19                    I,L                    C                    I   \n",
       "20                    I,L                    C                    I   \n",
       "21                      L                    C                    I   \n",
       "22                      L                    C                    I   \n",
       "23                      L                    C                    I   \n",
       "24                      L                    C                    I   \n",
       "25                      L                    C                    I   \n",
       "26                      L                    C                    I   \n",
       "27                      L                    C                    I   \n",
       "28                      L                    C                    I   \n",
       "29                      L                    C                    I   \n",
       "30                      L                    C                    I   \n",
       "31                      L                    C                    I   \n",
       "32                      L                    C                    I   \n",
       "33                      L                    C                    I   \n",
       "34                      L                    C                    I   \n",
       "35                      L                    C                    I   \n",
       "36                      L                    C                    I   \n",
       "37                      L                    C                    I   \n",
       "38                      L                    C                    I   \n",
       "39                      L                    C                    I   \n",
       "40                      I                    C                    M   \n",
       "\n",
       "   core_mutation_crt_75 core_mutation_crt_76 core_mutation_crt_326  \\\n",
       "0                     D                    T                     N   \n",
       "1                     D                    T                     N   \n",
       "2                     E                    T                     N   \n",
       "3                     E                    T                     N   \n",
       "4                     E                    T                     S   \n",
       "5                     E                    T                     S   \n",
       "6                     E                    T                     S   \n",
       "7                     E                    T                     S   \n",
       "8                     E                    T                     S   \n",
       "9                     E                    T                     S   \n",
       "10                    E                    T                     S   \n",
       "11                    E                    T                     S   \n",
       "12                    E                    T                     S   \n",
       "13                    E                    T                     S   \n",
       "14                    E                    T                     S   \n",
       "15                    E                    T                     S   \n",
       "16                    E                    T                     S   \n",
       "17                    E                    T                     S   \n",
       "18                    E                    T                     S   \n",
       "19                    E                    T                     S   \n",
       "20                    E                    T                     S   \n",
       "21                    D                    T                     N   \n",
       "22                    E                    T                     S   \n",
       "23                    E                    T                     S   \n",
       "24                    E                    T                     S   \n",
       "25                    E                    T                     S   \n",
       "26                    E                    T                     S   \n",
       "27                    E                    T                     S   \n",
       "28                    E                    T                     S   \n",
       "29                    E                    T                     S   \n",
       "30                    E                    T                     S   \n",
       "31                    E                    T                     S   \n",
       "32                    E                    T                     S   \n",
       "33                    E                    T                     S   \n",
       "34                    E                    T                     S   \n",
       "35                    E                    T                     S   \n",
       "36                    E                    T                     S   \n",
       "37                    E                    T                     S   \n",
       "38                    E                    T                     S   \n",
       "39                    E                    T                     S   \n",
       "40                    N                    K                     N   \n",
       "\n",
       "   core_mutation_crt_356  ... core_mutation_kelch13:S550P  \\\n",
       "0                      I  ...                       False   \n",
       "1                      I  ...                       False   \n",
       "2                      I  ...                       False   \n",
       "3                      I  ...                       False   \n",
       "4                      T  ...                       False   \n",
       "5                      T  ...                       False   \n",
       "6                      T  ...                       False   \n",
       "7                      T  ...                       False   \n",
       "8                      T  ...                       False   \n",
       "9                      T  ...                       False   \n",
       "10                     T  ...                       False   \n",
       "11                     T  ...                       False   \n",
       "12                     T  ...                       False   \n",
       "13                     T  ...                       False   \n",
       "14                     T  ...                       False   \n",
       "15                     T  ...                       False   \n",
       "16                     T  ...                       False   \n",
       "17                     T  ...                       False   \n",
       "18                     T  ...                       False   \n",
       "19                     T  ...                       False   \n",
       "20                     T  ...                       False   \n",
       "21                     I  ...                       False   \n",
       "22                     T  ...                       False   \n",
       "23                     T  ...                       False   \n",
       "24                     T  ...                       False   \n",
       "25                     T  ...                       False   \n",
       "26                     T  ...                       False   \n",
       "27                     T  ...                       False   \n",
       "28                     T  ...                       False   \n",
       "29                     T  ...                       False   \n",
       "30                     T  ...                       False   \n",
       "31                     T  ...                       False   \n",
       "32                     T  ...                       False   \n",
       "33                     T  ...                       False   \n",
       "34                     T  ...                       False   \n",
       "35                     T  ...                       False   \n",
       "36                     T  ...                       False   \n",
       "37                     T  ...                       False   \n",
       "38                     T  ...                       False   \n",
       "39                     T  ...                       False   \n",
       "40                     I  ...                       False   \n",
       "\n",
       "   core_mutation_kelch13:D464N core_mutation_kelch13:G358V  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "34                       False                       False   \n",
       "35                       False                       False   \n",
       "36                       False                       False   \n",
       "37                       False                       False   \n",
       "38                       False                       False   \n",
       "39                       False                       False   \n",
       "40                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:E362K core_mutation_kelch13:E431K  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "34                       False                       False   \n",
       "35                       False                       False   \n",
       "36                       False                       False   \n",
       "37                       False                       False   \n",
       "38                       False                       False   \n",
       "39                       False                       False   \n",
       "40                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:P615L core_mutation_kelch13:R471S  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                         True                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "34                       False                       False   \n",
       "35                       False                       False   \n",
       "36                       False                       False   \n",
       "37                       False                       False   \n",
       "38                       False                       False   \n",
       "39                       False                       False   \n",
       "40                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:G625E core_mutation_kelch13:N599D count  \n",
       "0                        False                       False     1  \n",
       "1                        False                       False    10  \n",
       "2                        False                       False     1  \n",
       "3                        False                       False     3  \n",
       "4                        False                       False    22  \n",
       "5                        False                       False    11  \n",
       "6                        False                       False    27  \n",
       "7                        False                       False     2  \n",
       "8                        False                       False     1  \n",
       "9                        False                       False     1  \n",
       "10                       False                       False     1  \n",
       "11                       False                       False     5  \n",
       "12                       False                       False     1  \n",
       "13                       False                       False     1  \n",
       "14                       False                       False     1  \n",
       "15                       False                       False    22  \n",
       "16                       False                       False    21  \n",
       "17                       False                       False     1  \n",
       "18                       False                       False     5  \n",
       "19                       False                       False     3  \n",
       "20                       False                       False     1  \n",
       "21                       False                       False     2  \n",
       "22                       False                       False     1  \n",
       "23                       False                       False    10  \n",
       "24                       False                       False     1  \n",
       "25                       False                       False     3  \n",
       "26                       False                       False     5  \n",
       "27                       False                       False     1  \n",
       "28                       False                       False     1  \n",
       "29                       False                       False     1  \n",
       "30                       False                       False     1  \n",
       "31                       False                       False     1  \n",
       "32                       False                       False     1  \n",
       "33                       False                       False   789  \n",
       "34                       False                       False     1  \n",
       "35                       False                       False     6  \n",
       "36                       False                       False     1  \n",
       "37                       False                       False     7  \n",
       "38                       False                       False     1  \n",
       "39                       False                       False     1  \n",
       "40                       False                       False     1  \n",
       "\n",
       "[41 rows x 117 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "core_mut_cols = [x for x in genre_pf8_merged_seq_gt_concord_small_df.columns.values if x.startswith('core_mutation')]\n",
    "genre_pf8_merged_seq_gt_concord_small_df.groupby( core_mut_cols).size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67b72d7a-e91e-409b-b078-23c9e9f6a368",
   "metadata": {},
   "source": [
    "There are 40 distinct patterns of genotypes.\n",
    "Extract a list of sample IDs of representative samples for each pattern."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e046eeea-3aeb-4a69-9a5b-2bd992f0c535",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>core_mutation_dhfr_51</th>\n",
       "      <th>core_mutation_dhfr_59</th>\n",
       "      <th>core_mutation_dhfr_108</th>\n",
       "      <th>core_mutation_dhfr_164</th>\n",
       "      <th>core_mutation_crt_72</th>\n",
       "      <th>core_mutation_crt_74</th>\n",
       "      <th>core_mutation_crt_75</th>\n",
       "      <th>core_mutation_crt_76</th>\n",
       "      <th>core_mutation_crt_326</th>\n",
       "      <th>...</th>\n",
       "      <th>core_mutation_kelch13:D373G</th>\n",
       "      <th>core_mutation_kelch13:S550P</th>\n",
       "      <th>core_mutation_kelch13:D464N</th>\n",
       "      <th>core_mutation_kelch13:G358V</th>\n",
       "      <th>core_mutation_kelch13:E362K</th>\n",
       "      <th>core_mutation_kelch13:E431K</th>\n",
       "      <th>core_mutation_kelch13:P615L</th>\n",
       "      <th>core_mutation_kelch13:R471S</th>\n",
       "      <th>core_mutation_kelch13:G625E</th>\n",
       "      <th>core_mutation_kelch13:N599D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RCN13568</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RCN13560</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RCN15107</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RCN14580</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RCN13532</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RCN13481</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RCN01778</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RCN13501</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RCN01871</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RCN12873</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>RCN12710</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RCN13368</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>RCN22834</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RCN12729</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>RCN13528</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>RCN12087</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>RCN12033</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>RCN13585</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RCN12865</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RCN18593</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I,L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>RCN13407</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I,L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>RCN15343</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>RCN22842</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>RCN12796</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>RCN25190</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>RCN01859</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>RCN13329</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>RCN14642</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>RCN15274</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>RCN22897</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>RCN12862</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>RCN12855</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>RCN12480</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>RCN01775</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>RCN25946</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>RCN12789</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>RCN12854</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>RCN12484</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>RCN15359</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>RCN22852</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>RCN14011</td>\n",
       "      <td>N</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>M</td>\n",
       "      <td>N</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41 rows × 117 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample core_mutation_dhfr_51 core_mutation_dhfr_59  \\\n",
       "0   RCN13568                     I                     R   \n",
       "1   RCN13560                     I                     R   \n",
       "2   RCN15107                     I                     R   \n",
       "3   RCN14580                     I                     R   \n",
       "4   RCN13532                     I                     R   \n",
       "5   RCN13481                     I                     R   \n",
       "6   RCN01778                     I                     R   \n",
       "7   RCN13501                     I                     R   \n",
       "8   RCN01871                     I                     R   \n",
       "9   RCN12873                     I                     R   \n",
       "10  RCN12710                     I                     R   \n",
       "11  RCN13368                     I                     R   \n",
       "12  RCN22834                     I                     R   \n",
       "13  RCN12729                     I                     R   \n",
       "14  RCN13528                     I                     R   \n",
       "15  RCN12087                     I                     R   \n",
       "16  RCN12033                     I                     R   \n",
       "17  RCN13585                     I                     R   \n",
       "18  RCN12865                     I                     R   \n",
       "19  RCN18593                     I                     R   \n",
       "20  RCN13407                     I                     R   \n",
       "21  RCN15343                     I                     R   \n",
       "22  RCN22842                     I                     R   \n",
       "23  RCN12796                     I                     R   \n",
       "24  RCN25190                     I                     R   \n",
       "25  RCN01859                     I                     R   \n",
       "26  RCN13329                     I                     R   \n",
       "27  RCN14642                     I                     R   \n",
       "28  RCN15274                     I                     R   \n",
       "29  RCN22897                     I                     R   \n",
       "30  RCN12862                     I                     R   \n",
       "31  RCN12855                     I                     R   \n",
       "32  RCN12480                     I                     R   \n",
       "33  RCN01775                     I                     R   \n",
       "34  RCN25946                     I                     R   \n",
       "35  RCN12789                     I                     R   \n",
       "36  RCN12854                     I                     R   \n",
       "37  RCN12484                     I                     R   \n",
       "38  RCN15359                     I                     R   \n",
       "39  RCN22852                     I                     R   \n",
       "40  RCN14011                     N                     R   \n",
       "\n",
       "   core_mutation_dhfr_108 core_mutation_dhfr_164 core_mutation_crt_72  \\\n",
       "0                       N                      I                    C   \n",
       "1                       N                      I                    C   \n",
       "2                       N                      I                    C   \n",
       "3                       N                      I                    C   \n",
       "4                       N                      I                    C   \n",
       "5                       N                      I                    C   \n",
       "6                       N                      I                    C   \n",
       "7                       N                      I                    C   \n",
       "8                       N                      I                    C   \n",
       "9                       N                      I                    C   \n",
       "10                      N                      I                    C   \n",
       "11                      N                      I                    C   \n",
       "12                      N                      I                    C   \n",
       "13                      N                      I                    C   \n",
       "14                      N                      I                    C   \n",
       "15                      N                      I                    C   \n",
       "16                      N                      I                    C   \n",
       "17                      N                      I                    C   \n",
       "18                      N                      I                    C   \n",
       "19                      N                    I,L                    C   \n",
       "20                      N                    I,L                    C   \n",
       "21                      N                      L                    C   \n",
       "22                      N                      L                    C   \n",
       "23                      N                      L                    C   \n",
       "24                      N                      L                    C   \n",
       "25                      N                      L                    C   \n",
       "26                      N                      L                    C   \n",
       "27                      N                      L                    C   \n",
       "28                      N                      L                    C   \n",
       "29                      N                      L                    C   \n",
       "30                      N                      L                    C   \n",
       "31                      N                      L                    C   \n",
       "32                      N                      L                    C   \n",
       "33                      N                      L                    C   \n",
       "34                      N                      L                    C   \n",
       "35                      N                      L                    C   \n",
       "36                      N                      L                    C   \n",
       "37                      N                      L                    C   \n",
       "38                      N                      L                    C   \n",
       "39                      N                      L                    C   \n",
       "40                      N                      I                    C   \n",
       "\n",
       "   core_mutation_crt_74 core_mutation_crt_75 core_mutation_crt_76  \\\n",
       "0                     I                    D                    T   \n",
       "1                     I                    D                    T   \n",
       "2                     I                    E                    T   \n",
       "3                     I                    E                    T   \n",
       "4                     I                    E                    T   \n",
       "5                     I                    E                    T   \n",
       "6                     I                    E                    T   \n",
       "7                     I                    E                    T   \n",
       "8                     I                    E                    T   \n",
       "9                     I                    E                    T   \n",
       "10                    I                    E                    T   \n",
       "11                    I                    E                    T   \n",
       "12                    I                    E                    T   \n",
       "13                    I                    E                    T   \n",
       "14                    I                    E                    T   \n",
       "15                    I                    E                    T   \n",
       "16                    I                    E                    T   \n",
       "17                    I                    E                    T   \n",
       "18                    I                    E                    T   \n",
       "19                    I                    E                    T   \n",
       "20                    I                    E                    T   \n",
       "21                    I                    D                    T   \n",
       "22                    I                    E                    T   \n",
       "23                    I                    E                    T   \n",
       "24                    I                    E                    T   \n",
       "25                    I                    E                    T   \n",
       "26                    I                    E                    T   \n",
       "27                    I                    E                    T   \n",
       "28                    I                    E                    T   \n",
       "29                    I                    E                    T   \n",
       "30                    I                    E                    T   \n",
       "31                    I                    E                    T   \n",
       "32                    I                    E                    T   \n",
       "33                    I                    E                    T   \n",
       "34                    I                    E                    T   \n",
       "35                    I                    E                    T   \n",
       "36                    I                    E                    T   \n",
       "37                    I                    E                    T   \n",
       "38                    I                    E                    T   \n",
       "39                    I                    E                    T   \n",
       "40                    M                    N                    K   \n",
       "\n",
       "   core_mutation_crt_326  ... core_mutation_kelch13:D373G  \\\n",
       "0                      N  ...                       False   \n",
       "1                      N  ...                       False   \n",
       "2                      N  ...                       False   \n",
       "3                      N  ...                       False   \n",
       "4                      S  ...                       False   \n",
       "5                      S  ...                       False   \n",
       "6                      S  ...                       False   \n",
       "7                      S  ...                       False   \n",
       "8                      S  ...                       False   \n",
       "9                      S  ...                       False   \n",
       "10                     S  ...                       False   \n",
       "11                     S  ...                       False   \n",
       "12                     S  ...                       False   \n",
       "13                     S  ...                       False   \n",
       "14                     S  ...                       False   \n",
       "15                     S  ...                       False   \n",
       "16                     S  ...                       False   \n",
       "17                     S  ...                       False   \n",
       "18                     S  ...                       False   \n",
       "19                     S  ...                       False   \n",
       "20                     S  ...                       False   \n",
       "21                     N  ...                       False   \n",
       "22                     S  ...                       False   \n",
       "23                     S  ...                       False   \n",
       "24                     S  ...                       False   \n",
       "25                     S  ...                       False   \n",
       "26                     S  ...                       False   \n",
       "27                     S  ...                       False   \n",
       "28                     S  ...                       False   \n",
       "29                     S  ...                       False   \n",
       "30                     S  ...                       False   \n",
       "31                     S  ...                       False   \n",
       "32                     S  ...                       False   \n",
       "33                     S  ...                       False   \n",
       "34                     S  ...                       False   \n",
       "35                     S  ...                       False   \n",
       "36                     S  ...                       False   \n",
       "37                     S  ...                       False   \n",
       "38                     S  ...                       False   \n",
       "39                     S  ...                       False   \n",
       "40                     N  ...                       False   \n",
       "\n",
       "   core_mutation_kelch13:S550P core_mutation_kelch13:D464N  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "34                       False                       False   \n",
       "35                       False                       False   \n",
       "36                       False                       False   \n",
       "37                       False                       False   \n",
       "38                       False                       False   \n",
       "39                       False                       False   \n",
       "40                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:G358V core_mutation_kelch13:E362K  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "34                       False                       False   \n",
       "35                       False                       False   \n",
       "36                       False                       False   \n",
       "37                       False                       False   \n",
       "38                       False                       False   \n",
       "39                       False                       False   \n",
       "40                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:E431K core_mutation_kelch13:P615L  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                        True   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "34                       False                       False   \n",
       "35                       False                       False   \n",
       "36                       False                       False   \n",
       "37                       False                       False   \n",
       "38                       False                       False   \n",
       "39                       False                       False   \n",
       "40                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:R471S core_mutation_kelch13:G625E  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "34                       False                       False   \n",
       "35                       False                       False   \n",
       "36                       False                       False   \n",
       "37                       False                       False   \n",
       "38                       False                       False   \n",
       "39                       False                       False   \n",
       "40                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:N599D  \n",
       "0                        False  \n",
       "1                        False  \n",
       "2                        False  \n",
       "3                        False  \n",
       "4                        False  \n",
       "5                        False  \n",
       "6                        False  \n",
       "7                        False  \n",
       "8                        False  \n",
       "9                        False  \n",
       "10                       False  \n",
       "11                       False  \n",
       "12                       False  \n",
       "13                       False  \n",
       "14                       False  \n",
       "15                       False  \n",
       "16                       False  \n",
       "17                       False  \n",
       "18                       False  \n",
       "19                       False  \n",
       "20                       False  \n",
       "21                       False  \n",
       "22                       False  \n",
       "23                       False  \n",
       "24                       False  \n",
       "25                       False  \n",
       "26                       False  \n",
       "27                       False  \n",
       "28                       False  \n",
       "29                       False  \n",
       "30                       False  \n",
       "31                       False  \n",
       "32                       False  \n",
       "33                       False  \n",
       "34                       False  \n",
       "35                       False  \n",
       "36                       False  \n",
       "37                       False  \n",
       "38                       False  \n",
       "39                       False  \n",
       "40                       False  \n",
       "\n",
       "[41 rows x 117 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_gt_concord_patterns_df = genre_pf8_merged_seq_gt_concord_small_df.groupby( core_mut_cols)['sample'].min().reset_index()\n",
    "\n",
    "# move the samples column to first\n",
    "cols = list(genre_pf8_merged_seq_gt_concord_patterns_df.columns)\n",
    "cols = [cols[-1]] + cols[:-1]\n",
    "genre_pf8_merged_seq_gt_concord_patterns_df = genre_pf8_merged_seq_gt_concord_patterns_df[cols]\n",
    "genre_pf8_merged_seq_gt_concord_patterns_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b725a297-db30-43b5-ab0e-9bf3665298f9",
   "metadata": {},
   "source": [
    "### Save file\n",
    "Save the file of representative samples for every genotype pattern."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d6f6b2f4-10a1-4fa8-949f-8d8d6cb1800f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_patterns_df.to_csv('Pf8-GenReMekong_fully_concordant_genotype_patterns_representative_samples.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
