{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2e0e043e-1fe1-48e2-86ad-493abc1588b9",
   "metadata": {},
   "source": [
    "# Creating a gold-standard dataset from pf8 and GenRE MEKONG data\n",
    "\n",
    "The aim of this notebook is to create a gold-standard dataset for malaria genomics pipelines that predict drug resistance.\n",
    "\n",
    "The following resources will be used:\n",
    "1. [MalariaGEN Pf8](https://www.malariagen.net/data_package/open-dataset-plasmodium-falciparum-v80/)\n",
    "2. [GenRE Mekong](https://www.malariagen.net/resource/29/)\n",
    "\n",
    "## Rationale for selection of resources\n",
    "Pipelines are usually tailored to a particular sequences technology, such as whole-genome or amplicon sequencing. While Pf8 is based on (selective) Whole Genome Sequencing (sWGS) data, GenRE Mekong utilised the SPotMalaria panel for amplicon sequencing (download available at [https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx](https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx).\n",
    "Some GenRE Mekong samples have been re-analysed with sWGS for Pf8, thus providing data for the same sample from two different technologies and two independent resistance phenotype calls.\n",
    "We will use the intersect between Pf8 and GenRE Mekong as the gold-standard dataset.\n",
    "\n",
    "In Pf8, drug-resistance phenotypes are inferred based on criteria described in [this PDF](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf)\n",
    "\n",
    "## Goal\n",
    "THe notebook creates a csv file of sample IDs and metadata. The sample IDs can be used to retrieve the raw read data from public databases.  \n",
    "This can be done manually or by using a tool such as [nf-core fetchngs](https://nf-co.re/fetchngs/1.12.0).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7e72b18-623d-48c5-b585-c15c9524b4e4",
   "metadata": {},
   "source": [
    "## Set up environment\n",
    "All packages imported here must be installed locally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "947cd6a8-8192-4831-a821-4c1072c19799",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import re\n",
    "from warnings import simplefilter \n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "pd.options.mode.copy_on_write = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6cea270-e01d-45c3-8cbc-cdc2c0fe09c5",
   "metadata": {},
   "source": [
    "___custom ENA data helper module___\n",
    "If this import fails, check that the ENA_data_helper.py script exists in the same directory as this Jupyter notebook. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2f513a60-8979-46f2-b00f-ea2067e3d9d5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from ENA_data_helper import create_ena_data_frame, merge_ena_results_into_sample_data_genre_pf8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b76cd977-c763-4e6e-b8cd-dad98e7898b3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load Pf8 data\n",
    "Obtain sample metadata and drug resisance inference data from Pf8.\n",
    "Details can be found [here](https://www.malariagen.net/data_package/open-dataset-plasmodium-falciparum-v80/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee77eaca-480f-4054-92f4-034d436dc1ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_infer_resistance_df=pd.read_csv(\"https://pf8-release.cog.sanger.ac.uk/Pf8_inferred_resistance_status_classification.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d6645dee-44ba-4cba-a84d-fa60aec1c5c6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_samples_df=pd.read_csv(\"https://pf8-release.cog.sanger.ac.uk/metadata/Pf8_samples.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96193a2d-f629-4438-9237-5d3716f4f153",
   "metadata": {},
   "source": [
    "## Merge Pf8 resistance and sample metadata\n",
    "Also rename the sample column with lower case for consistency across data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "07200b79-bc54-4207-b6b9-f42506098629",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_samples_df.rename(columns={'Sample': 'sample'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "af63fb33-857b-47ce-96c4-931143ad50d3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_df = pd.merge(pf8_infer_resistance_df, pf8_samples_df, on=[\"sample\"], how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "74cbe906-aa62-4e06-8ba5-e7a47f90d90c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>SP (uncomplicated)</th>\n",
       "      <th>SP (IPTp)</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>...</th>\n",
       "      <th>Admin level 1 longitude</th>\n",
       "      <th>Year</th>\n",
       "      <th>ENA</th>\n",
       "      <th>All samples same case</th>\n",
       "      <th>Population</th>\n",
       "      <th>% callable</th>\n",
       "      <th>QC pass</th>\n",
       "      <th>Exclusion reason</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>Sample was in Pf7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081237</td>\n",
       "      <td>FP0008-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>82.48</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR1081238</td>\n",
       "      <td>FP0009-C</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>88.95</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889621</td>\n",
       "      <td>FP0010-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>87.01</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FP0011-CW</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889624</td>\n",
       "      <td>FP0011-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>86.95</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FP0012-CW</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.832345</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>ERR2889627</td>\n",
       "      <td>FP0012-CW</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>89.86</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33320</th>\n",
       "      <td>SPT92268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>ERR10940681</td>\n",
       "      <td>SPT92268</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>0.10</td>\n",
       "      <td>False</td>\n",
       "      <td>Low_coverage</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33321</th>\n",
       "      <td>SPT92269</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>ERR11009733</td>\n",
       "      <td>SPT92269</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>0.12</td>\n",
       "      <td>False</td>\n",
       "      <td>Low_coverage</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33322</th>\n",
       "      <td>SPT92270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3.618846</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>ERR11009737</td>\n",
       "      <td>SPT92270</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>0.01</td>\n",
       "      <td>False</td>\n",
       "      <td>Low_coverage</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33323</th>\n",
       "      <td>SPT94772</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.401559</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>ERR10789456</td>\n",
       "      <td>SPT94772</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>64.96</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33324</th>\n",
       "      <td>SPT94773</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.401559</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>ERR10789457</td>\n",
       "      <td>SPT94773</td>\n",
       "      <td>AF-W</td>\n",
       "      <td>90.53</td>\n",
       "      <td>True</td>\n",
       "      <td>Analysis_set</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33325 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample   Chloroquine Pyrimethamine   Sulfadoxine    Mefloquine  \\\n",
       "0       FP0008-C  Undetermined  Undetermined  Undetermined     Sensitive   \n",
       "1       FP0009-C     Resistant     Resistant     Sensitive     Sensitive   \n",
       "2      FP0010-CW  Undetermined     Resistant     Resistant     Sensitive   \n",
       "3      FP0011-CW  Undetermined     Resistant  Undetermined     Sensitive   \n",
       "4      FP0012-CW     Resistant     Resistant     Sensitive     Sensitive   \n",
       "...          ...           ...           ...           ...           ...   \n",
       "33320   SPT92268           NaN           NaN           NaN           NaN   \n",
       "33321   SPT92269           NaN           NaN           NaN           NaN   \n",
       "33322   SPT92270           NaN           NaN           NaN           NaN   \n",
       "33323   SPT94772     Resistant     Resistant     Resistant  Undetermined   \n",
       "33324   SPT94773     Resistant     Resistant     Resistant     Sensitive   \n",
       "\n",
       "        Artemisinin   Piperaquine SP (uncomplicated)  SP (IPTp)         AS-MQ  \\\n",
       "0         Sensitive     Sensitive          Sensitive  Sensitive     Sensitive   \n",
       "1         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "2         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "3         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "4         Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "...             ...           ...                ...        ...           ...   \n",
       "33320           NaN           NaN                NaN        NaN           NaN   \n",
       "33321           NaN           NaN                NaN        NaN           NaN   \n",
       "33322           NaN           NaN                NaN        NaN           NaN   \n",
       "33323  Undetermined  Undetermined          Resistant  Sensitive  Undetermined   \n",
       "33324     Sensitive     Sensitive          Resistant  Sensitive     Sensitive   \n",
       "\n",
       "       ... Admin level 1 longitude    Year          ENA All samples same case  \\\n",
       "0      ...               -9.832345  2014.0   ERR1081237              FP0008-C   \n",
       "1      ...               -9.832345  2014.0   ERR1081238              FP0009-C   \n",
       "2      ...               -9.832345  2014.0   ERR2889621             FP0010-CW   \n",
       "3      ...               -9.832345  2014.0   ERR2889624             FP0011-CW   \n",
       "4      ...               -9.832345  2014.0   ERR2889627             FP0012-CW   \n",
       "...    ...                     ...     ...          ...                   ...   \n",
       "33320  ...                3.618846  2000.0  ERR10940681              SPT92268   \n",
       "33321  ...                3.618846  2000.0  ERR11009733              SPT92269   \n",
       "33322  ...                3.618846  2000.0  ERR11009737              SPT92270   \n",
       "33323  ...              -16.401559  2017.0  ERR10789456              SPT94772   \n",
       "33324  ...              -16.401559  2017.0  ERR10789457              SPT94773   \n",
       "\n",
       "       Population  % callable  QC pass  Exclusion reason  Sample type  \\\n",
       "0            AF-W       82.48     True      Analysis_set         gDNA   \n",
       "1            AF-W       88.95     True      Analysis_set         gDNA   \n",
       "2            AF-W       87.01     True      Analysis_set         sWGA   \n",
       "3            AF-W       86.95     True      Analysis_set         sWGA   \n",
       "4            AF-W       89.86     True      Analysis_set         sWGA   \n",
       "...           ...         ...      ...               ...          ...   \n",
       "33320        AF-W        0.10    False      Low_coverage         sWGA   \n",
       "33321        AF-W        0.12    False      Low_coverage         sWGA   \n",
       "33322        AF-W        0.01    False      Low_coverage         sWGA   \n",
       "33323        AF-W       64.96     True      Analysis_set         sWGA   \n",
       "33324        AF-W       90.53     True      Analysis_set         sWGA   \n",
       "\n",
       "      Sample was in Pf7  \n",
       "0                  True  \n",
       "1                  True  \n",
       "2                  True  \n",
       "3                  True  \n",
       "4                  True  \n",
       "...                 ...  \n",
       "33320             False  \n",
       "33321             False  \n",
       "33322             False  \n",
       "33323             False  \n",
       "33324             False  \n",
       "\n",
       "[33325 rows x 27 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pf8_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f0df980-9e83-491f-b072-c5f6666861c8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load GenRe Mekong data\n",
    "Using this file from the GenRe Mekong webstie, release v1.4 14/05/24 (latest release):  \n",
    "\n",
    "https://assets.super.so/abd63e6d-92df-4a9f-b19c-3a65d59b8a34/files/4accef63-0436-4893-a00a-6ea0e8a55249/Pf-GenRe-GRCv1.4-Full-PublicRelease-20240514.xlsx  \n",
    "Tab: GRC\n",
    "\n",
    "Sample IDs for samples that were re-analysed in Pf8 are the same in both projects, i.e. it is possible to merge GenRe Mekong and Pf8 data on sample ID.  \n",
    "\n",
    "For more information on the data in this spreadsheet, consult the [GenRe Mekong data dictionary](https://github.com/GenRe-Mekong/Documents/blob/e331501b8e93752221ebd7ae72e5eddd3e6e924d/Pf-GenRe-GRCv1.4-DataDictionary.xlsx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4a01fb15-45b6-4f21-ac00-e052a5fb6b1d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GenRe_df = pd.read_excel(\n",
    "    \"https://assets.super.so/abd63e6d-92df-4a9f-b19c-3a65d59b8a34/files/4accef63-0436-4893-a00a-6ea0e8a55249/Pf-GenRe-GRCv1.4-Full-PublicRelease-20240514.xlsx\",\n",
    "    sheet_name=\"GRC\"\n",
    ")\n",
    "\n",
    "# make sample ID column name match Pf8\n",
    "GenRe_df.rename(columns={'SampleId': 'sample'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d6fd8d9-03ef-49c5-9d2a-b270dd1affff",
   "metadata": {},
   "source": [
    "## Merge GenRe Mekong and Pf8 data\n",
    "Sample IDs for samples that were analysed in Pf8 are not changed, i.e. we can merge GenRe and Pf8 on sample ID.  \n",
    "Retain only rows that have a sample ID that is found in GenRe and Pf8."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1c560ee2-5090-43f2-a6a9-687791c6a8cb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_df = pd.merge(GenRe_df, pf8_df, on=[\"sample\"], suffixes=('_GRMK','_pf8'), how=\"inner\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a989ef4-07e8-46ad-90a8-7e3db4d6e656",
   "metadata": {},
   "source": [
    "## Remove GenRe Agena data\n",
    "In the GenRe Mekong project, three techniques were used:\n",
    "- Agena\n",
    "- AmpSeqV1\n",
    "- AmpSeqV2\n",
    "\n",
    "These are recorded in the \"Process\" column of the GenRe Mekong data. Only AmpSeq data is relevant here because we are building a dataset for genotyping by sequencing. \"Agena\" refers to [\"Agena Bioscience MassARRAY\"](https://link.springer.com/protocol/10.1007/978-1-4939-6442-0_5), a technique based on PCR amplification of SNP regions that does not use sequencing and should thus not be used in this dataset.  \n",
    "\n",
    "Count samples by process used in GenRe Mekong and \"Sample type\" in Pf8.  The Pf8 sample type, according to the [Pf8 data release README](https://pf8-release.cog.sanger.ac.uk/Pf8_README.txt), refers to\n",
    "\"Amplification technology used on the sample (MDA, gDNA or sWGA)\", where \"gDNA\" means non-selective WGS, as opposed to selective sWGS. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d5ad7bbe-6480-4c5d-9132-3fc867449438",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Process</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Agena</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>3690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AmpSeqV1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AmpSeqV1</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>2730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Process Sample type  count\n",
       "0     Agena        sWGA   3690\n",
       "1  AmpSeqV1        gDNA     64\n",
       "2  AmpSeqV1        sWGA    558\n",
       "3  AmpSeqV2        sWGA   2730"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_df.groupby(['Process','Sample type']).size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d89019a1-b2e7-4c38-b3c1-71d336823831",
   "metadata": {},
   "source": [
    "-> 3690 samples in GenRe Mekong are based on Agena array, not on sequencing.  \n",
    "Remove those samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7fe5360f-3aa2-4955-b861-9cc31464fbf7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_df = genre_pf8_merged_df[ genre_pf8_merged_df['Process'] != \"Agena\" ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "57b156cd-8f7e-4681-be7b-2510e167daaa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Process</th>\n",
       "      <th>Sample type</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AmpSeqV1</td>\n",
       "      <td>gDNA</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AmpSeqV1</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>sWGA</td>\n",
       "      <td>2730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Process Sample type  count\n",
       "0  AmpSeqV1        gDNA     64\n",
       "1  AmpSeqV1        sWGA    558\n",
       "2  AmpSeqV2        sWGA   2730"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_df.groupby(['Process','Sample type']).size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b393a83-e1cc-47d7-9fdd-96c342253024",
   "metadata": {},
   "source": [
    "## Identify concordant drug-resistance calls and high-quality data\n",
    "Both projects, GenRe Mekong and Pf8, provide drug-resistance phenotype predictions.  the final gold-standard dataset should contain only samples that have the same phenotype predictions in both projects and are of high quality.  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc76211-4591-4fbe-a0d7-c0306e6c7b89",
   "metadata": {},
   "source": [
    "Identify columns that are found in both GenRe and Pf8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1650eb93-5e19-4b2c-a8f0-857a0cc7a145",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sample',\n",
       " 'Study',\n",
       " 'Year',\n",
       " 'Country',\n",
       " 'Artemisinin',\n",
       " 'Piperaquine',\n",
       " 'Mefloquine',\n",
       " 'Chloroquine',\n",
       " 'Pyrimethamine',\n",
       " 'Sulfadoxine',\n",
       " 'DHA-PPQ',\n",
       " 'AS-MQ']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(GenRe_df.columns.intersection( pf8_df.columns).values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bde1e446-1821-40fc-95e7-d48182751c4b",
   "metadata": {},
   "source": [
    "-> all except 'sample', 'Study', 'Year' and 'Country'are drug-resistance phenotypes.  \n",
    "Create a list of field names to compare."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9c016c3c-4027-48d6-8b5b-aae89aa361ae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "phenotypes = [\n",
    " 'Artemisinin',\n",
    " 'Piperaquine',\n",
    " 'Mefloquine',\n",
    " 'Chloroquine',\n",
    " 'Pyrimethamine',\n",
    " 'Sulfadoxine',\n",
    " 'DHA-PPQ',\n",
    " 'AS-MQ']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2627f9f7-1b25-4b65-998b-da6fccf513b6",
   "metadata": {},
   "source": [
    "Get counts for concordant vs. discordant phenotype calls in Pf8 and GenRe Mekong"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "30d9612e-16cd-46a4-8af2-8a3b1fb8b602",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>discordant_count</th>\n",
       "      <th>concordant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Artemisinin</td>\n",
       "      <td>832</td>\n",
       "      <td>2520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Piperaquine</td>\n",
       "      <td>1807</td>\n",
       "      <td>1545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mefloquine</td>\n",
       "      <td>1853</td>\n",
       "      <td>1499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chloroquine</td>\n",
       "      <td>413</td>\n",
       "      <td>2939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>480</td>\n",
       "      <td>2872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Sulfadoxine</td>\n",
       "      <td>478</td>\n",
       "      <td>2874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DHA-PPQ</td>\n",
       "      <td>1784</td>\n",
       "      <td>1568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AS-MQ</td>\n",
       "      <td>1834</td>\n",
       "      <td>1518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            drug  discordant_count  concordant_count\n",
       "0    Artemisinin               832              2520\n",
       "1    Piperaquine              1807              1545\n",
       "2     Mefloquine              1853              1499\n",
       "3    Chloroquine               413              2939\n",
       "4  Pyrimethamine               480              2872\n",
       "5    Sulfadoxine               478              2874\n",
       "6        DHA-PPQ              1784              1568\n",
       "7          AS-MQ              1834              1518"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "discordant_counts = []\n",
    "for p in phenotypes:\n",
    "    discordant_counts.append(\n",
    "        {\n",
    "            'drug': p, \n",
    "            'discordant_count': len(\n",
    "                genre_pf8_merged_seq_df[\n",
    "                    genre_pf8_merged_seq_df[ p + '_GRMK'] != genre_pf8_merged_seq_df[ p + '_pf8']\n",
    "                ]\n",
    "            ),\n",
    "            'concordant_count': len(\n",
    "                genre_pf8_merged_seq_df[\n",
    "                    genre_pf8_merged_seq_df[ p + '_GRMK'] == genre_pf8_merged_seq_df[ p + '_pf8']\n",
    "                ]\n",
    "            )\n",
    "    })\n",
    "pd.DataFrame(discordant_counts)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a510d0dc-98d5-480e-8e8d-00360044cf1d",
   "metadata": {},
   "source": [
    "-> the data shows that there are discordant as well as concordant calls for all drug resistance phenotypes, thus it is important to filter for samples that are concordant."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef297e01-b702-498f-acbb-8748b5a0ee30",
   "metadata": {},
   "source": [
    "Filter for rows that are concordant on all 8 phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0f9c59ff-27e6-4d92-af44-7d1d7e539606",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_conc_df = genre_pf8_merged_seq_df\n",
    "for p in phenotypes:\n",
    "    cols = [ p + '_GRMK', p + '_pf8' ]\n",
    "    genre_pf8_conc_df = genre_pf8_conc_df[\n",
    "        ( genre_pf8_conc_df[cols].notnull().all(axis='columns') ) &\n",
    "        ( genre_pf8_conc_df[cols[0]] == genre_pf8_conc_df[cols[1]] )\n",
    "    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17c4e3e5-fca7-4ec2-a792-436a527b945a",
   "metadata": {},
   "source": [
    "Quality filter: percent of callable bases > 85 and 'QC pass' is True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7eb29de4-2b9e-4518-be2e-d11bd2e4a242",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_conc_df = genre_pf8_conc_df[ \n",
    "    (genre_pf8_conc_df['% callable'] >= 85 ) &\n",
    "    (genre_pf8_conc_df['QC pass'] == True )\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0a811c9-30f2-4495-89db-2f9045f0c5bd",
   "metadata": {},
   "source": [
    "Number of rows after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0e44f560-527a-4029-a3c8-ff70b83325d6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "968"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genre_pf8_conc_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f4b7452-1d78-4aa7-b8b2-c99dad5a338a",
   "metadata": {},
   "source": [
    "Reduce the dataset to a smaller number of relevant columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "79566537-e06b-44b2-9522-df88cf7fb40f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "keep_cols = [\n",
    "        'sample', 'Artemisinin_GRMK', 'Piperaquine_GRMK',\n",
    "       'Mefloquine_GRMK', 'Chloroquine_GRMK', 'Pyrimethamine_GRMK',\n",
    "       'Sulfadoxine_GRMK', 'DHA-PPQ_GRMK', 'AS-MQ_GRMK', 'Species', \n",
    "       'Chloroquine_pf8', 'Pyrimethamine_pf8',\n",
    "       'Sulfadoxine_pf8', 'Mefloquine_pf8', 'Artemisinin_pf8',\n",
    "       'Piperaquine_pf8', 'AS-MQ_pf8',\n",
    "       'DHA-PPQ_pf8', 'ENA', '% callable'\n",
    "]\n",
    "\n",
    "genre_pf8_conc_s_df = genre_pf8_conc_df[keep_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bed55f8a-654b-4dd8-be1d-cd24e58369ff",
   "metadata": {
    "tags": []
   },
   "source": [
    "Combine the phenotype calls into single column each (they are all concordant due to the filter used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8c740df8-51eb-47d6-9b65-ffff74b80fc9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for p in phenotypes:\n",
    "    # just to be safe\n",
    "    if not genre_pf8_conc_s_df[p + '_GRMK'].equals(genre_pf8_conc_s_df[p + '_pf8']):\n",
    "        raise Exception('phenotypes are not concordant but they should be if filtered correctly')\n",
    "    genre_pf8_conc_s_df = genre_pf8_conc_s_df.rename(columns={p + '_GRMK': p}, errors=\"raise\")\n",
    "    genre_pf8_conc_s_df = genre_pf8_conc_s_df.drop([ p + '_pf8' ], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e236a175-ed1b-4e3f-bfc4-475da4752f0e",
   "metadata": {},
   "source": [
    "Overview of final data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cae41e8d-44a6-4af0-bced-975395447c78",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>DHA-PPQ</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>Species</th>\n",
       "      <th>ENA</th>\n",
       "      <th>% callable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3730</th>\n",
       "      <td>RCN12025</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409305</td>\n",
       "      <td>89.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3731</th>\n",
       "      <td>RCN12026</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409312</td>\n",
       "      <td>85.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3733</th>\n",
       "      <td>RCN12028</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409373</td>\n",
       "      <td>87.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3734</th>\n",
       "      <td>RCN12031</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409314</td>\n",
       "      <td>87.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3735</th>\n",
       "      <td>RCN12032</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409311</td>\n",
       "      <td>88.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6975</th>\n",
       "      <td>RCN26767</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221085</td>\n",
       "      <td>87.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6976</th>\n",
       "      <td>RCN26775</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221086</td>\n",
       "      <td>87.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6987</th>\n",
       "      <td>RCN26820</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221102</td>\n",
       "      <td>89.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7021</th>\n",
       "      <td>RCN26932</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221141</td>\n",
       "      <td>86.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7025</th>\n",
       "      <td>RCN26938</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221132</td>\n",
       "      <td>89.21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>968 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample Artemisinin Piperaquine Mefloquine Chloroquine Pyrimethamine  \\\n",
       "3730  RCN12025   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "3731  RCN12026   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3733  RCN12028   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3734  RCN12031   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3735  RCN12032   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "...        ...         ...         ...        ...         ...           ...   \n",
       "6975  RCN26767   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "6976  RCN26775   Sensitive   Sensitive  Sensitive   Resistant     Resistant   \n",
       "6987  RCN26820   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "7021  RCN26932   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "7025  RCN26938   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "\n",
       "     Sulfadoxine    DHA-PPQ      AS-MQ Species         ENA  % callable  \n",
       "3730   Resistant  Sensitive  Sensitive      Pf  ERR3409305       89.14  \n",
       "3731   Resistant  Resistant  Sensitive      Pf  ERR3409312       85.83  \n",
       "3733   Resistant  Resistant  Sensitive      Pf  ERR3409373       87.23  \n",
       "3734   Resistant  Resistant  Sensitive      Pf  ERR3409314       87.04  \n",
       "3735   Resistant  Sensitive  Sensitive      Pf  ERR3409311       88.65  \n",
       "...          ...        ...        ...     ...         ...         ...  \n",
       "6975   Resistant  Sensitive  Sensitive      Pf  ERR7221085       87.92  \n",
       "6976   Sensitive  Sensitive  Sensitive      Pf  ERR7221086       87.30  \n",
       "6987   Resistant  Sensitive  Sensitive      Pf  ERR7221102       89.35  \n",
       "7021   Resistant  Sensitive  Sensitive      Pf  ERR7221141       86.03  \n",
       "7025   Resistant  Sensitive  Sensitive      Pf  ERR7221132       89.21  \n",
       "\n",
       "[968 rows x 12 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_conc_s_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60a640ed-68df-4ea6-9a56-3130c374d969",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add ENA accessions and download links\n",
    "Using functions from the custom ENA data helper tool (file ENA_data_helper.py in the working directory), perform a search on ENA by sample ID and retrieve data for run accessions (unique accession for 1 sample on 1 run) and FTP download URLs for FASTQ files (read1/read2).  \n",
    "This step performs a live search on ENA and may not work if your internet connection is very slow or unstable."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4753eff9-a7d2-40bf-9aea-3055a59f6e4b",
   "metadata": {},
   "source": [
    "The first function performs a search for ENA data by sample IDs in sample data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0f7cc0a5-d22a-4e78-9966-a1bc7ec2c8d8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_conc_s_ena_result = create_ena_data_frame(genre_pf8_conc_s_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea5f918c-1b95-4b36-9509-da0a157e8b42",
   "metadata": {},
   "source": [
    "Second function merges the ENA search results (which contain more than one result per sample) back into the original sample data frame where sample ID is unique. Columns are added for the run accessions for all threee panels in GenRe as well as the WGS data for Pf8. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0981e915-a5e7-4e08-84bc-241f19fcdaf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "genre_pf8_conc_s_ena_df = merge_ena_results_into_sample_data_genre_pf8(sample_data=genre_pf8_conc_s_df, ena_result=genre_pf8_conc_s_ena_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "155e4490-4d77-44e4-91ca-d956978816fc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>DHA-PPQ</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>Species</th>\n",
       "      <th>ENA</th>\n",
       "      <th>...</th>\n",
       "      <th>ENA_acc_GenRe_GRC2</th>\n",
       "      <th>ENA_acc_GenRe_SPEC</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_1</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_2</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RCN12025</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409305</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388604</td>\n",
       "      <td>ERR14388605</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12026</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409312</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388607</td>\n",
       "      <td>ERR14388608</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12028</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409373</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388613</td>\n",
       "      <td>ERR14388614</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12031</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409314</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388622</td>\n",
       "      <td>ERR14388623</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12032</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409311</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388625</td>\n",
       "      <td>ERR14388626</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26767</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221085</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397547</td>\n",
       "      <td>ERR14397548</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26775</th>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221086</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397571</td>\n",
       "      <td>ERR14397572</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26820</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221102</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397664</td>\n",
       "      <td>ERR14397665</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26932</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221141</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397976</td>\n",
       "      <td>ERR14397977</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26938</th>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221132</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397991</td>\n",
       "      <td>ERR14397992</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>968 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Artemisinin Piperaquine Mefloquine Chloroquine Pyrimethamine  \\\n",
       "sample                                                                  \n",
       "RCN12025   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "RCN12026   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "RCN12028   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "RCN12031   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "RCN12032   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "...              ...         ...        ...         ...           ...   \n",
       "RCN26767   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "RCN26775   Sensitive   Sensitive  Sensitive   Resistant     Resistant   \n",
       "RCN26820   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "RCN26932   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "RCN26938   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "\n",
       "         Sulfadoxine    DHA-PPQ      AS-MQ Species         ENA  ...  \\\n",
       "sample                                                          ...   \n",
       "RCN12025   Resistant  Sensitive  Sensitive      Pf  ERR3409305  ...   \n",
       "RCN12026   Resistant  Resistant  Sensitive      Pf  ERR3409312  ...   \n",
       "RCN12028   Resistant  Resistant  Sensitive      Pf  ERR3409373  ...   \n",
       "RCN12031   Resistant  Resistant  Sensitive      Pf  ERR3409314  ...   \n",
       "RCN12032   Resistant  Sensitive  Sensitive      Pf  ERR3409311  ...   \n",
       "...              ...        ...        ...     ...         ...  ...   \n",
       "RCN26767   Resistant  Sensitive  Sensitive      Pf  ERR7221085  ...   \n",
       "RCN26775   Sensitive  Sensitive  Sensitive      Pf  ERR7221086  ...   \n",
       "RCN26820   Resistant  Sensitive  Sensitive      Pf  ERR7221102  ...   \n",
       "RCN26932   Resistant  Sensitive  Sensitive      Pf  ERR7221141  ...   \n",
       "RCN26938   Resistant  Sensitive  Sensitive      Pf  ERR7221132  ...   \n",
       "\n",
       "          ENA_acc_GenRe_GRC2 ENA_acc_GenRe_SPEC  \\\n",
       "sample                                            \n",
       "RCN12025         ERR14388604        ERR14388605   \n",
       "RCN12026         ERR14388607        ERR14388608   \n",
       "RCN12028         ERR14388613        ERR14388614   \n",
       "RCN12031         ERR14388622        ERR14388623   \n",
       "RCN12032         ERR14388625        ERR14388626   \n",
       "...                      ...                ...   \n",
       "RCN26767         ERR14397547        ERR14397548   \n",
       "RCN26775         ERR14397571        ERR14397572   \n",
       "RCN26820         ERR14397664        ERR14397665   \n",
       "RCN26932         ERR14397976        ERR14397977   \n",
       "RCN26938         ERR14397991        ERR14397992   \n",
       "\n",
       "                                 GenRe_GRC1_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   \n",
       "RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...   \n",
       "RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...   \n",
       "RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "\n",
       "                                 GenRe_GRC2_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   \n",
       "RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   \n",
       "RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...   \n",
       "RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...   \n",
       "\n",
       "                                 GenRe_SPEC_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...   \n",
       "RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...   \n",
       "RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...   \n",
       "RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...   \n",
       "\n",
       "                                        Pf8_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...   \n",
       "RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...   \n",
       "RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...   \n",
       "RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...   \n",
       "RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...   \n",
       "RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...   \n",
       "\n",
       "                                 GenRe_GRC1_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   \n",
       "RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...   \n",
       "RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...   \n",
       "RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "\n",
       "                                 GenRe_GRC2_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   \n",
       "RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   \n",
       "RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...   \n",
       "RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...   \n",
       "\n",
       "                                 GenRe_SPEC_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...   \n",
       "RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...   \n",
       "RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...   \n",
       "RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...   \n",
       "\n",
       "                                        Pf8_ENA_FASTQ_FTP_2  \n",
       "sample                                                       \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...  \n",
       "RCN12026  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...  \n",
       "RCN12028  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...  \n",
       "RCN12031  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...  \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...  \n",
       "...                                                     ...  \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...  \n",
       "RCN26775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...  \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...  \n",
       "RCN26932  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...  \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...  \n",
       "\n",
       "[968 rows x 23 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_conc_s_ena_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4cd088b-3438-4c2a-914a-e456eab18c41",
   "metadata": {},
   "source": [
    "Keep only samples that have downloadable data for all three AmpSeq panels and WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "351ba9b5-7145-43c2-9549-8b67d9191342",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_conc_s_ena2_df = genre_pf8_conc_s_ena_df[\n",
    "    genre_pf8_conc_s_ena_df[['GenRe_GRC1_ENA_FASTQ_FTP_1',\n",
    "       'GenRe_GRC2_ENA_FASTQ_FTP_1', 'GenRe_SPEC_ENA_FASTQ_FTP_1',\n",
    "       'Pf8_ENA_FASTQ_FTP_1', 'GenRe_GRC1_ENA_FASTQ_FTP_2',\n",
    "       'GenRe_GRC2_ENA_FASTQ_FTP_2', 'GenRe_SPEC_ENA_FASTQ_FTP_2',\n",
    "       'Pf8_ENA_FASTQ_FTP_2']].notnull().all(1)\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11f8f332-65a3-4531-99d0-30fdaa2d640c",
   "metadata": {},
   "source": [
    "number of samples with and without the filter for FTP links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0c9cdabf-293b-4d24-a3bd-dd1f71588d5c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n samples before FTP link filter: 968\n",
      "n samples with all FASTQ files available via FTP: 907\n"
     ]
    }
   ],
   "source": [
    "print(f'n samples before FTP link filter: {len(genre_pf8_conc_s_ena_df)}')\n",
    "print(f'n samples with all FASTQ files available via FTP: {len(genre_pf8_conc_s_ena2_df)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "258ef86e-45a3-4ba8-b46a-3487883d7610",
   "metadata": {},
   "source": [
    "### Save results: concordant phenotype data\n",
    "Save to csv file\n",
    "```Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv```  \n",
    "\n",
    "In summary: this table of 907 rows represents samples that have been sequenced by amplicon sequencing for GenRe Mekong and WGS for Pf8 and where the  sample is marked as high-quality in Pf8 and all phenotye calls (inferred from genotypes) are fully conconrdant between Pf8 and GenRe Mekong.  \n",
    "The samples have also been filtered to include only those that have FTP download links for all GenRe AmpSeq panels and the Pf8 WGS data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e4fd520c-6075-443c-bddf-b6830283427d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_conc_s_ena2_df.to_csv('Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d367d3b-6b2e-4bbe-9517-0e1c4bb42d02",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffc5fe1c-db61-42eb-a62d-11e99a53c153",
   "metadata": {},
   "source": [
    "## Add genotypes\n",
    "In the above steps, concordance was defined on high-level drug resistance phenotype calls.  \n",
    "Adding the underlying genotype calls will provide a lower-level truth dataset that can be used to validate genotype calls of a novel pipeline.  \n",
    "___Note___ that \"genotype\" calls are provided at amino acid level - not nucelotide level - in both studies. \n",
    "\n",
    "### Genes and haplotypes\n",
    "For the purpose of this gold-standard dataset, the intention is to identify samples that are fully concordant on their haplotype calls for important genes between Pf8 and GenRe Mekong, i.e. two different sequencing techinques and two pipelines have resulted in the same haplotype call.  \n",
    "\n",
    "This dataset will focus on 8 key drug resistance genes where drug resistance is determined by non-synonymous point mutations:\n",
    "- dhfr\n",
    "- crt\n",
    "- dhps\n",
    "- kelch13\n",
    "- mdr1\n",
    "- mdr2\n",
    "- fd\n",
    "- arps10\n",
    "\n",
    "This list of genes and \"core\" amino acid positions was compiled according to Table SM1 (Markers associated with drug resistance for P. falciparum) in the [SpotMalaria Platform Technical Notes and Methods](https://ngs.sanger.ac.uk/production/malaria/Resource/29/20200705-GenRe-04a-SpotMalaria-0.39.pdf). For Pf8, details about drug resistance classification can be found in in [this document](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf).  \n",
    "\n",
    "This results in the following list of drug-resistance haplotypes to be considered for the gold- standard dataset:\n",
    "\n",
    "| Gene | Amino Acid position |\n",
    "|---|---|\n",
    "| dhfr | 51, 59, 108, 164 |\n",
    "| crt | 72, ~~73~~\\*, 74, 75, 76, 326, 356 |\n",
    "| dhps | 436, 437, 540, 581, 613 |\n",
    "| kelch13 | 349-726 |\n",
    "| mdr1 | 86, 184, 1246 |\n",
    "| mdr2 | 484 |\n",
    "| fd | 193 |\n",
    "| arps10 | 127, 128 |\n",
    "\\* Position crt:73 is listed as a core position in the reference table but it is not provided in GenRe Mekong data and it is always WT in Pf8 data, hence this position will be ignored in the following code.  \n",
    "\n",
    "For more details on gene haplotypes in P. falciparum see the [Pf-HaploAtlas](https://www.malariagen.net/article/introducing-pf-haploatlas-a-new-app-to-track-malaria-parasite-mutations/). \n",
    "       "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f97089cf-3736-4dfb-9687-2a316d4eb3b0",
   "metadata": {},
   "source": [
    "### Load Pf8 genotype calls from data release\n",
    "The GenRe data loaded earlier does contain genotype data already."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "44396456-e86e-49ad-8ffa-3000ffea4dd3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_gt_df = pd.read_csv(\"https://pf8-release.cog.sanger.ac.uk/Pf8_drug_resistance_marker_genotypes.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01e86e1c-3b44-443c-963d-9ed75428895f",
   "metadata": {},
   "source": [
    "### Merge Pf8 genotypes \n",
    "Merge into into existing dataframe  \n",
    "```genre_pf8_merged_seq_df```  \n",
    "created above, which is the merged data for Pf8 and GenRe Mekong, with array data removed. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b15e53af-a105-4c49-acc4-86cd5ffbd7b4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pf8_gt_df.rename(columns={'Sample': 'sample'}, inplace=True)\n",
    "\n",
    "genre_pf8_merged_seq_gt_df = pd.merge(\n",
    "    genre_pf8_merged_seq_df, \n",
    "    pf8_gt_df, \n",
    "    on=[\"sample\"], \n",
    "    suffixes=('_GRMK','_pf8'), \n",
    "    how=\"inner\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6fee5a6-340e-4d1d-82a0-47187820625f",
   "metadata": {},
   "source": [
    "The dataframe now contains Pf8 and GenRe Mekong genotypes for all samples that have been sequenced (excluding Array data) in both projects and can be linked via sample ID."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba3e1534-bf4d-42d8-b5dd-51202e785cfd",
   "metadata": {},
   "source": [
    "### Create new dataframe with amino acid genotype calls\n",
    "Create a list of gene/aa-positions from the above table and use it to comile a new data frame with amino acid calls from Pf8 and GenRe Mekong for these core positions. The kelch13 list needs to be built separately because the data is provided in separate column in GenRe Mekong and as a sigle column of comma-separated lists in Pf8.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2664f9c-256b-4f14-985a-423a07ab9df3",
   "metadata": {},
   "source": [
    "The new columns will have a unified naming scheme:  \n",
    "\"core_mutation_{GENE}_{AAPos}_\\[pf8/GMRK\\]\"\n",
    "\n",
    "The majority of genotypes can be extracted from the Pf8 and GenRe data by a simple pattern.  \n",
    "The respective pattern of genotype calls by gene and AA position is as follows:  \n",
    "__Pf8__: {GENE}\\_{AAPos} (example \"crt_75\\[N\\]\")  \n",
    "__GenRe Mekong__: Pf{GENE}:{AAPos} (example: \"PfCRT:75\")  \n",
    "\n",
    "Create a set of new column names and respective Pf8 and GenRe column names accordingly.  \n",
    "There are ___two excpetions___: \n",
    "1. kelch13: in both datasets, kelch13 data is provided as a column of lists of AA positions that need to be converted separately\n",
    "2. arps10: provided as AA pos 127-128 in Pf8 and just pos 127 in GenRe. Needs to be converted separately\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b2a0b271-8de9-444e-9bbb-4693bd29f66d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genes_pos = {\n",
    "    'dhfr': [51, 59, 108, 164],\n",
    "    'crt': [72, 74, 75, 76, 326, 356], # omitting pos 73, see above for explanation\n",
    "    'dhps': [436, 437, 540, 581, 613],\n",
    "    'mdr1': [86, 184, 1246],\n",
    "    'mdr2': [484],\n",
    "    'fd': [193]\n",
    "}\n",
    "\n",
    "gt_cols = {}\n",
    "for gene, pos_list in genes_pos.items():\n",
    "    gt_cols[ gene ] = {}\n",
    "    for pos in pos_list:\n",
    "        gt_cols[ gene ][ pos ] = {\n",
    "            'pf8_pattern': gene + '_' + str(pos) + '\\[[A-Z]\\]',\n",
    "            'GRMK_name': 'Pf' + gene.upper() + ':' + str(pos),\n",
    "            'new_col_basename': 'core_mutation_' + gene + '_' + str(pos)\n",
    "        }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aded4aa5-d782-488b-9734-c32cbf10ac32",
   "metadata": {},
   "source": [
    "### Add new genotype columns \n",
    "Add new genotype columns for the core mutation positions to dataframe   \n",
    "```genre_pf8_merged_seq_gt_df```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "95c121c3-12e2-4945-b4b3-fe5c68f01a92",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in gt_cols:\n",
    "    for pos in gt_cols[ gene ]:\n",
    "        \n",
    "        # genotype from Pf8\n",
    "        pf8_pattern = gt_cols[gene][pos]['pf8_pattern']\n",
    "        new_col_pf8 = gt_cols[gene][pos]['new_col_basename'] + '_pf8'\n",
    "        pf8_col = [ c for c in genre_pf8_merged_seq_gt_df if re.match(pf8_pattern, c) ]\n",
    "        if not pf8_col:\n",
    "            raise ValueError(f'could not find Pf8 column starting with \"{pf8_pattern}\"')\n",
    "        elif len(pf8_col) > 1:\n",
    "            raise ValueError(f'found more than one Pf8 column starting with \"{pf8_pattern}\":{\", \".join(pf8_col)}')\n",
    "        else:\n",
    "            genre_pf8_merged_seq_gt_df[new_col_pf8] = genre_pf8_merged_seq_gt_df[pf8_col[0]].copy()\n",
    "            \n",
    "        # genotype from GenRe\n",
    "        genre_pattern = gt_cols[gene][pos]['GRMK_name']\n",
    "        new_col_genre = gt_cols[gene][pos]['new_col_basename'] + '_GRMK'\n",
    "        if genre_pattern in genre_pf8_merged_seq_gt_df:\n",
    "            genre_pf8_merged_seq_gt_df[new_col_genre] = genre_pf8_merged_seq_gt_df[genre_pattern].copy()\n",
    "        else:\n",
    "            raise ValueError(f'could not find GenRe column \"{genre_pattern}\"')\n",
    "                                                                         "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c812509a-8c40-4953-acda-99731ea726cf",
   "metadata": {
    "tags": []
   },
   "source": [
    "Add columns for arps10: use only position arsp10:127, which is provided as a column inf GenRe and needs to be extracted from column 'arps10_127-128\\[VD\\]' in Pf8."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4031cbdb-274d-40a7-baf1-73ff4e7ddd3d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "arps10_basename = 'core_mutation_arps10_127'\n",
    "genre_pf8_merged_seq_gt_df[ arps10_basename + '_GRMK'] = genre_pf8_merged_seq_gt_df['PfARPS10:127'].copy()\n",
    "genre_pf8_merged_seq_gt_df[ arps10_basename + '_pf8'] = genre_pf8_merged_seq_gt_df['arps10_127-128[VD]'].astype(str).str[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "184395e7-5bfb-48aa-b948-c612874257e7",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Add columns for kelch13\n",
    "Add columns for all kelch13 AA positions for which data exists in at least one sample in both projects.  \n",
    "__Pf8__:  \n",
    "    _column_: 'kelch13_349-726_ns_changes'  \n",
    "    _content_: comma separated list of haplotypes at Kelch13 positions 349-726. Each haplotype contains one or more non-synonymous variations, separateed by \"/\". Homozygous mutations are shown in upper case and heterozygous in lower case.   Details are [here (heading \"Pf8_drug_resistance_marker_genotypes.tsv\"](https://pf8-release.cog.sanger.ac.uk/Pf8_README.txt)  \n",
    "    \n",
    "__GenRe__:  \n",
    "    _column_: 'Pfkelch13'  \n",
    "    _content_: Comma-separated list of amino acid mutations. ___NOTE___ this differs from Pf8. In GenRe, the comma delimits mutations in the same haplotype, while in Pf8 (see above) a comma delimts haplotypes. There is no disctiontion between homo- and heterozygous mutations.  For details see the [GenRe Mekong data dictionary](https://github.com/GenRe-Mekong/Documents/blob/e331501b8e93752221ebd7ae72e5eddd3e6e924d/Pf-GenRe-GRCv1.4-DataDictionary.xlsx)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15d08078-852a-4f15-89b7-5cb66871c089",
   "metadata": {},
   "source": [
    "Show the unique contents of the respective kelch13 columns in Pf8 and GenRe Mekong: "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f3acdf7-cf41-48ed-aa18-a12cc32e7b0d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create individual kelch13 mutation columns\n",
    "For each individual mutation found in any sample in any project, create a column with a name following this schema:  \n",
    "```core_mutation_kelch13:{MUTATION}_{pf8/GRMK}```  \n",
    "MUTATION is of the format ```{WT allele}{Pos}{mutant allele}```   \n",
    "The column contains boolean values where True indicates that this mutation is present in a given sample. Due to the differences in data provided between the two projects, no resolution at haplotype level is provided and no information about zygosity is recorded. Consequently, the \"WT\" allele recorded in some samples in GenRe Mekong is ignored. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "36079b97-30b2-4743-8ddf-3558675d629f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "kelch13_mutations = set()\n",
    "\n",
    "# Pf8\n",
    "for mut_list in genre_pf8_merged_seq_gt_df['kelch13_349-726_ns_changes'].unique():\n",
    "    mut_list = str(mut_list)\n",
    "    if mut_list and re.search(r'[1-9]', mut_list):\n",
    "        mut_list = mut_list.replace('/',',') # treat all mutations individually, regardless of haplotype grouping\n",
    "        mut_list = mut_list.replace('*','')\n",
    "        for mut in mut_list.split(','):\n",
    "            kelch13_mutations.add(mut.upper())\n",
    "            \n",
    "# GenRe Mekong\n",
    "for mut_list in genre_pf8_merged_seq_gt_df['Pfkelch13'].unique():\n",
    "    mut_list = str(mut_list)\n",
    "    if mut_list and re.search(r'[1-9]', mut_list):\n",
    "        for mut in mut_list.split(','):\n",
    "            if mut != 'WT':\n",
    "                kelch13_mutations.add(mut.upper())\n",
    "\n",
    "# add the new columns\n",
    "genre_pf8_merged_seq_gt_df = genre_pf8_merged_seq_gt_df.copy()\n",
    "for mut in kelch13_mutations:\n",
    "    base_name = 'core_mutation_kelch13:' + mut\n",
    "    genre_pf8_merged_seq_gt_df[base_name + '_pf8'] = genre_pf8_merged_seq_gt_df['kelch13_349-726_ns_changes'].str.contains(mut, case=False)\n",
    "    genre_pf8_merged_seq_gt_df[base_name + '_GRMK'] = genre_pf8_merged_seq_gt_df['Pfkelch13'].str.contains(mut, case=False)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e70bfe4-9425-4b16-a32c-721f272288a7",
   "metadata": {},
   "source": [
    "### Identify concordant samples\n",
    "Filter for samples that are fully concordant on all core mutation genotype calls, i.e. all columns added with the \"core_mutation_\" prefix above.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6339e484-22cc-46f4-b1a0-34c37efe3d13",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def is_concordant(row, *core_mut_col_basenames):\n",
    "    \"\"\"\n",
    "    Return True if all 'core_mutation' columns match between \n",
    "    Pf8 and GenRe Mekong\n",
    "    \"\"\"\n",
    "    for col in core_mut_col_basenames:\n",
    "        if row[col + '_pf8'] != row[col + '_GRMK']:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "regex = re.compile(r'^(core_mutation_.+)_(pf8|GRMK)$')\n",
    "core_mut_col_basenames = set(\n",
    "    [ regex.sub(r'\\1', x) for x in genre_pf8_merged_seq_gt_df.columns.values if regex.match(x) ]\n",
    ")\n",
    "genre_pf8_merged_seq_gt_df = genre_pf8_merged_seq_gt_df.copy()\n",
    "genre_pf8_merged_seq_gt_df['fully_concordant_gt'] = genre_pf8_merged_seq_gt_df.apply(is_concordant, axis=1, args=core_mut_col_basenames)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "927fa9b9-5931-45b0-8cc5-ec0b6fee3309",
   "metadata": {},
   "source": [
    "Count fully concordant (on all genotypes reported) vs. not fully concordant samples across the two projects:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9dd4475b-991f-4e50-a64d-aee4b65434bf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fully_concordant_gt</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>1572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>1394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   fully_concordant_gt  count\n",
       "0                False   1572\n",
       "1                 True   1394"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_gt_df.groupby( ['fully_concordant_gt']).size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44ca7bb6-23ad-4764-abaf-807d295cdff6",
   "metadata": {},
   "source": [
    "47% of samples are fully concordant on all drug-reistance \"core mutation\" genotypes between Pf8 (WGS) and GenRe Mekong (AmpSeq)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f680f155-accf-44ea-8516-cbc787cdb588",
   "metadata": {},
   "source": [
    "### Create a filtered dataset containing only fully concordant samples\n",
    "In addition, apply filters on Pf8 quality, as done above for phenotypes (%callable >=85 and QC pass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "63a5407c-9d79-4376-8362-51753a6e39f7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_df = genre_pf8_merged_seq_gt_df[\n",
    "    (genre_pf8_merged_seq_gt_df['fully_concordant_gt'] == True ) &\n",
    "    (genre_pf8_merged_seq_gt_df['% callable'] >= 85 ) &\n",
    "    (genre_pf8_merged_seq_gt_df['QC pass'] == True )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a41d5774-56ab-4ef6-aff9-cb1acab44759",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "976"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genre_pf8_merged_seq_gt_concord_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab193a8e-b2d7-4e68-b323-547400a2072b",
   "metadata": {},
   "source": [
    "There are 976 high-quality samples where all reported drug-resitance genotype positions are fully concordant between Pf8 and GenRe Mekong."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a42608a5-d677-475f-80b3-f0a02a6a454d",
   "metadata": {},
   "source": [
    "Keep only relevant columns. As all samples are fully concordant, keep only one column per core mutation and drop the pf8/GRMK suffix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "79fb2e2d-98a4-41c2-8e30-91acb68e1016",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "keep_cols = ['sample', 'Source', 'Process', \n",
    "    'TimePoint', 'Country_GRMK', 'Species',\n",
    "    'fully_concordant_gt']\n",
    "\n",
    "core_mut_cols = [x for x in genre_pf8_merged_seq_gt_concord_df.columns.values if x.startswith('core_mutation')]\n",
    "keep_cols.extend( core_mut_cols )\n",
    "genre_pf8_merged_seq_gt_concord_small_df = genre_pf8_merged_seq_gt_concord_df[keep_cols]\n",
    "genre_pf8_merged_seq_gt_concord_small_df.rename({'Country_GRMK':'Country'})\n",
    "\n",
    "regex = re.compile(r'^(core_mutation_.+)_(pf8|GRMK)$')\n",
    "for col in core_mut_cols:\n",
    "    col_basename = regex.sub(r'\\1', col)\n",
    "    genre_pf8_merged_seq_gt_concord_small_df[col_basename] = genre_pf8_merged_seq_gt_concord_small_df[col].copy()\n",
    "    genre_pf8_merged_seq_gt_concord_small_df.drop(col, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2531203d-adc1-4563-8d33-e17af3ed1f74",
   "metadata": {},
   "source": [
    "## Add ENA accessions and download links\n",
    "Using functions from the custom ENA data helper tool (file ENA_data_helper.py in the working directory), perform a search on ENA by sample ID and retrieve data for run accessions (unique accession for 1 sample on 1 run) and FTP download URLs for FASTQ files (read1/read2).  \n",
    "This step performs a live search on ENA and may not work if your internet connection is very slow or unstable."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4e5e86c-6246-46d5-9d98-66882e4a69a3",
   "metadata": {},
   "source": [
    "The first function performs a search for ENA data by sample IDs in sample data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "02dbe2f7-4180-4f5b-b12b-d84240fb4ac1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_small_ena_result = create_ena_data_frame(genre_pf8_merged_seq_gt_concord_small_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fbdf5ec-c2d5-4ffc-85ab-bd21db270074",
   "metadata": {},
   "source": [
    "Second function merges the ENA search results (which contain more than one result per sample) back into the original sample data frame where sample ID is unique. Columns are added for the run accessions for all threee panels in GenRe as well as the WGS data for Pf8."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "909046db-6189-4fdd-bbd0-2b9db6d24764",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_small_ena_df = merge_ena_results_into_sample_data_genre_pf8(sample_data=genre_pf8_merged_seq_gt_concord_small_df, ena_result=genre_pf8_merged_seq_gt_concord_small_ena_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7865960b-0e0e-4a47-952e-891d357a693a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>Process</th>\n",
       "      <th>TimePoint</th>\n",
       "      <th>Country_GRMK</th>\n",
       "      <th>Species</th>\n",
       "      <th>fully_concordant_gt</th>\n",
       "      <th>core_mutation_dhfr_51</th>\n",
       "      <th>core_mutation_dhfr_59</th>\n",
       "      <th>core_mutation_dhfr_108</th>\n",
       "      <th>core_mutation_dhfr_164</th>\n",
       "      <th>...</th>\n",
       "      <th>ENA_acc_GenRe_GRC2</th>\n",
       "      <th>ENA_acc_GenRe_SPEC</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_1</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_2</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RCN01775</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>-</td>\n",
       "      <td>VN</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/078/...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/078/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN01776</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>-</td>\n",
       "      <td>VN</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/079/...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/079/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN01777</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>-</td>\n",
       "      <td>VN</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/080/...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/080/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN01778</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>-</td>\n",
       "      <td>VN</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/081/...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/081/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN01779</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>-</td>\n",
       "      <td>VN</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/082/...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/082/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26767</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>D00H00</td>\n",
       "      <td>LA</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397547</td>\n",
       "      <td>ERR14397548</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26782</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>D00H00</td>\n",
       "      <td>LA</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397589</td>\n",
       "      <td>ERR14397590</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/088/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/089/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/057/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/088/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/089/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/057/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26820</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>D00H00</td>\n",
       "      <td>LA</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397664</td>\n",
       "      <td>ERR14397665</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26852</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>D00H00</td>\n",
       "      <td>LA</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397748</td>\n",
       "      <td>ERR14397749</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/049/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/074/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/049/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/074/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN26938</th>\n",
       "      <td>GenRe-Mekong</td>\n",
       "      <td>AmpSeqV2</td>\n",
       "      <td>D00H00</td>\n",
       "      <td>LA</td>\n",
       "      <td>Pf</td>\n",
       "      <td>True</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397991</td>\n",
       "      <td>ERR14397992</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>976 rows × 134 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Source   Process TimePoint Country_GRMK Species  \\\n",
       "sample                                                            \n",
       "RCN01775  GenRe-Mekong  AmpSeqV2         -           VN      Pf   \n",
       "RCN01776  GenRe-Mekong  AmpSeqV2         -           VN      Pf   \n",
       "RCN01777  GenRe-Mekong  AmpSeqV2         -           VN      Pf   \n",
       "RCN01778  GenRe-Mekong  AmpSeqV2         -           VN      Pf   \n",
       "RCN01779  GenRe-Mekong  AmpSeqV2         -           VN      Pf   \n",
       "...                ...       ...       ...          ...     ...   \n",
       "RCN26767  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   \n",
       "RCN26782  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   \n",
       "RCN26820  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   \n",
       "RCN26852  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   \n",
       "RCN26938  GenRe-Mekong  AmpSeqV2    D00H00           LA      Pf   \n",
       "\n",
       "          fully_concordant_gt core_mutation_dhfr_51 core_mutation_dhfr_59  \\\n",
       "sample                                                                      \n",
       "RCN01775                 True                     I                     R   \n",
       "RCN01776                 True                     I                     R   \n",
       "RCN01777                 True                     I                     R   \n",
       "RCN01778                 True                     I                     R   \n",
       "RCN01779                 True                     I                     R   \n",
       "...                       ...                   ...                   ...   \n",
       "RCN26767                 True                     I                     R   \n",
       "RCN26782                 True                     I                     R   \n",
       "RCN26820                 True                     I                     R   \n",
       "RCN26852                 True                     I                     R   \n",
       "RCN26938                 True                     I                     R   \n",
       "\n",
       "         core_mutation_dhfr_108 core_mutation_dhfr_164  ...  \\\n",
       "sample                                                  ...   \n",
       "RCN01775                      N                      L  ...   \n",
       "RCN01776                      N                      L  ...   \n",
       "RCN01777                      N                      L  ...   \n",
       "RCN01778                      N                      I  ...   \n",
       "RCN01779                      N                      L  ...   \n",
       "...                         ...                    ...  ...   \n",
       "RCN26767                      N                      I  ...   \n",
       "RCN26782                      N                      I  ...   \n",
       "RCN26820                      N                      I  ...   \n",
       "RCN26852                      N                      I  ...   \n",
       "RCN26938                      N                      I  ...   \n",
       "\n",
       "         ENA_acc_GenRe_GRC2 ENA_acc_GenRe_SPEC  \\\n",
       "sample                                           \n",
       "RCN01775               None               None   \n",
       "RCN01776               None               None   \n",
       "RCN01777               None               None   \n",
       "RCN01778               None               None   \n",
       "RCN01779               None               None   \n",
       "...                     ...                ...   \n",
       "RCN26767        ERR14397547        ERR14397548   \n",
       "RCN26782        ERR14397589        ERR14397590   \n",
       "RCN26820        ERR14397664        ERR14397665   \n",
       "RCN26852        ERR14397748        ERR14397749   \n",
       "RCN26938        ERR14397991        ERR14397992   \n",
       "\n",
       "                                 GenRe_GRC1_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN01775                                               None   \n",
       "RCN01776                                               None   \n",
       "RCN01777                                               None   \n",
       "RCN01778                                               None   \n",
       "RCN01779                                               None   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/088/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "RCN26852  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "\n",
       "                                 GenRe_GRC2_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN01775                                               None   \n",
       "RCN01776                                               None   \n",
       "RCN01777                                               None   \n",
       "RCN01778                                               None   \n",
       "RCN01779                                               None   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/089/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "RCN26852  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...   \n",
       "\n",
       "                                 GenRe_SPEC_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN01775                                               None   \n",
       "RCN01776                                               None   \n",
       "RCN01777                                               None   \n",
       "RCN01778                                               None   \n",
       "RCN01779                                               None   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "RCN26852  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/049/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...   \n",
       "\n",
       "                                        Pf8_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN01775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/078/...   \n",
       "RCN01776  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/079/...   \n",
       "RCN01777  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/080/...   \n",
       "RCN01778  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/081/...   \n",
       "RCN01779  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/082/...   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...   \n",
       "RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/057/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...   \n",
       "RCN26852  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/074/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...   \n",
       "\n",
       "                                 GenRe_GRC1_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN01775                                               None   \n",
       "RCN01776                                               None   \n",
       "RCN01777                                               None   \n",
       "RCN01778                                               None   \n",
       "RCN01779                                               None   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/088/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "RCN26852  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "\n",
       "                                 GenRe_GRC2_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN01775                                               None   \n",
       "RCN01776                                               None   \n",
       "RCN01777                                               None   \n",
       "RCN01778                                               None   \n",
       "RCN01779                                               None   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/089/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "RCN26852  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...   \n",
       "\n",
       "                                 GenRe_SPEC_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN01775                                               None   \n",
       "RCN01776                                               None   \n",
       "RCN01777                                               None   \n",
       "RCN01778                                               None   \n",
       "RCN01779                                               None   \n",
       "...                                                     ...   \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "RCN26852  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/049/...   \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...   \n",
       "\n",
       "                                        Pf8_ENA_FASTQ_FTP_2  \n",
       "sample                                                       \n",
       "RCN01775  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/078/...  \n",
       "RCN01776  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/079/...  \n",
       "RCN01777  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/080/...  \n",
       "RCN01778  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/081/...  \n",
       "RCN01779  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/082/...  \n",
       "...                                                     ...  \n",
       "RCN26767  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...  \n",
       "RCN26782  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/057/...  \n",
       "RCN26820  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...  \n",
       "RCN26852  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/074/...  \n",
       "RCN26938  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...  \n",
       "\n",
       "[976 rows x 134 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_gt_concord_small_ena_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5eb7e83e-c6c7-45e4-988c-25a0ede0edd5",
   "metadata": {},
   "source": [
    "Keep only samples that have downloadable data for all three AmpSeq panels and WGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fd3b36c6-8355-4c5a-bd3d-2cf44c84ae24",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_small_ena2_df = genre_pf8_merged_seq_gt_concord_small_ena_df[\n",
    "    genre_pf8_merged_seq_gt_concord_small_ena_df[['GenRe_GRC1_ENA_FASTQ_FTP_1',\n",
    "       'GenRe_GRC2_ENA_FASTQ_FTP_1', 'GenRe_SPEC_ENA_FASTQ_FTP_1',\n",
    "       'Pf8_ENA_FASTQ_FTP_1', 'GenRe_GRC1_ENA_FASTQ_FTP_2',\n",
    "       'GenRe_GRC2_ENA_FASTQ_FTP_2', 'GenRe_SPEC_ENA_FASTQ_FTP_2',\n",
    "       'Pf8_ENA_FASTQ_FTP_2']].notnull().all(1)\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bb03c37-2809-464d-be5a-1a96b0405807",
   "metadata": {},
   "source": [
    "number of samples with and without the filter for FTP links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a2790902-9885-4bae-b595-3f5081822845",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n samples before FTP link filter: 976\n",
      "n samples with all FASTQ files available via FTP: 873\n"
     ]
    }
   ],
   "source": [
    "print(f'n samples before FTP link filter: {len(genre_pf8_merged_seq_gt_concord_small_ena_df)}')\n",
    "print(f'n samples with all FASTQ files available via FTP: {len(genre_pf8_merged_seq_gt_concord_small_ena2_df)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2782f187-6467-475e-9583-48a9d32f9a1c",
   "metadata": {},
   "source": [
    "### Save file of fully concordant samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "0de529f7-1eb7-418c-b30a-078d8b541b0c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_small_ena2_df.to_csv('Pf8-GenReMekong_fully_concordant_genotype_high_quality_samples.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2da663ef-5673-4bd9-8649-62a3482b46e7",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dce7e08-fc81-46a0-bd4a-dc112d7543b2",
   "metadata": {},
   "source": [
    "## Distinct genotype patterns and representative samples\n",
    "Final dataset is based on the fully concordant genotypes dataset, but will only contain one representative sample for every distinct pattern of haplotypes.   \n",
    "This dataset is based on the previous dataset, which means only samples with downloadable FASTQ files are taken into consideration here."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f409d61d-ddcd-47d3-9366-7bc5290ac0d8",
   "metadata": {},
   "source": [
    "First, get a count of different patterns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a03c1d4b-bd88-4075-9047-2e3fff1785e9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>core_mutation_dhfr_51</th>\n",
       "      <th>core_mutation_dhfr_59</th>\n",
       "      <th>core_mutation_dhfr_108</th>\n",
       "      <th>core_mutation_dhfr_164</th>\n",
       "      <th>core_mutation_crt_72</th>\n",
       "      <th>core_mutation_crt_74</th>\n",
       "      <th>core_mutation_crt_75</th>\n",
       "      <th>core_mutation_crt_76</th>\n",
       "      <th>core_mutation_crt_326</th>\n",
       "      <th>core_mutation_crt_356</th>\n",
       "      <th>...</th>\n",
       "      <th>core_mutation_kelch13:T474I</th>\n",
       "      <th>core_mutation_kelch13:G358V</th>\n",
       "      <th>core_mutation_kelch13:G639D</th>\n",
       "      <th>core_mutation_kelch13:I376M</th>\n",
       "      <th>core_mutation_kelch13:W470R</th>\n",
       "      <th>core_mutation_kelch13:I723V</th>\n",
       "      <th>core_mutation_kelch13:I540V</th>\n",
       "      <th>core_mutation_kelch13:P553L</th>\n",
       "      <th>core_mutation_kelch13:E705K</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I,L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I,L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34 rows × 117 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   core_mutation_dhfr_51 core_mutation_dhfr_59 core_mutation_dhfr_108  \\\n",
       "0                      I                     R                      N   \n",
       "1                      I                     R                      N   \n",
       "2                      I                     R                      N   \n",
       "3                      I                     R                      N   \n",
       "4                      I                     R                      N   \n",
       "5                      I                     R                      N   \n",
       "6                      I                     R                      N   \n",
       "7                      I                     R                      N   \n",
       "8                      I                     R                      N   \n",
       "9                      I                     R                      N   \n",
       "10                     I                     R                      N   \n",
       "11                     I                     R                      N   \n",
       "12                     I                     R                      N   \n",
       "13                     I                     R                      N   \n",
       "14                     I                     R                      N   \n",
       "15                     I                     R                      N   \n",
       "16                     I                     R                      N   \n",
       "17                     I                     R                      N   \n",
       "18                     I                     R                      N   \n",
       "19                     I                     R                      N   \n",
       "20                     I                     R                      N   \n",
       "21                     I                     R                      N   \n",
       "22                     I                     R                      N   \n",
       "23                     I                     R                      N   \n",
       "24                     I                     R                      N   \n",
       "25                     I                     R                      N   \n",
       "26                     I                     R                      N   \n",
       "27                     I                     R                      N   \n",
       "28                     I                     R                      N   \n",
       "29                     I                     R                      N   \n",
       "30                     I                     R                      N   \n",
       "31                     I                     R                      N   \n",
       "32                     I                     R                      N   \n",
       "33                     I                     R                      N   \n",
       "\n",
       "   core_mutation_dhfr_164 core_mutation_crt_72 core_mutation_crt_74  \\\n",
       "0                       I                    C                    I   \n",
       "1                       I                    C                    I   \n",
       "2                       I                    C                    I   \n",
       "3                       I                    C                    I   \n",
       "4                       I                    C                    I   \n",
       "5                       I                    C                    I   \n",
       "6                       I                    C                    I   \n",
       "7                       I                    C                    I   \n",
       "8                       I                    C                    I   \n",
       "9                       I                    C                    I   \n",
       "10                      I                    C                    I   \n",
       "11                      I                    C                    I   \n",
       "12                      I                    C                    I   \n",
       "13                      I                    C                    I   \n",
       "14                      I                    C                    I   \n",
       "15                      I                    C                    I   \n",
       "16                      I                    C                    I   \n",
       "17                    I,L                    C                    I   \n",
       "18                    I,L                    C                    I   \n",
       "19                      L                    C                    I   \n",
       "20                      L                    C                    I   \n",
       "21                      L                    C                    I   \n",
       "22                      L                    C                    I   \n",
       "23                      L                    C                    I   \n",
       "24                      L                    C                    I   \n",
       "25                      L                    C                    I   \n",
       "26                      L                    C                    I   \n",
       "27                      L                    C                    I   \n",
       "28                      L                    C                    I   \n",
       "29                      L                    C                    I   \n",
       "30                      L                    C                    I   \n",
       "31                      L                    C                    I   \n",
       "32                      L                    C                    I   \n",
       "33                      L                    C                    I   \n",
       "\n",
       "   core_mutation_crt_75 core_mutation_crt_76 core_mutation_crt_326  \\\n",
       "0                     D                    T                     N   \n",
       "1                     D                    T                     N   \n",
       "2                     E                    T                     N   \n",
       "3                     E                    T                     N   \n",
       "4                     E                    T                     S   \n",
       "5                     E                    T                     S   \n",
       "6                     E                    T                     S   \n",
       "7                     E                    T                     S   \n",
       "8                     E                    T                     S   \n",
       "9                     E                    T                     S   \n",
       "10                    E                    T                     S   \n",
       "11                    E                    T                     S   \n",
       "12                    E                    T                     S   \n",
       "13                    E                    T                     S   \n",
       "14                    E                    T                     S   \n",
       "15                    E                    T                     S   \n",
       "16                    E                    T                     S   \n",
       "17                    E                    T                     S   \n",
       "18                    E                    T                     S   \n",
       "19                    D                    T                     N   \n",
       "20                    E                    T                     S   \n",
       "21                    E                    T                     S   \n",
       "22                    E                    T                     S   \n",
       "23                    E                    T                     S   \n",
       "24                    E                    T                     S   \n",
       "25                    E                    T                     S   \n",
       "26                    E                    T                     S   \n",
       "27                    E                    T                     S   \n",
       "28                    E                    T                     S   \n",
       "29                    E                    T                     S   \n",
       "30                    E                    T                     S   \n",
       "31                    E                    T                     S   \n",
       "32                    E                    T                     S   \n",
       "33                    E                    T                     S   \n",
       "\n",
       "   core_mutation_crt_356  ... core_mutation_kelch13:T474I  \\\n",
       "0                      I  ...                       False   \n",
       "1                      I  ...                       False   \n",
       "2                      I  ...                       False   \n",
       "3                      I  ...                       False   \n",
       "4                      T  ...                       False   \n",
       "5                      T  ...                       False   \n",
       "6                      T  ...                       False   \n",
       "7                      T  ...                       False   \n",
       "8                      T  ...                       False   \n",
       "9                      T  ...                       False   \n",
       "10                     T  ...                       False   \n",
       "11                     T  ...                       False   \n",
       "12                     T  ...                       False   \n",
       "13                     T  ...                       False   \n",
       "14                     T  ...                       False   \n",
       "15                     T  ...                       False   \n",
       "16                     T  ...                       False   \n",
       "17                     T  ...                       False   \n",
       "18                     T  ...                       False   \n",
       "19                     I  ...                       False   \n",
       "20                     T  ...                       False   \n",
       "21                     T  ...                       False   \n",
       "22                     T  ...                       False   \n",
       "23                     T  ...                       False   \n",
       "24                     T  ...                       False   \n",
       "25                     T  ...                       False   \n",
       "26                     T  ...                       False   \n",
       "27                     T  ...                       False   \n",
       "28                     T  ...                       False   \n",
       "29                     T  ...                       False   \n",
       "30                     T  ...                       False   \n",
       "31                     T  ...                       False   \n",
       "32                     T  ...                       False   \n",
       "33                     T  ...                       False   \n",
       "\n",
       "   core_mutation_kelch13:G358V core_mutation_kelch13:G639D  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:I376M core_mutation_kelch13:W470R  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:I723V core_mutation_kelch13:I540V  \\\n",
       "0                        False                       False   \n",
       "1                        False                       False   \n",
       "2                        False                       False   \n",
       "3                        False                       False   \n",
       "4                        False                       False   \n",
       "5                        False                       False   \n",
       "6                        False                       False   \n",
       "7                        False                       False   \n",
       "8                        False                       False   \n",
       "9                        False                       False   \n",
       "10                       False                       False   \n",
       "11                       False                       False   \n",
       "12                       False                       False   \n",
       "13                       False                       False   \n",
       "14                       False                       False   \n",
       "15                       False                       False   \n",
       "16                       False                       False   \n",
       "17                       False                       False   \n",
       "18                       False                       False   \n",
       "19                       False                       False   \n",
       "20                       False                       False   \n",
       "21                       False                       False   \n",
       "22                       False                       False   \n",
       "23                       False                       False   \n",
       "24                       False                       False   \n",
       "25                       False                       False   \n",
       "26                       False                       False   \n",
       "27                       False                       False   \n",
       "28                       False                       False   \n",
       "29                       False                       False   \n",
       "30                       False                       False   \n",
       "31                       False                       False   \n",
       "32                       False                       False   \n",
       "33                       False                       False   \n",
       "\n",
       "   core_mutation_kelch13:P553L core_mutation_kelch13:E705K count  \n",
       "0                        False                       False     1  \n",
       "1                        False                       False    10  \n",
       "2                        False                       False     1  \n",
       "3                        False                       False     3  \n",
       "4                        False                       False    22  \n",
       "5                        False                       False     5  \n",
       "6                        False                       False    26  \n",
       "7                        False                       False     2  \n",
       "8                        False                       False     1  \n",
       "9                        False                       False     1  \n",
       "10                       False                       False     5  \n",
       "11                       False                       False     1  \n",
       "12                       False                       False     1  \n",
       "13                       False                       False    17  \n",
       "14                       False                       False    21  \n",
       "15                       False                       False     1  \n",
       "16                       False                       False     4  \n",
       "17                       False                       False     3  \n",
       "18                       False                       False     1  \n",
       "19                       False                       False     2  \n",
       "20                       False                       False     1  \n",
       "21                       False                       False     1  \n",
       "22                       False                       False     2  \n",
       "23                       False                       False     1  \n",
       "24                       False                       False     1  \n",
       "25                       False                       False     1  \n",
       "26                       False                       False     1  \n",
       "27                       False                       False     1  \n",
       "28                       False                       False   721  \n",
       "29                       False                       False     1  \n",
       "30                       False                       False     5  \n",
       "31                       False                       False     1  \n",
       "32                       False                       False     7  \n",
       "33                       False                       False     1  \n",
       "\n",
       "[34 rows x 117 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "core_mut_cols = [x for x in genre_pf8_merged_seq_gt_concord_small_ena2_df.columns.values if x.startswith('core_mutation')]\n",
    "genre_pf8_merged_seq_gt_concord_small_ena2_df.groupby( core_mut_cols).size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67b72d7a-e91e-409b-b078-23c9e9f6a368",
   "metadata": {},
   "source": [
    "There are 34 distinct patterns of genotypes.\n",
    "Extract a list of sample IDs of representative samples for each pattern."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e046eeea-3aeb-4a69-9a5b-2bd992f0c535",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>core_mutation_dhfr_51</th>\n",
       "      <th>core_mutation_dhfr_59</th>\n",
       "      <th>core_mutation_dhfr_108</th>\n",
       "      <th>core_mutation_dhfr_164</th>\n",
       "      <th>core_mutation_crt_72</th>\n",
       "      <th>core_mutation_crt_74</th>\n",
       "      <th>core_mutation_crt_75</th>\n",
       "      <th>core_mutation_crt_76</th>\n",
       "      <th>core_mutation_crt_326</th>\n",
       "      <th>core_mutation_crt_356</th>\n",
       "      <th>...</th>\n",
       "      <th>ENA_acc_GenRe_GRC2</th>\n",
       "      <th>ENA_acc_GenRe_SPEC</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_1</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_2</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RCN13568</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390746</td>\n",
       "      <td>ERR14390747</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/045/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/045/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/095/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13560</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390722</td>\n",
       "      <td>ERR14390723</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/087/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/087/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN15107</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14392567</td>\n",
       "      <td>ERR14392568</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/066/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/067/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/068/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/087/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/066/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/067/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/068/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/087/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN14580</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14391280</td>\n",
       "      <td>ERR14391281</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/080/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/029/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/080/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/029/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13532</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390644</td>\n",
       "      <td>ERR14390645</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/043/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/044/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/045/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/060/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/043/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/044/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/045/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/060/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13481</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390500</td>\n",
       "      <td>ERR14390501</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/099/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/001/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/013/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/099/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/001/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/013/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12032</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388625</td>\n",
       "      <td>ERR14388626</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13501</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390560</td>\n",
       "      <td>ERR14390561</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/059/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/060/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/061/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/033/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/059/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/060/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/061/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/033/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12873</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389801</td>\n",
       "      <td>ERR14389802</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/001/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/002/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/001/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/002/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/075/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12710</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389408</td>\n",
       "      <td>ERR14389409</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/009/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/050/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/009/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/050/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13368</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390182</td>\n",
       "      <td>ERR14390183</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/008/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12729</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389462</td>\n",
       "      <td>ERR14389463</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/061/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/062/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/060/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/061/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/062/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/060/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13528</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390632</td>\n",
       "      <td>ERR14390633</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/031/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/032/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/033/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/056/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/031/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/032/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/033/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/056/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12087</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388775</td>\n",
       "      <td>ERR14388776</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/074/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/074/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12033</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388628</td>\n",
       "      <td>ERR14388629</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/027/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/028/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/029/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/012/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/027/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/028/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/029/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/012/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13585</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390794</td>\n",
       "      <td>ERR14390795</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/093/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/093/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12865</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389780</td>\n",
       "      <td>ERR14389781</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/080/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/067/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/080/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/067/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN18593</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I,L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14394295</td>\n",
       "      <td>ERR14394296</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/096/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/034/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/096/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/034/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13407</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>I,L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390296</td>\n",
       "      <td>ERR14390297</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/096/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/097/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/045/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/096/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/097/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/045/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN15343</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>D</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14393233</td>\n",
       "      <td>ERR14393234</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/032/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/033/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/034/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/097/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/032/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/033/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/034/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/097/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12796</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389594</td>\n",
       "      <td>ERR14389595</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/093/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/001/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/093/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/001/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN25190</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14395864</td>\n",
       "      <td>ERR14395865</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN13333</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14390083</td>\n",
       "      <td>ERR14390084</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/084/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/073/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/084/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/073/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN14642</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14391457</td>\n",
       "      <td>ERR14391458</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/056/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/057/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/058/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/056/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/057/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/058/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/090/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN15274</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14393038</td>\n",
       "      <td>ERR14393039</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/037/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/038/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/039/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/031/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/037/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/038/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/039/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/031/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12862</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389777</td>\n",
       "      <td>ERR14389778</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/078/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/078/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12855</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389759</td>\n",
       "      <td>ERR14389760</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/058/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/059/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/060/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/057/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/058/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/059/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/060/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/057/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12480</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388970</td>\n",
       "      <td>ERR14388971</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/069/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/050/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/069/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/050/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12025</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388604</td>\n",
       "      <td>ERR14388605</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN25946</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14396599</td>\n",
       "      <td>ERR14396600</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/098/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/099/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/059/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/098/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/099/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/059/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12789</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389576</td>\n",
       "      <td>ERR14389577</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/094/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/094/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12854</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14389756</td>\n",
       "      <td>ERR14389757</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/055/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/056/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/057/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/056/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/055/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/056/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/057/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/056/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN12484</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388982</td>\n",
       "      <td>ERR14388983</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/054/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/054/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCN15359</th>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14393278</td>\n",
       "      <td>ERR14393279</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/078/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/012/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/078/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/012/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34 rows × 128 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         core_mutation_dhfr_51 core_mutation_dhfr_59 core_mutation_dhfr_108  \\\n",
       "sample                                                                        \n",
       "RCN13568                     I                     R                      N   \n",
       "RCN13560                     I                     R                      N   \n",
       "RCN15107                     I                     R                      N   \n",
       "RCN14580                     I                     R                      N   \n",
       "RCN13532                     I                     R                      N   \n",
       "RCN13481                     I                     R                      N   \n",
       "RCN12032                     I                     R                      N   \n",
       "RCN13501                     I                     R                      N   \n",
       "RCN12873                     I                     R                      N   \n",
       "RCN12710                     I                     R                      N   \n",
       "RCN13368                     I                     R                      N   \n",
       "RCN12729                     I                     R                      N   \n",
       "RCN13528                     I                     R                      N   \n",
       "RCN12087                     I                     R                      N   \n",
       "RCN12033                     I                     R                      N   \n",
       "RCN13585                     I                     R                      N   \n",
       "RCN12865                     I                     R                      N   \n",
       "RCN18593                     I                     R                      N   \n",
       "RCN13407                     I                     R                      N   \n",
       "RCN15343                     I                     R                      N   \n",
       "RCN12796                     I                     R                      N   \n",
       "RCN25190                     I                     R                      N   \n",
       "RCN13333                     I                     R                      N   \n",
       "RCN14642                     I                     R                      N   \n",
       "RCN15274                     I                     R                      N   \n",
       "RCN12862                     I                     R                      N   \n",
       "RCN12855                     I                     R                      N   \n",
       "RCN12480                     I                     R                      N   \n",
       "RCN12025                     I                     R                      N   \n",
       "RCN25946                     I                     R                      N   \n",
       "RCN12789                     I                     R                      N   \n",
       "RCN12854                     I                     R                      N   \n",
       "RCN12484                     I                     R                      N   \n",
       "RCN15359                     I                     R                      N   \n",
       "\n",
       "         core_mutation_dhfr_164 core_mutation_crt_72 core_mutation_crt_74  \\\n",
       "sample                                                                      \n",
       "RCN13568                      I                    C                    I   \n",
       "RCN13560                      I                    C                    I   \n",
       "RCN15107                      I                    C                    I   \n",
       "RCN14580                      I                    C                    I   \n",
       "RCN13532                      I                    C                    I   \n",
       "RCN13481                      I                    C                    I   \n",
       "RCN12032                      I                    C                    I   \n",
       "RCN13501                      I                    C                    I   \n",
       "RCN12873                      I                    C                    I   \n",
       "RCN12710                      I                    C                    I   \n",
       "RCN13368                      I                    C                    I   \n",
       "RCN12729                      I                    C                    I   \n",
       "RCN13528                      I                    C                    I   \n",
       "RCN12087                      I                    C                    I   \n",
       "RCN12033                      I                    C                    I   \n",
       "RCN13585                      I                    C                    I   \n",
       "RCN12865                      I                    C                    I   \n",
       "RCN18593                    I,L                    C                    I   \n",
       "RCN13407                    I,L                    C                    I   \n",
       "RCN15343                      L                    C                    I   \n",
       "RCN12796                      L                    C                    I   \n",
       "RCN25190                      L                    C                    I   \n",
       "RCN13333                      L                    C                    I   \n",
       "RCN14642                      L                    C                    I   \n",
       "RCN15274                      L                    C                    I   \n",
       "RCN12862                      L                    C                    I   \n",
       "RCN12855                      L                    C                    I   \n",
       "RCN12480                      L                    C                    I   \n",
       "RCN12025                      L                    C                    I   \n",
       "RCN25946                      L                    C                    I   \n",
       "RCN12789                      L                    C                    I   \n",
       "RCN12854                      L                    C                    I   \n",
       "RCN12484                      L                    C                    I   \n",
       "RCN15359                      L                    C                    I   \n",
       "\n",
       "         core_mutation_crt_75 core_mutation_crt_76 core_mutation_crt_326  \\\n",
       "sample                                                                     \n",
       "RCN13568                    D                    T                     N   \n",
       "RCN13560                    D                    T                     N   \n",
       "RCN15107                    E                    T                     N   \n",
       "RCN14580                    E                    T                     N   \n",
       "RCN13532                    E                    T                     S   \n",
       "RCN13481                    E                    T                     S   \n",
       "RCN12032                    E                    T                     S   \n",
       "RCN13501                    E                    T                     S   \n",
       "RCN12873                    E                    T                     S   \n",
       "RCN12710                    E                    T                     S   \n",
       "RCN13368                    E                    T                     S   \n",
       "RCN12729                    E                    T                     S   \n",
       "RCN13528                    E                    T                     S   \n",
       "RCN12087                    E                    T                     S   \n",
       "RCN12033                    E                    T                     S   \n",
       "RCN13585                    E                    T                     S   \n",
       "RCN12865                    E                    T                     S   \n",
       "RCN18593                    E                    T                     S   \n",
       "RCN13407                    E                    T                     S   \n",
       "RCN15343                    D                    T                     N   \n",
       "RCN12796                    E                    T                     S   \n",
       "RCN25190                    E                    T                     S   \n",
       "RCN13333                    E                    T                     S   \n",
       "RCN14642                    E                    T                     S   \n",
       "RCN15274                    E                    T                     S   \n",
       "RCN12862                    E                    T                     S   \n",
       "RCN12855                    E                    T                     S   \n",
       "RCN12480                    E                    T                     S   \n",
       "RCN12025                    E                    T                     S   \n",
       "RCN25946                    E                    T                     S   \n",
       "RCN12789                    E                    T                     S   \n",
       "RCN12854                    E                    T                     S   \n",
       "RCN12484                    E                    T                     S   \n",
       "RCN15359                    E                    T                     S   \n",
       "\n",
       "         core_mutation_crt_356  ... ENA_acc_GenRe_GRC2 ENA_acc_GenRe_SPEC  \\\n",
       "sample                          ...                                         \n",
       "RCN13568                     I  ...        ERR14390746        ERR14390747   \n",
       "RCN13560                     I  ...        ERR14390722        ERR14390723   \n",
       "RCN15107                     I  ...        ERR14392567        ERR14392568   \n",
       "RCN14580                     I  ...        ERR14391280        ERR14391281   \n",
       "RCN13532                     T  ...        ERR14390644        ERR14390645   \n",
       "RCN13481                     T  ...        ERR14390500        ERR14390501   \n",
       "RCN12032                     T  ...        ERR14388625        ERR14388626   \n",
       "RCN13501                     T  ...        ERR14390560        ERR14390561   \n",
       "RCN12873                     T  ...        ERR14389801        ERR14389802   \n",
       "RCN12710                     T  ...        ERR14389408        ERR14389409   \n",
       "RCN13368                     T  ...        ERR14390182        ERR14390183   \n",
       "RCN12729                     T  ...        ERR14389462        ERR14389463   \n",
       "RCN13528                     T  ...        ERR14390632        ERR14390633   \n",
       "RCN12087                     T  ...        ERR14388775        ERR14388776   \n",
       "RCN12033                     T  ...        ERR14388628        ERR14388629   \n",
       "RCN13585                     T  ...        ERR14390794        ERR14390795   \n",
       "RCN12865                     T  ...        ERR14389780        ERR14389781   \n",
       "RCN18593                     T  ...        ERR14394295        ERR14394296   \n",
       "RCN13407                     T  ...        ERR14390296        ERR14390297   \n",
       "RCN15343                     I  ...        ERR14393233        ERR14393234   \n",
       "RCN12796                     T  ...        ERR14389594        ERR14389595   \n",
       "RCN25190                     T  ...        ERR14395864        ERR14395865   \n",
       "RCN13333                     T  ...        ERR14390083        ERR14390084   \n",
       "RCN14642                     T  ...        ERR14391457        ERR14391458   \n",
       "RCN15274                     T  ...        ERR14393038        ERR14393039   \n",
       "RCN12862                     T  ...        ERR14389777        ERR14389778   \n",
       "RCN12855                     T  ...        ERR14389759        ERR14389760   \n",
       "RCN12480                     T  ...        ERR14388970        ERR14388971   \n",
       "RCN12025                     T  ...        ERR14388604        ERR14388605   \n",
       "RCN25946                     T  ...        ERR14396599        ERR14396600   \n",
       "RCN12789                     T  ...        ERR14389576        ERR14389577   \n",
       "RCN12854                     T  ...        ERR14389756        ERR14389757   \n",
       "RCN12484                     T  ...        ERR14388982        ERR14388983   \n",
       "RCN15359                     T  ...        ERR14393278        ERR14393279   \n",
       "\n",
       "                                 GenRe_GRC1_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN13568  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/045/...   \n",
       "RCN13560  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   \n",
       "RCN15107  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/066/...   \n",
       "RCN14580  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...   \n",
       "RCN13532  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/043/...   \n",
       "RCN13481  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/099/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   \n",
       "RCN13501  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/059/...   \n",
       "RCN12873  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...   \n",
       "RCN12710  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   \n",
       "RCN13368  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...   \n",
       "RCN12729  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/061/...   \n",
       "RCN13528  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/031/...   \n",
       "RCN12087  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/074/...   \n",
       "RCN12033  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/027/...   \n",
       "RCN13585  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/093/...   \n",
       "RCN12865  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...   \n",
       "RCN18593  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...   \n",
       "RCN13407  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...   \n",
       "RCN15343  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/032/...   \n",
       "RCN12796  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/093/...   \n",
       "RCN25190  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "RCN13333  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...   \n",
       "RCN14642  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/056/...   \n",
       "RCN15274  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/037/...   \n",
       "RCN12862  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "RCN12855  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/058/...   \n",
       "RCN12480  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/069/...   \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   \n",
       "RCN25946  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/098/...   \n",
       "RCN12789  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "RCN12854  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/055/...   \n",
       "RCN12484  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...   \n",
       "RCN15359  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "\n",
       "                                 GenRe_GRC2_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN13568  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "RCN13560  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   \n",
       "RCN15107  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/067/...   \n",
       "RCN14580  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/080/...   \n",
       "RCN13532  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/044/...   \n",
       "RCN13481  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   \n",
       "RCN13501  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/060/...   \n",
       "RCN12873  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/001/...   \n",
       "RCN12710  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...   \n",
       "RCN13368  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...   \n",
       "RCN12729  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/062/...   \n",
       "RCN13528  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/032/...   \n",
       "RCN12087  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "RCN12033  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/028/...   \n",
       "RCN13585  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...   \n",
       "RCN12865  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/080/...   \n",
       "RCN18593  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...   \n",
       "RCN13407  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/096/...   \n",
       "RCN15343  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/033/...   \n",
       "RCN12796  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...   \n",
       "RCN25190  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "RCN13333  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...   \n",
       "RCN14642  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/057/...   \n",
       "RCN15274  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/038/...   \n",
       "RCN12862  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "RCN12855  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/059/...   \n",
       "RCN12480  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   \n",
       "RCN25946  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/099/...   \n",
       "RCN12789  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "RCN12854  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/056/...   \n",
       "RCN12484  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...   \n",
       "RCN15359  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/078/...   \n",
       "\n",
       "                                 GenRe_SPEC_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN13568  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "RCN13560  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...   \n",
       "RCN15107  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/068/...   \n",
       "RCN14580  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...   \n",
       "RCN13532  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/045/...   \n",
       "RCN13481  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/001/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...   \n",
       "RCN13501  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/061/...   \n",
       "RCN12873  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/002/...   \n",
       "RCN12710  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/009/...   \n",
       "RCN13368  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...   \n",
       "RCN12729  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "RCN13528  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/033/...   \n",
       "RCN12087  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "RCN12033  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/029/...   \n",
       "RCN13585  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...   \n",
       "RCN12865  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...   \n",
       "RCN18593  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/096/...   \n",
       "RCN13407  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/097/...   \n",
       "RCN15343  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/034/...   \n",
       "RCN12796  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...   \n",
       "RCN25190  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "RCN13333  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/084/...   \n",
       "RCN14642  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/058/...   \n",
       "RCN15274  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/039/...   \n",
       "RCN12862  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/078/...   \n",
       "RCN12855  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/060/...   \n",
       "RCN12480  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...   \n",
       "RCN25946  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...   \n",
       "RCN12789  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "RCN12854  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/057/...   \n",
       "RCN12484  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...   \n",
       "RCN15359  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...   \n",
       "\n",
       "                                        Pf8_ENA_FASTQ_FTP_1  \\\n",
       "sample                                                        \n",
       "RCN13568  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/095/...   \n",
       "RCN13560  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/087/...   \n",
       "RCN15107  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/087/...   \n",
       "RCN14580  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/029/...   \n",
       "RCN13532  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/060/...   \n",
       "RCN13481  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/013/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...   \n",
       "RCN13501  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/033/...   \n",
       "RCN12873  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/075/...   \n",
       "RCN12710  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/050/...   \n",
       "RCN13368  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/008/...   \n",
       "RCN12729  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/060/...   \n",
       "RCN13528  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/056/...   \n",
       "RCN12087  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...   \n",
       "RCN12033  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/012/...   \n",
       "RCN13585  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...   \n",
       "RCN12865  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/067/...   \n",
       "RCN18593  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/034/...   \n",
       "RCN13407  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/045/...   \n",
       "RCN15343  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/097/...   \n",
       "RCN12796  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/001/...   \n",
       "RCN25190  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...   \n",
       "RCN13333  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/073/...   \n",
       "RCN14642  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/090/...   \n",
       "RCN15274  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/031/...   \n",
       "RCN12862  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...   \n",
       "RCN12855  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/057/...   \n",
       "RCN12480  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/050/...   \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...   \n",
       "RCN25946  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/059/...   \n",
       "RCN12789  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/094/...   \n",
       "RCN12854  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/056/...   \n",
       "RCN12484  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/054/...   \n",
       "RCN15359  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/012/...   \n",
       "\n",
       "                                 GenRe_GRC1_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN13568  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/045/...   \n",
       "RCN13560  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   \n",
       "RCN15107  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/066/...   \n",
       "RCN14580  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...   \n",
       "RCN13532  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/043/...   \n",
       "RCN13481  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/099/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   \n",
       "RCN13501  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/059/...   \n",
       "RCN12873  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...   \n",
       "RCN12710  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   \n",
       "RCN13368  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...   \n",
       "RCN12729  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/061/...   \n",
       "RCN13528  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/031/...   \n",
       "RCN12087  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/074/...   \n",
       "RCN12033  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/027/...   \n",
       "RCN13585  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/093/...   \n",
       "RCN12865  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...   \n",
       "RCN18593  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...   \n",
       "RCN13407  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...   \n",
       "RCN15343  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/032/...   \n",
       "RCN12796  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/093/...   \n",
       "RCN25190  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "RCN13333  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...   \n",
       "RCN14642  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/056/...   \n",
       "RCN15274  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/037/...   \n",
       "RCN12862  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "RCN12855  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/058/...   \n",
       "RCN12480  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/069/...   \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   \n",
       "RCN25946  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/098/...   \n",
       "RCN12789  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "RCN12854  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/055/...   \n",
       "RCN12484  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...   \n",
       "RCN15359  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "\n",
       "                                 GenRe_GRC2_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN13568  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "RCN13560  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   \n",
       "RCN15107  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/067/...   \n",
       "RCN14580  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/080/...   \n",
       "RCN13532  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/044/...   \n",
       "RCN13481  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   \n",
       "RCN13501  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/060/...   \n",
       "RCN12873  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/001/...   \n",
       "RCN12710  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...   \n",
       "RCN13368  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...   \n",
       "RCN12729  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/062/...   \n",
       "RCN13528  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/032/...   \n",
       "RCN12087  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "RCN12033  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/028/...   \n",
       "RCN13585  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...   \n",
       "RCN12865  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/080/...   \n",
       "RCN18593  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...   \n",
       "RCN13407  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/096/...   \n",
       "RCN15343  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/033/...   \n",
       "RCN12796  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/094/...   \n",
       "RCN25190  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "RCN13333  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...   \n",
       "RCN14642  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/057/...   \n",
       "RCN15274  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/038/...   \n",
       "RCN12862  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "RCN12855  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/059/...   \n",
       "RCN12480  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   \n",
       "RCN25946  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/099/...   \n",
       "RCN12789  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "RCN12854  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/056/...   \n",
       "RCN12484  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/082/...   \n",
       "RCN15359  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/078/...   \n",
       "\n",
       "                                 GenRe_SPEC_ENA_FASTQ_FTP_2  \\\n",
       "sample                                                        \n",
       "RCN13568  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "RCN13560  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...   \n",
       "RCN15107  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/068/...   \n",
       "RCN14580  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...   \n",
       "RCN13532  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/045/...   \n",
       "RCN13481  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/001/...   \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...   \n",
       "RCN13501  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/061/...   \n",
       "RCN12873  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/002/...   \n",
       "RCN12710  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/009/...   \n",
       "RCN13368  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...   \n",
       "RCN12729  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "RCN13528  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/033/...   \n",
       "RCN12087  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "RCN12033  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/029/...   \n",
       "RCN13585  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...   \n",
       "RCN12865  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/081/...   \n",
       "RCN18593  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/096/...   \n",
       "RCN13407  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/097/...   \n",
       "RCN15343  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/034/...   \n",
       "RCN12796  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/095/...   \n",
       "RCN25190  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "RCN13333  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/084/...   \n",
       "RCN14642  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/058/...   \n",
       "RCN15274  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/039/...   \n",
       "RCN12862  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/078/...   \n",
       "RCN12855  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/060/...   \n",
       "RCN12480  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...   \n",
       "RCN25946  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/000/...   \n",
       "RCN12789  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "RCN12854  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/057/...   \n",
       "RCN12484  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/083/...   \n",
       "RCN15359  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/079/...   \n",
       "\n",
       "                                        Pf8_ENA_FASTQ_FTP_2  \n",
       "sample                                                       \n",
       "RCN13568  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/095/...  \n",
       "RCN13560  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/087/...  \n",
       "RCN15107  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/087/...  \n",
       "RCN14580  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/029/...  \n",
       "RCN13532  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/060/...  \n",
       "RCN13481  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/013/...  \n",
       "RCN12032  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...  \n",
       "RCN13501  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/033/...  \n",
       "RCN12873  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/075/...  \n",
       "RCN12710  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/050/...  \n",
       "RCN13368  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/008/...  \n",
       "RCN12729  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/060/...  \n",
       "RCN13528  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/056/...  \n",
       "RCN12087  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...  \n",
       "RCN12033  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/012/...  \n",
       "RCN13585  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...  \n",
       "RCN12865  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/067/...  \n",
       "RCN18593  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/034/...  \n",
       "RCN13407  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/045/...  \n",
       "RCN15343  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/097/...  \n",
       "RCN12796  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/001/...  \n",
       "RCN25190  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...  \n",
       "RCN13333  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/073/...  \n",
       "RCN14642  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/090/...  \n",
       "RCN15274  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/031/...  \n",
       "RCN12862  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...  \n",
       "RCN12855  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/057/...  \n",
       "RCN12480  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/050/...  \n",
       "RCN12025  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...  \n",
       "RCN25946  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/059/...  \n",
       "RCN12789  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/094/...  \n",
       "RCN12854  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/056/...  \n",
       "RCN12484  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/054/...  \n",
       "RCN15359  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/012/...  \n",
       "\n",
       "[34 rows x 128 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genre_pf8_merged_seq_gt_concord_patterns_df = genre_pf8_merged_seq_gt_concord_small_ena2_df.reset_index().groupby( core_mut_cols)['sample'].min().reset_index().set_index('sample')\n",
    "\n",
    "# move the samples column to first\n",
    "cols = list(genre_pf8_merged_seq_gt_concord_patterns_df.columns)\n",
    "genre_pf8_merged_seq_gt_concord_patterns_df = genre_pf8_merged_seq_gt_concord_patterns_df[cols]\n",
    "\n",
    "# add the ENA accession and download link columns back in\n",
    "genre_pf8_merged_seq_gt_concord_patterns_final_df = pd.merge(\n",
    "    genre_pf8_merged_seq_gt_concord_patterns_df,\n",
    "    genre_pf8_merged_seq_gt_concord_small_ena2_df.filter(regex='ENA_'),\n",
    "    left_index=True,\n",
    "    right_index=True\n",
    ")\n",
    "genre_pf8_merged_seq_gt_concord_patterns_final_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b725a297-db30-43b5-ab0e-9bf3665298f9",
   "metadata": {},
   "source": [
    "### Save file\n",
    "Sumary: this file contains 34 \"representative samples\", each represent a distinct pattern of drug-resistance haplotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d6f6b2f4-10a1-4fa8-949f-8d8d6cb1800f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genre_pf8_merged_seq_gt_concord_patterns_final_df.to_csv('Pf8-GenReMekong_fully_concordant_genotype_patterns_representative_samples.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bca53019-c04b-421b-8f98-79ee41b9bb48",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4181137-c188-45b2-b385-a0b65b63c6c4",
   "metadata": {},
   "source": [
    "## Retrieve FASTQ files\n",
    "A Python script \"ENA_data_helper.py\" is provided alongside this notebook and can be used to download the FASTQ files from ENA.  \n",
    "The full download is time intenstive and uses a large amount of disk space and is done performed in this notbook.  \n",
    "\n",
    "The script works with the ENA search results produced earlier, but it needs the FASTQ FTP URLs to be split into two columns.  \n",
    "In addition, the ENA search results should be filtered by sample ID against the final data saved for each of the three datasets to ensure that data is downloaded only for those samples that are still in the final dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cc4deee-5e69-43d0-9fcd-3bae91ebf3b9",
   "metadata": {},
   "source": [
    "Concordant phenotype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "a7fb7ecd-08a1-491e-92e5-fad18a9af61d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = genre_pf8_conc_s_ena_result[\n",
    "    genre_pf8_conc_s_ena_result['sample'].isin(genre_pf8_conc_s_ena2_df.index)\n",
    "]\n",
    "df[['ftp_url_read_1', 'ftp_url_read_2']] = df['fastq_ftp'].str.split(';', expand=True)\n",
    "df = df[['run_accession','sample','ftp_url_read_1','ftp_url_read_2']]\n",
    "df.to_csv('ENA_download_manifest_concordant_phenotype_high_quality_samples.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "88538f7d-0f9c-4985-9c17-6fed9e0f7fe8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = genre_pf8_merged_seq_gt_concord_small_ena_result[\n",
    "    genre_pf8_merged_seq_gt_concord_small_ena_result['sample'].isin(genre_pf8_merged_seq_gt_concord_small_ena2_df.index)\n",
    "]\n",
    "df[['ftp_url_read_1', 'ftp_url_read_2']] = df['fastq_ftp'].str.split(';', expand=True)\n",
    "df = df[['run_accession','sample','ftp_url_read_1','ftp_url_read_2']]\n",
    "df.to_csv('ENA_download_manifest_fully_concordant_genotype_high_quality_samples.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "19cb2054-5390-45cb-8151-d14c02eebc10",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = genre_pf8_merged_seq_gt_concord_small_ena_result[\n",
    "    genre_pf8_merged_seq_gt_concord_small_ena_result['sample'].isin(genre_pf8_merged_seq_gt_concord_patterns_final_df.index)\n",
    "]\n",
    "df[['ftp_url_read_1', 'ftp_url_read_2']] = df['fastq_ftp'].str.split(';', expand=True)\n",
    "df = df[['run_accession','sample','ftp_url_read_1','ftp_url_read_2']]\n",
    "df.to_csv('ENA_download_manifest_fully_concordant_genotype_patterns_representative_samples.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9c145fa-e1f9-4f3e-b8bf-c67d521922f2",
   "metadata": {},
   "source": [
    "***"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
