{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e0088e6-1fc9-4e89-bfdc-3490c7385928",
   "metadata": {},
   "source": [
    "# Prepartion for read count stats for simulated data\n",
    "## Aim of this notebook\n",
    "This notebook prepares read-count stats for a selection of real-world data from the [GenRe Mekong project](https://www.malariagen.net/resource/29/) for the three AmpSeq primer panels used in SpotMalaria (GRC1, GRC2 and SPEC).  \n",
    "Data for the SpotMalaria panels is available in [this spreadsheet](https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xls).  \n",
    "The read counts are used as guides for the creation of simulated data.  \n",
    "\n",
    "## FASTQ data\n",
    "The FASTQ files used in this notebook are based on the ENA data download manifests created in [this notebook](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb). But __note__ that the FASTQ files are __not__ icluded in this repo because of the large amount of data contained in them. If you would like to re-run this notebook, please refer to the explanation provided [here](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb#Example:-how-to-use-the-downloader-tool) to download the FASTQ files from ENA using the downloader tool provided in this repo.\n",
    "\n",
    "For the purpose of this analysis, the dataset [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv), created [here](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb#Save-results:-concordant-phenotype-data) was used. This contains an essentially random subset of 907 samples of high-quality data from the GenRe Mekong project. This is sufficient for the purpose of obtaining guides for real-world read counts of the AmpSeq panels.  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9033074e-7289-4f7e-856d-696489e35e7b",
   "metadata": {},
   "source": [
    "## Creating the read counts data file\n",
    "A file \"fastq_readcounts.csv\" is already prepared in this directory and used by this notebook. \n",
    "\n",
    "It is recommended to use that file as provided in this folder. The following instructions are only needed to re-create that file and to document how it was generated. \n",
    "\n",
    "Ensure that the following Python packages are installed on your system:\n",
    "- requests\n",
    "- pandas\n",
    "\n",
    "Then, in a terminal, run the following commands. The commands can be found alongside this notebook.   \n",
    "Both commands will take a potentially long time to run (depending on your system and network) and the first command will download a large amount of data over the internet. If you are working in a HPC/cloud environment, you may want to split the data into chunks and parallelise these tasks. This is beyond the scope of this recipe.  \n",
    "\n",
    "__Step 1: Download FASTQ files__    \n",
    "Run the following command in a terminal:\n",
    "```bash\n",
    "python ../real-world_gold-standard_data/Pf8-GenReMekong/ENA_data_helper.py \\\n",
    "    download \\\n",
    "    --data ../real-world_gold-standard_data/Pf8 GenReMekong/ENA_download_manifest_concordant_phenotype_high_quality_samples.csv \\\n",
    "    --skip_errors \\\n",
    "    --download_attempts 10 \\\n",
    "    --out ENA_download_concordant_phenotype_high_quality_sample \\\n",
    "```\n",
    "It will create a folder \"ENA_download_concordant_phenotype_high_quality_sample\" in this directory. The folder contains the 3628 FASTQ (gzipped) files belonging to the [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv), created [here](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb#Save-results:-concordant-phenotype-data).  \n",
    "\n",
    "__Step 2: Obtain read counts__  \n",
    "Once the FASTQ download has completed, use the script \"run_counts.sh\", also in this folder, like this:  \n",
    "(or change the path to the downloaded FASTQ files if different)\n",
    "\n",
    "```bash\n",
    "./run_counts.sh ENA_download_concordant_phenotype_high_quality_sample\n",
    "```  \n",
    "This will create a file \"fastq_readcounts.csv\" with file names and read counts for all mate1 files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4671e11d-e3c2-4644-aab0-0220ac9035ad",
   "metadata": {},
   "source": [
    "## Calculate AmpSeq panel stats\n",
    "Using the dataset [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv), link the FASTQ read counts to AmpSeq panels and calcualte stats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "5619e6ee-c39b-4cdf-b0cf-b1cfc32b0880",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8dc8d3bc-d7ea-4954-a9f2-09f92f564cbb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR14388603_1.fastq.gz</td>\n",
       "      <td>32819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR14388604_1.fastq.gz</td>\n",
       "      <td>29409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR14388605_1.fastq.gz</td>\n",
       "      <td>1094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR14388606_1.fastq.gz</td>\n",
       "      <td>38189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR14388607_1.fastq.gz</td>\n",
       "      <td>35804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3254</th>\n",
       "      <td>ERR15627188_1.fastq.gz</td>\n",
       "      <td>14927105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3255</th>\n",
       "      <td>ERR15627189_1.fastq.gz</td>\n",
       "      <td>12862572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3256</th>\n",
       "      <td>ERR15627190_1.fastq.gz</td>\n",
       "      <td>11781727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3257</th>\n",
       "      <td>ERR15627191_1.fastq.gz</td>\n",
       "      <td>12052828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3258</th>\n",
       "      <td>ERR15627192_1.fastq.gz</td>\n",
       "      <td>9728637</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3259 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file     count\n",
       "0     ERR14388603_1.fastq.gz     32819\n",
       "1     ERR14388604_1.fastq.gz     29409\n",
       "2     ERR14388605_1.fastq.gz      1094\n",
       "3     ERR14388606_1.fastq.gz     38189\n",
       "4     ERR14388607_1.fastq.gz     35804\n",
       "...                      ...       ...\n",
       "3254  ERR15627188_1.fastq.gz  14927105\n",
       "3255  ERR15627189_1.fastq.gz  12862572\n",
       "3256  ERR15627190_1.fastq.gz  11781727\n",
       "3257  ERR15627191_1.fastq.gz  12052828\n",
       "3258  ERR15627192_1.fastq.gz   9728637\n",
       "\n",
       "[3259 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df = pd.read_csv('fastq_readcounts.csv')\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d6160fab-2fc3-4700-bda3-8bac7c28f01b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>DHA-PPQ</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>Species</th>\n",
       "      <th>...</th>\n",
       "      <th>ENA_acc_GenRe_GRC2</th>\n",
       "      <th>ENA_acc_GenRe_SPEC</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_1</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_2</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RCN12025</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388604</td>\n",
       "      <td>ERR14388605</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RCN12026</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388607</td>\n",
       "      <td>ERR14388608</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RCN12028</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388613</td>\n",
       "      <td>ERR14388614</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RCN12031</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388622</td>\n",
       "      <td>ERR14388623</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RCN12032</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388625</td>\n",
       "      <td>ERR14388626</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>902</th>\n",
       "      <td>RCN26767</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397547</td>\n",
       "      <td>ERR14397548</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>RCN26775</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397571</td>\n",
       "      <td>ERR14397572</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>RCN26820</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397664</td>\n",
       "      <td>ERR14397665</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>RCN26932</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397976</td>\n",
       "      <td>ERR14397977</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906</th>\n",
       "      <td>RCN26938</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397991</td>\n",
       "      <td>ERR14397992</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>907 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample Artemisinin Piperaquine Mefloquine Chloroquine Pyrimethamine  \\\n",
       "0    RCN12025   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "1    RCN12026   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "2    RCN12028   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3    RCN12031   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "4    RCN12032   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "..        ...         ...         ...        ...         ...           ...   \n",
       "902  RCN26767   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "903  RCN26775   Sensitive   Sensitive  Sensitive   Resistant     Resistant   \n",
       "904  RCN26820   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "905  RCN26932   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "906  RCN26938   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "\n",
       "    Sulfadoxine    DHA-PPQ      AS-MQ Species  ... ENA_acc_GenRe_GRC2  \\\n",
       "0     Resistant  Sensitive  Sensitive      Pf  ...        ERR14388604   \n",
       "1     Resistant  Resistant  Sensitive      Pf  ...        ERR14388607   \n",
       "2     Resistant  Resistant  Sensitive      Pf  ...        ERR14388613   \n",
       "3     Resistant  Resistant  Sensitive      Pf  ...        ERR14388622   \n",
       "4     Resistant  Sensitive  Sensitive      Pf  ...        ERR14388625   \n",
       "..          ...        ...        ...     ...  ...                ...   \n",
       "902   Resistant  Sensitive  Sensitive      Pf  ...        ERR14397547   \n",
       "903   Sensitive  Sensitive  Sensitive      Pf  ...        ERR14397571   \n",
       "904   Resistant  Sensitive  Sensitive      Pf  ...        ERR14397664   \n",
       "905   Resistant  Sensitive  Sensitive      Pf  ...        ERR14397976   \n",
       "906   Resistant  Sensitive  Sensitive      Pf  ...        ERR14397991   \n",
       "\n",
       "     ENA_acc_GenRe_SPEC                         GenRe_GRC1_ENA_FASTQ_FTP_1  \\\n",
       "0           ERR14388605  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   \n",
       "1           ERR14388608  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...   \n",
       "2           ERR14388614  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...   \n",
       "3           ERR14388623  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   \n",
       "4           ERR14388626  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   \n",
       "..                  ...                                                ...   \n",
       "902         ERR14397548  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "903         ERR14397572  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   \n",
       "904         ERR14397665  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "905         ERR14397977  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "906         ERR14397992  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "\n",
       "                            GenRe_GRC2_ENA_FASTQ_FTP_1  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...   \n",
       "\n",
       "                            GenRe_SPEC_ENA_FASTQ_FTP_1  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...   \n",
       "\n",
       "                                   Pf8_ENA_FASTQ_FTP_1  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...   \n",
       "\n",
       "                            GenRe_GRC1_ENA_FASTQ_FTP_2  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "\n",
       "                            GenRe_GRC2_ENA_FASTQ_FTP_2  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...   \n",
       "\n",
       "                            GenRe_SPEC_ENA_FASTQ_FTP_2  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...   \n",
       "\n",
       "                                   Pf8_ENA_FASTQ_FTP_2  \n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...  \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...  \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...  \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...  \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...  \n",
       "..                                                 ...  \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...  \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...  \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...  \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...  \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...  \n",
       "\n",
       "[907 rows x 24 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_df = pd.read_csv('../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv')\n",
    "samples_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c8d0e88-7d68-4a23-bd31-92ed22f872b7",
   "metadata": {},
   "source": [
    "Assign FASTQ files back to GenRe AmpSeq primer panel or to Pf8 (WGS), then drop the Pf8 rows, which are not relevant here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9c876cc3-0795-40e6-8d9e-f11372cab460",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "counts_df['panel']=counts_df.apply(\n",
    "    lambda row: \n",
    "    'SPEC' if samples_df.GenRe_SPEC_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'GRC1' if samples_df.GenRe_GRC1_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'GRC2' if samples_df.GenRe_GRC2_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'Pf8' if samples_df.Pf8_ENA_FASTQ_FTP_1.str.contains(row.file).any() else None, axis=1)\n",
    "counts_df = counts_df[ counts_df['panel'].isin(['SPEC','GRC1','GRC2'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "731cdf31-bd8a-421c-8aee-f04f4eeb4ba1",
   "metadata": {},
   "source": [
    "Sanity check: we should have 907 samples for each GenRe primer panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "51e781cd-c5c9-4637-a499-36e5651ba811",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>panel</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRC1</td>\n",
       "      <td>907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRC2</td>\n",
       "      <td>907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPEC</td>\n",
       "      <td>907</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  panel  count\n",
       "0  GRC1    907\n",
       "1  GRC2    907\n",
       "2  SPEC    907"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.groupby('panel').size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea3afc4d-9f6e-497c-ad9c-abfe7bf14b50",
   "metadata": {},
   "source": [
    "Calculate stats for read counts per primer panel: minimum, maximum, mean and percentiles at 25%, 50% and 75%.  \n",
    "These will be used to guide the number of reads to simulate for each panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f0d717a1-8d71-437f-b92a-49f61f00f6cd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"6\" halign=\"left\">count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>q25</th>\n",
       "      <th>q50</th>\n",
       "      <th>q75</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>panel</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRC1</th>\n",
       "      <td>33917.249173</td>\n",
       "      <td>18</td>\n",
       "      <td>96442</td>\n",
       "      <td>25568.5</td>\n",
       "      <td>32424.0</td>\n",
       "      <td>40281.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRC2</th>\n",
       "      <td>33500.464168</td>\n",
       "      <td>21</td>\n",
       "      <td>78617</td>\n",
       "      <td>26732.5</td>\n",
       "      <td>32519.0</td>\n",
       "      <td>39864.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPEC</th>\n",
       "      <td>1291.057332</td>\n",
       "      <td>0</td>\n",
       "      <td>17924</td>\n",
       "      <td>654.5</td>\n",
       "      <td>988.0</td>\n",
       "      <td>1445.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              count                                      \n",
       "               mean min    max      q25      q50      q75\n",
       "panel                                                    \n",
       "GRC1   33917.249173  18  96442  25568.5  32424.0  40281.5\n",
       "GRC2   33500.464168  21  78617  26732.5  32519.0  39864.5\n",
       "SPEC    1291.057332   0  17924    654.5    988.0   1445.5"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def q25(x):\n",
    "    return x.quantile(0.25)\n",
    "def q50(x):\n",
    "    return x.quantile(0.5)\n",
    "def q75(x):\n",
    "    return x.quantile(0.75)\n",
    "\n",
    "counts_df.groupby('panel').agg({'count': ['mean', 'min','max', q25,q50,q75]})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
