{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e0088e6-1fc9-4e89-bfdc-3490c7385928",
   "metadata": {},
   "source": [
    "# Prepartion for creating simulated data\n",
    "## Aim of this notebook\n",
    "This notebook contains the preparation required to run read simulations using the read simulation pipeline published in this repository.  \n",
    "The main aim of this notebook is to provide a working recipe for a read simulation run that is based on real-world data. In contrast to whole-genome data, for which real-world datasets are readily availble from the [Pf8 resource](https://www.malariagen.net/resource/36/), amplicon data for a pipeline that may have been developed for a specific amplicon primer panel, my not be available from public resources. In this case, simulating reads may be the only option for pipeline validation. Use this notebook as a guideline on how to generate the input files for a read simulation pipeline run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97c3368e-a9ff-4f8c-b4dc-45c7fc23a23c",
   "metadata": {},
   "source": [
    "## Part 1: read counts to guide simulation\n",
    "An important consideration for a simulated dataset is the number of reads to create. This should be within the range of expected read numbers from real-world experiments because unrealistically high or low read counts will not provide useful insights into the validity of pipeline results.  \n",
    "\n",
    "In this example, we are simulating reads for a wild-type and a mutant _P. falciparum_ strain. We will simulate an AmpSeq run using the [SpotMalaria](https://www.malariagen.net/project/spotmalaria/) panel, as used for the [GenRe Mekong project](https://www.malariagen.net/resource/29/) project. \n",
    "\n",
    "This primer panel consists of three sub-panels: SPEC (speciation), GRC1 and GRC2.  \n",
    "The data for these panels is provided in [this spreadsheet](https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xls).  \n",
    "\n",
    "The notebook and data provided here is not intended to be used as-is, but to serve as a template for creating custom simulated datasets, tailored to the pipeline that is to be tested. That said, you are welcome to use the data produced here for your own pipeline, if you wish.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfee8c68-86d3-4554-adc9-05716ca8d18c",
   "metadata": {},
   "source": [
    "### FASTQ data\n",
    "The FASTQ files used in this notebook are based on the ENA data download manifests created in [this notebook](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb). But __note__ that the FASTQ files are __not__ icluded in this repo because of the large amount of data contained in them. If you would like to re-run this notebook, please refer to the explanation provided [here](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb#Example:-how-to-use-the-downloader-tool) to download the FASTQ files from ENA using the downloader tool provided in this repo.\n",
    "\n",
    "For the purpose of this analysis, the dataset [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv), created [here](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb#Save-results:-concordant-phenotype-data) was used. This contains a subset of 907 samples of all high-quality data from the GenRe Mekong project where drug resistance phenotype agrees with the WGS sample analysis in Pf8.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9033074e-7289-4f7e-856d-696489e35e7b",
   "metadata": {},
   "source": [
    "### Creating the read counts data file\n",
    "A file \"fastq_readcounts.csv\" is already prepared in this directory and used by this notebook. \n",
    "\n",
    "It is recommended to use that file as provided in this folder. The following instructions are only needed to re-create that file and to document how it was generated. \n",
    "\n",
    "Ensure that the following Python packages are installed on your system:\n",
    "- requests\n",
    "- pandas\n",
    "\n",
    "Then, in a terminal, run the following commands. The commands can be found alongside this notebook.   \n",
    "Both commands will take a potentially long time to run (depending on your system and network) and the first command will download a large amount of data over the internet. If you are working in a HPC/cloud environment, you may want to split the data into chunks and parallelise these tasks. This is beyond the scope of this recipe.  \n",
    "\n",
    "__Step 1: Download FASTQ files__    \n",
    "Run the following command in a terminal:\n",
    "```bash\n",
    "python ../real-world_gold-standard_data/Pf8-GenReMekong/ENA_data_helper.py \\\n",
    "    download \\\n",
    "    --data ../real-world_gold-standard_data/Pf8 GenReMekong/ENA_download_manifest_concordant_phenotype_high_quality_samples.csv \\\n",
    "    --skip_errors \\\n",
    "    --download_attempts 10 \\\n",
    "    --out ENA_download_concordant_phenotype_high_quality_sample \\\n",
    "```\n",
    "It will create a folder \"ENA_download_concordant_phenotype_high_quality_sample\" in this directory. The folder contains the 3628 FASTQ (gzipped) files belonging to the [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv), created [here](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb#Save-results:-concordant-phenotype-data).  \n",
    "\n",
    "__Step 2: Obtain read counts__  \n",
    "Once the FASTQ download has completed, use the script \"run_counts.sh\", also in this folder, like this:  \n",
    "(or change the path to the downloaded FASTQ files if different)\n",
    "\n",
    "```bash\n",
    "./run_counts.sh ENA_download_concordant_phenotype_high_quality_sample\n",
    "```  \n",
    "This will create a file \"fastq_readcounts.csv\" with file names and read counts for all mate1 files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4671e11d-e3c2-4644-aab0-0220ac9035ad",
   "metadata": {},
   "source": [
    "### Calculate AmpSeq panel stats\n",
    "Using the dataset [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv) and the pre-computed (see above) csv file of read counts, create read count stats for the three SpotMalaria amplicon panels.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5619e6ee-c39b-4cdf-b0cf-b1cfc32b0880",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8dc8d3bc-d7ea-4954-a9f2-09f92f564cbb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR14388603_1.fastq.gz</td>\n",
       "      <td>32819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR14388604_1.fastq.gz</td>\n",
       "      <td>29409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR14388605_1.fastq.gz</td>\n",
       "      <td>1094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR14388606_1.fastq.gz</td>\n",
       "      <td>38189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR14388607_1.fastq.gz</td>\n",
       "      <td>35804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3623</th>\n",
       "      <td>ERR15628752_1.fastq.gz</td>\n",
       "      <td>18788821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3624</th>\n",
       "      <td>ERR15628753_1.fastq.gz</td>\n",
       "      <td>13952241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3625</th>\n",
       "      <td>ERR15628764_1.fastq.gz</td>\n",
       "      <td>14159323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3626</th>\n",
       "      <td>ERR15628798_1.fastq.gz</td>\n",
       "      <td>7910579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3627</th>\n",
       "      <td>ERR15628802_1.fastq.gz</td>\n",
       "      <td>9683996</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3628 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file     count\n",
       "0     ERR14388603_1.fastq.gz     32819\n",
       "1     ERR14388604_1.fastq.gz     29409\n",
       "2     ERR14388605_1.fastq.gz      1094\n",
       "3     ERR14388606_1.fastq.gz     38189\n",
       "4     ERR14388607_1.fastq.gz     35804\n",
       "...                      ...       ...\n",
       "3623  ERR15628752_1.fastq.gz  18788821\n",
       "3624  ERR15628753_1.fastq.gz  13952241\n",
       "3625  ERR15628764_1.fastq.gz  14159323\n",
       "3626  ERR15628798_1.fastq.gz   7910579\n",
       "3627  ERR15628802_1.fastq.gz   9683996\n",
       "\n",
       "[3628 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df = pd.read_csv('fastq_readcounts.csv')\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6160fab-2fc3-4700-bda3-8bac7c28f01b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>DHA-PPQ</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>Species</th>\n",
       "      <th>...</th>\n",
       "      <th>ENA_acc_GenRe_GRC2</th>\n",
       "      <th>ENA_acc_GenRe_SPEC</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_1</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_1</th>\n",
       "      <th>GenRe_GRC1_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_GRC2_ENA_FASTQ_FTP_2</th>\n",
       "      <th>GenRe_SPEC_ENA_FASTQ_FTP_2</th>\n",
       "      <th>Pf8_ENA_FASTQ_FTP_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RCN12025</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388604</td>\n",
       "      <td>ERR14388605</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RCN12026</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388607</td>\n",
       "      <td>ERR14388608</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RCN12028</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388613</td>\n",
       "      <td>ERR14388614</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RCN12031</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388622</td>\n",
       "      <td>ERR14388623</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RCN12032</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14388625</td>\n",
       "      <td>ERR14388626</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>902</th>\n",
       "      <td>RCN26767</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397547</td>\n",
       "      <td>ERR14397548</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>RCN26775</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397571</td>\n",
       "      <td>ERR14397572</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>RCN26820</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397664</td>\n",
       "      <td>ERR14397665</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>RCN26932</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397976</td>\n",
       "      <td>ERR14397977</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906</th>\n",
       "      <td>RCN26938</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR14397991</td>\n",
       "      <td>ERR14397992</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>907 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample Artemisinin Piperaquine Mefloquine Chloroquine Pyrimethamine  \\\n",
       "0    RCN12025   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "1    RCN12026   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "2    RCN12028   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3    RCN12031   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "4    RCN12032   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "..        ...         ...         ...        ...         ...           ...   \n",
       "902  RCN26767   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "903  RCN26775   Sensitive   Sensitive  Sensitive   Resistant     Resistant   \n",
       "904  RCN26820   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "905  RCN26932   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "906  RCN26938   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "\n",
       "    Sulfadoxine    DHA-PPQ      AS-MQ Species  ... ENA_acc_GenRe_GRC2  \\\n",
       "0     Resistant  Sensitive  Sensitive      Pf  ...        ERR14388604   \n",
       "1     Resistant  Resistant  Sensitive      Pf  ...        ERR14388607   \n",
       "2     Resistant  Resistant  Sensitive      Pf  ...        ERR14388613   \n",
       "3     Resistant  Resistant  Sensitive      Pf  ...        ERR14388622   \n",
       "4     Resistant  Sensitive  Sensitive      Pf  ...        ERR14388625   \n",
       "..          ...        ...        ...     ...  ...                ...   \n",
       "902   Resistant  Sensitive  Sensitive      Pf  ...        ERR14397547   \n",
       "903   Sensitive  Sensitive  Sensitive      Pf  ...        ERR14397571   \n",
       "904   Resistant  Sensitive  Sensitive      Pf  ...        ERR14397664   \n",
       "905   Resistant  Sensitive  Sensitive      Pf  ...        ERR14397976   \n",
       "906   Resistant  Sensitive  Sensitive      Pf  ...        ERR14397991   \n",
       "\n",
       "     ENA_acc_GenRe_SPEC                         GenRe_GRC1_ENA_FASTQ_FTP_1  \\\n",
       "0           ERR14388605  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   \n",
       "1           ERR14388608  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...   \n",
       "2           ERR14388614  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...   \n",
       "3           ERR14388623  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   \n",
       "4           ERR14388626  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   \n",
       "..                  ...                                                ...   \n",
       "902         ERR14397548  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "903         ERR14397572  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   \n",
       "904         ERR14397665  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "905         ERR14397977  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "906         ERR14397992  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "\n",
       "                            GenRe_GRC2_ENA_FASTQ_FTP_1  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...   \n",
       "\n",
       "                            GenRe_SPEC_ENA_FASTQ_FTP_1  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...   \n",
       "\n",
       "                                   Pf8_ENA_FASTQ_FTP_1  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...   \n",
       "\n",
       "                            GenRe_GRC1_ENA_FASTQ_FTP_2  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/003/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/006/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/012/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/021/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/024/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/046/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/070/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/063/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/075/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/090/...   \n",
       "\n",
       "                            GenRe_GRC2_ENA_FASTQ_FTP_2  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/004/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/007/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/013/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/022/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/025/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/047/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/071/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/064/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/076/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/091/...   \n",
       "\n",
       "                            GenRe_SPEC_ENA_FASTQ_FTP_2  \\\n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/005/...   \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/008/...   \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/014/...   \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/023/...   \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/026/...   \n",
       "..                                                 ...   \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/048/...   \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/072/...   \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/065/...   \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/077/...   \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR143/092/...   \n",
       "\n",
       "                                   Pf8_ENA_FASTQ_FTP_2  \n",
       "0    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/006/...  \n",
       "1    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/007/...  \n",
       "2    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/009/...  \n",
       "3    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/010/...  \n",
       "4    ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/011/...  \n",
       "..                                                 ...  \n",
       "902  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/052/...  \n",
       "903  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/053/...  \n",
       "904  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/064/...  \n",
       "905  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/098/...  \n",
       "906  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR156/002/...  \n",
       "\n",
       "[907 rows x 24 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_df = pd.read_csv('../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv')\n",
    "samples_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c8d0e88-7d68-4a23-bd31-92ed22f872b7",
   "metadata": {},
   "source": [
    "Read counts in ```counts_df``` are assigned to file names that do not contain the name of the primer panel to which the file belongs. Use the ```samples_df``` data to assign the panel name (SPEC, GRC1 or GRC2) to each read count by looking up the file name.  \n",
    "Drop Pf8 (WGS) data, which is not relevant here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c876cc3-0795-40e6-8d9e-f11372cab460",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "counts_df['panel']=counts_df.apply(\n",
    "    lambda row: \n",
    "    'SPEC' if samples_df.GenRe_SPEC_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'GRC1' if samples_df.GenRe_GRC1_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'GRC2' if samples_df.GenRe_GRC2_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'Pf8' if samples_df.Pf8_ENA_FASTQ_FTP_1.str.contains(row.file).any() else None, axis=1)\n",
    "counts_df = counts_df[ counts_df['panel'].isin(['SPEC','GRC1','GRC2'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "731cdf31-bd8a-421c-8aee-f04f4eeb4ba1",
   "metadata": {},
   "source": [
    "Sanity check: we should have 907 samples for each GenRe primer panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "51e781cd-c5c9-4637-a499-36e5651ba811",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>panel</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GRC1</td>\n",
       "      <td>907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GRC2</td>\n",
       "      <td>907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPEC</td>\n",
       "      <td>907</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  panel  count\n",
       "0  GRC1    907\n",
       "1  GRC2    907\n",
       "2  SPEC    907"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.groupby('panel').size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea3afc4d-9f6e-497c-ad9c-abfe7bf14b50",
   "metadata": {},
   "source": [
    "Calculate stats for read counts per primer panel: minimum, maximum, mean and percentiles at 25%, 50% and 75%.  \n",
    "These will be used to guide the number of reads to simulate for each panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f0d717a1-8d71-437f-b92a-49f61f00f6cd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"6\" halign=\"left\">count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>q25</th>\n",
       "      <th>q50</th>\n",
       "      <th>q75</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>panel</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GRC1</th>\n",
       "      <td>33917.249173</td>\n",
       "      <td>18</td>\n",
       "      <td>96442</td>\n",
       "      <td>25568.5</td>\n",
       "      <td>32424.0</td>\n",
       "      <td>40281.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRC2</th>\n",
       "      <td>33500.464168</td>\n",
       "      <td>21</td>\n",
       "      <td>78617</td>\n",
       "      <td>26732.5</td>\n",
       "      <td>32519.0</td>\n",
       "      <td>39864.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPEC</th>\n",
       "      <td>1291.057332</td>\n",
       "      <td>0</td>\n",
       "      <td>17924</td>\n",
       "      <td>654.5</td>\n",
       "      <td>988.0</td>\n",
       "      <td>1445.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              count                                      \n",
       "               mean min    max      q25      q50      q75\n",
       "panel                                                    \n",
       "GRC1   33917.249173  18  96442  25568.5  32424.0  40281.5\n",
       "GRC2   33500.464168  21  78617  26732.5  32519.0  39864.5\n",
       "SPEC    1291.057332   0  17924    654.5    988.0   1445.5"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def q25(x):\n",
    "    return x.quantile(0.25)\n",
    "def q50(x):\n",
    "    return x.quantile(0.5)\n",
    "def q75(x):\n",
    "    return x.quantile(0.75)\n",
    "\n",
    "counts_df.groupby('panel').agg({'count': ['mean', 'min','max', q25,q50,q75]})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "300d0a58-b5b1-44f9-8320-fd97d78a8c9f",
   "metadata": {},
   "source": [
    "### Conclusion\n",
    "The read count stats shown above can be used as guidelines for simulation runs. Specifically, the 25, 50 and 75 quantile are useful guides for three datasets to be created.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0f5b215-754a-490a-89cf-53a47de1d642",
   "metadata": {},
   "source": [
    "## Part 2: Configuring the read simulation pipeline\n",
    "The [read simulation pipeline](https://github.com/ISOinaBox/iib_readsim) requires a set of input files to configure the simulation run. This part is about creating these files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2af2109-7db4-4074-8267-88e08e244347",
   "metadata": {},
   "source": [
    "### Create a bed-like file of primer positions\n",
    "A file of primer positions in a bed-like format is rquired to configure start/end positions of amplicons to simulate.  \n",
    "As we are simulating a SpotMalaria AmpSeq run, we are using primer positions for the GenRe-04b-SpotMalaria AmpSeq panel, available from www.malariagen.net in form of an Excel file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "19cf1db4-fe4b-44a4-99c1-4400d2d47f2a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df = pd.read_excel(\n",
    "    \"https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx\",\n",
    "    sheet_name=\"P. falciparum amplicon primers\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09944b0f-fc25-4ebc-a24a-4da09dc2c206",
   "metadata": {},
   "source": [
    "Change column names in line with expected headers for bed-like config file for the simulation pipeline.  \n",
    "Add a column DESC, expected by the simulation pipeline. Use it to record the primer panel ('GRC1','GRC2' or 'SPEC') and drop rows (if any) that do not have a primer panel assigned.  \n",
    "Only keep the columns relevant to the simulation pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "19353f64-97d0-480a-8ebb-9f1327488aa2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df.rename(\n",
    "    columns={\n",
    "        'Chromosome':'#CHROM', \n",
    "        'Amplicon Start Position':'START',\n",
    "        'Amplicon Stop Position':'END'\n",
    "    },inplace=True)\n",
    "\n",
    "panel_df['DESC'] = panel_df.apply(\n",
    "    lambda row:\n",
    "    'SPEC' if 'SPEC' in row['Multiplex'] else \n",
    "    'GRC1' if 'GRC1' in row['Multiplex'] else \n",
    "    'GRC2' if 'GRC2' in row['Multiplex'] else\n",
    "    np.NaN, \n",
    "    axis=1\n",
    ")\n",
    "panel_df=panel_df[panel_df['DESC'].notna()]\n",
    "keep_cols=['#CHROM','START','END','DESC']\n",
    "panel_df=panel_df[keep_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b5e61af-2b1c-41cb-a060-4d6cfe38bc03",
   "metadata": {},
   "source": [
    "Coordinates are 1-based in this file but need to be converted to 0-based for the simulation pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7d27fe78-082c-4f89-928a-358206b71e51",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df['START']=pd.to_numeric(panel_df['START'],errors='coerce')-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4b784ca2-82b6-4f23-9a42-8abb263b63bc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df['END']=pd.to_numeric(panel_df['END'],errors='coerce')-1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27b8dae3-645a-4da4-b342-9ddecfa152aa",
   "metadata": {},
   "source": [
    "Keep only rows where primer start and end is provided"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "76312dd6-8979-42e4-af23-78ea17ea3aaf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df = panel_df[(panel_df['START'].notna()) & (panel_df['END'].notna())]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09b344a8-0302-4800-a57c-3c542df14f8d",
   "metadata": {},
   "source": [
    "convert positions to integer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eb2d703e-f356-4e53-bf86-207b09c497de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df['START']=panel_df['START'].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7d44be3a-997c-4966-946f-70cccb6b4ac4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df['END']=panel_df['END'].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "41dff90d-2c24-4064-a629-b13eb30a037c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>DESC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pf_M76611</td>\n",
       "      <td>88</td>\n",
       "      <td>314</td>\n",
       "      <td>SPEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pf_M76611</td>\n",
       "      <td>1637</td>\n",
       "      <td>1892</td>\n",
       "      <td>SPEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pf3D7_01_v3</td>\n",
       "      <td>145386</td>\n",
       "      <td>145590</td>\n",
       "      <td>GRC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pf3D7_01_v3</td>\n",
       "      <td>179274</td>\n",
       "      <td>179471</td>\n",
       "      <td>GRC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pf3D7_01_v3</td>\n",
       "      <td>180439</td>\n",
       "      <td>180683</td>\n",
       "      <td>GRC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>Pf3D7_14_v3</td>\n",
       "      <td>2480904</td>\n",
       "      <td>2481116</td>\n",
       "      <td>GRC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>Pf3D7_14_v3</td>\n",
       "      <td>2625793</td>\n",
       "      <td>2626011</td>\n",
       "      <td>GRC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Pf3D7_14_v3</td>\n",
       "      <td>2733564</td>\n",
       "      <td>2733803</td>\n",
       "      <td>GRC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Pf3D7_14_v3</td>\n",
       "      <td>3045934</td>\n",
       "      <td>3046141</td>\n",
       "      <td>GRC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>Pf3D7_14_v3</td>\n",
       "      <td>3126025</td>\n",
       "      <td>3126256</td>\n",
       "      <td>GRC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          #CHROM    START      END  DESC\n",
       "0      Pf_M76611       88      314  SPEC\n",
       "1      Pf_M76611     1637     1892  SPEC\n",
       "2    Pf3D7_01_v3   145386   145590  GRC1\n",
       "3    Pf3D7_01_v3   179274   179471  GRC2\n",
       "4    Pf3D7_01_v3   180439   180683  GRC1\n",
       "..           ...      ...      ...   ...\n",
       "132  Pf3D7_14_v3  2480904  2481116  GRC1\n",
       "133  Pf3D7_14_v3  2625793  2626011  GRC2\n",
       "134  Pf3D7_14_v3  2733564  2733803  GRC1\n",
       "135  Pf3D7_14_v3  3045934  3046141  GRC2\n",
       "136  Pf3D7_14_v3  3126025  3126256  GRC1\n",
       "\n",
       "[136 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "panel_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f389c749-4dd8-4c42-a7dd-b260156c6d13",
   "metadata": {},
   "source": [
    "Create versions of the panel file for each of the three sub-panels and save files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6955b22d-26d9-406d-a2bb-6af200127279",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "spec_panel_df = panel_df[panel_df['DESC']=='SPEC']\n",
    "grc1_panel_df = panel_df[panel_df['DESC']=='GRC1']\n",
    "grc2_panel_df = panel_df[panel_df['DESC']=='GRC2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "692521b7-b326-4c09-abd4-d93d1e8b726a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "spec_panel_df.to_csv('SpotMalaria-SPEC_Pf_amplicon_primers.bed', index=False)\n",
    "grc1_panel_df.to_csv('SpotMalaria-GRC1_Pf_amplicon_primers.bed', index=False)\n",
    "grc2_panel_df.to_csv('SpotMalaria-GRC2_Pf_amplicon_primers.bed', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8c339af-40a3-4e4a-90cd-ad41c956541e",
   "metadata": {},
   "source": [
    "## Reference genome and annotation\n",
    "The read simulation pipeline requires a reference genome. The _P. falciparum_ 3D7 reference genome was obtained from \n",
    "https://plasmodb.org/common/downloads/Current_Release/Pfalciparum3D7/fasta/data/PlasmoDB-68_Pfalciparum3D7_Genome.fasta\n",
    "and is provided in this folder.  \n",
    "In addition, a GFF genome annotation file is required for the next step, creating a variant file. It was obtained from \n",
    "https://plasmodb.org/a/service/raw-files/release-68/Pfalciparum3D7/gff/data/PlasmoDB-68_Pfalciparum3D7.gff and is also provided in this folder.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b493f0-22b4-4ad8-94e6-818650906621",
   "metadata": {},
   "source": [
    "## Variant file\n",
    "This section creates an example variants file for the read simulation pipeline. It can be used to simulate a run of the SpotMalaria panel with a genome that carries a resistance genotype. The simulation pipeline requires a vcf-like file with SNP data in order to create a mutant genome from the reference WT.\n",
    "\n",
    "As pointed out in the introduction, this is intended as a template for creating tailor-made simulated read datasets. In this example, reads will be created to simulate a chloroquine resistance encoded in the crt gene (PF3D7_0709000). This could be used to verify that a pipeline correctly calls this mutation and the resistance phenotype from the simulated read data.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70fc3489-7395-480f-8570-788ec787792e",
   "metadata": {},
   "source": [
    "Drug-resistance genotypes are recorded in the Pf8_resistance_classification.pdf file available [here](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf).  \n",
    "\n",
    "According to page 3 of this document, the mutation PF3D7_0709000:K76T leads to chloroquine resistance.  \n",
    "\n",
    "To create the variants (vcf-like) file for the simualtion pipeline, we need to convert this aminoacid position into a genome position and apply a SNP at DNA level.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a92d3eef-813f-4693-8dd9-67106b060741",
   "metadata": {},
   "source": [
    "A python module is provided in this folder, which can be used stand-alone but will be used via imported functions here in order to convert from gene/AA to genomic NA coordinates via a GFF annotation file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cb55aaba-c2d5-4a26-a75f-af4620127c46",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from snp2mutant_coords import find_cds_for_gene, aa_to_genomic_from_exons, sequence_for_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fb9f56f1-511a-4001-a893-c7553c26489a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PF3D7_0709000 AA pos 76 = genomic position (1-based) Pf3D7_07_v3:403624-403626\n"
     ]
    }
   ],
   "source": [
    "gene_id='PF3D7_0709000'\n",
    "aa_pos=76\n",
    "gff_file='PlasmoDB-68_Pfalciparum3D7.gff'\n",
    "\n",
    "chrom, strand, exons, diagnostics = find_cds_for_gene(gff_file, gene_id)\n",
    "start, end = aa_to_genomic_from_exons(exons, aa_pos, strand)\n",
    "\n",
    "print(f'{gene_id} AA pos {str(aa_pos)} = genomic position (1-based) {chrom}:{str(start)}-{str(end)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "322d20a0-a5b6-46f4-b1b8-e96227242bf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence at Pf3D7_07_v3:403624-403626:AAA\n"
     ]
    }
   ],
   "source": [
    "seq=sequence_for_pos(\n",
    "    'PlasmoDB-68_Pfalciparum3D7_Genome.fasta',\n",
    "    chrom, \n",
    "    strand,\n",
    "    start,\n",
    "    end)\n",
    "print(f'sequence at {chrom}:{str(start)}-{str(end)}:{seq}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fc66e66-f616-4aa4-a0d9-afa23c3d7df6",
   "metadata": {},
   "source": [
    "__RESULT:__ \n",
    "To create the chloroquine resistance mutaion PF3D7_0709000:K76T, codon 'AAA' at genomic position f3D7_07_v3:403624-403626 (encoding Lys/K) needs to be changed to 'ACA' to encode Thr/T. For reference see https://en.wikipedia.org/wiki/DNA_and_RNA_codon_tables.  \n",
    "\n",
    "The vcf-like file for this mutation looks like this:\n",
    "```\n",
    "#CHROM\tPOS\tREF\tALT\n",
    "Pf3D7_07_v3\t403625\tA\tC\n",
    "```\n",
    "This has been saved in this folder as file ```Pf3D7_07_v3_403625_a2c```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
