# A resource for malaria gold-standard validation data

## Aim of this resource
The aim of this resource is to provide datasets that may be used in validating pipelines for malaria genomics pipelines in order to provide evidence for accreditation of the pipeline and process.  
Two datasets are provided:  
- high-quality real-world data from public resources
- simulated data: example data and recipes

While real-world data from public resources are an important part of any validation strategy, such data suffer from two main issues:  
- data is generated using specific techniques that may not be be compatible with a given pipeline
- results are generated by other pipelines. Even if those pipelines are published and highly trusted, 


## Details about the data sets
This section provides details about the provided datasets, including recipes for creating the data, which may serve as a useful starting point for creating tailored datasets for different purposes.  

### Real-world data
The data in this section was obtained from two high-profile public malaria genomics resources:  
1. [MalariaGEN Pf8](https://www.malariagen.net/data_package/open-dataset-plasmodium-falciparum-v80/)
2. [GenRe Mekong](https://www.malariagen.net/resource/29/)

Analysis pipelines differ in scope and methodology. Thus, no single dataset will be applicable to every possible genomics pipeline. For the selection of real-world public datasets, we are assuming that the pipeline to be validated produces:  
- amino acid calls for variant loci that are relevant to drug resistance
and/or  
- high-level drug-resistance phenotype calls 
Furthermore, we assume that the pipeline can work with _P. falciparum_ data.

For pipelines that are using amplicon sequencing data, we assume that the pipeline can work with the SpotMalaria panel. For details about this panel, consult the [SpotMalaria technical manual](https://ngs.sanger.ac.uk/production/malaria/Resource/29/20200705-GenRe-04a-SpotMalaria-0.39.pdf) and [this SpotMalaria supplementary data file](https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx), which provides details for every primer used in the panel.





## Rationale for selection of resources
Pipelines are usually tailored to a particular sequences technology, such as whole-genome or amplicon sequencing. While Pf8 is based on (selective) Whole Genome Sequencing (sWGS) data, GenRE Mekong utilised the SPotMalaria panel for amplicon sequencing (download available at [https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx](https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx).
Some GenRE Mekong samples have been re-analysed with sWGS for Pf8, thus providing data for the same sample from two different technologies and two independent resistance phenotype calls.
We will use the intersect between Pf8 and GenRE Mekong as the gold-standard dataset.

In Pf8, drug-resistance phenotypes are inferred based on criteria described in [this PDF](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf)


#### Rationale for selecting the dataset
#### Recipe for creating the dataset

### Simulated data