{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e0088e6-1fc9-4e89-bfdc-3490c7385928",
   "metadata": {},
   "source": [
    "# Preparation for creating simulated data\n",
    "## Aim of this notebook\n",
    "This notebook contains the preparation required to run amplicon-sequencing (AmpSeq) read simulations using the read simulation pipeline \"pop_var_sim\", provided in a [separate repository](https://github.com/ISOinaBox/pop_var_sim).  \n",
    "\n",
    "While the files produced in this notebook can be used to create simulated reads, the main use of this notebook is to serve as a template and guide for generating simulation runs, tailor-made for a pipeline that is to be validated.  \n",
    "\n",
    "This notebook focuses on AmpSeq data because simulated data is particularly useful for this type of data. Whole-genome data for pipeline validataion can be obtained from real-world datasets that are readily available [from public resources](https://www.malariagen.net/resource/36/). AmpSeq data, on the other hand, depends on a specific primer panel, for which data may not be readily available. In addition, a pipeline written and validated for a specific AmpSeq panel, such as [SpotMalaria](https://ngs.sanger.ac.uk/production/malaria/Resource/29/20200705-GenRe-04a-SpotMalaria-0.39.pdf) may need to be validated against other primer panels, for which data does not exist yet.  \n",
    "\n",
    "In part 1 of this notebook, we obtain read counts from high-quality public data. This step is necessary for any simulation, so that the generated data provides realistic input for an analysis pipeline. Here, we simulate a [SpotMalaria](https://ngs.sanger.ac.uk/production/malaria/Resource/29/20200705-GenRe-04a-SpotMalaria-0.39.pdf) AmpSeq run, for which real-world samples exist in public repositories. This may not always be the case for simulating different AmpSeq primer panel runs, in which case the user may need to estimate realistic read counts in different ways.\n",
    "\n",
    "In part 2, the simulation run input files are created. This demonstrates how genome data and publicly available data on relevant mutations can be used to simulate WT and mutant strain results and should be modified according to the validation that is to be carried out. For example, a new pipeline might need to be tested in terms of how robustly it can call a mutant allele in a mix of WT and mutant parasites, which can be configured in the simulation pipeline run.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97c3368e-a9ff-4f8c-b4dc-45c7fc23a23c",
   "metadata": {},
   "source": [
    "## Part 1: read counts to guide simulation\n",
    "An important consideration for a simulated dataset is the number of reads to create. This should be within the range of expected read numbers from real-world experiments because unrealistically high or low read counts will not provide useful insights into the validity of pipeline results.  \n",
    "\n",
    "In this example, we are simulating reads for a wild-type and a mutant _P. falciparum_ strain. We will simulate an AmpSeq run using the [SpotMalaria](https://www.malariagen.net/project/spotmalaria/) panel, as used for the [GenRe Mekong project](https://www.malariagen.net/resource/29/) project. \n",
    "\n",
    "This primer panel consists of three sub-panels: SPEC (speciation), GRC1 and GRC2.  \n",
    "The data for these panels is provided in [this spreadsheet](https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xls).  \n",
    "\n",
    "As noted in the introduction, this analysis is specific to the SpotMalaria AmpSeq primer panel and will need to be modified - based on the template provided here - to accommodate other AmpSeq panels or whole-genome data.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfee8c68-86d3-4554-adc9-05716ca8d18c",
   "metadata": {},
   "source": [
    "### FASTQ data\n",
    "The FASTQ files used in this notebook are based on the ENA data download manifests created in [this notebook](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb). But __note__ that the FASTQ files are __not__ icluded in this repo because of the large amount of data contained in them. If you would like to re-run this notebook, please refer to the explanation provided [here](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb#Example:-how-to-use-the-downloader-tool) to download the FASTQ files from ENA using the downloader tool provided in this repo.\n",
    "\n",
    "For the purpose of this analysis, the dataset [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv), created [here](../real-world_gold-standard_data/Pf8-GenReMekong/pf8genre.ipynb#Save-results:-concordant-phenotype-data) was used. This contains a subset of 907 samples of all high-quality data from the GenRe Mekong project where drug resistance phenotype agrees with the WGS sample analysis in Pf8.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9033074e-7289-4f7e-856d-696489e35e7b",
   "metadata": {},
   "source": [
    "### Creating the read counts data file\n",
    "A file \"fastq_readcounts.csv\" is already prepared in this directory and used by this notebook. \n",
    "\n",
    "It is recommended to use that file as provided in this folder. The following instructions are only needed to re-create that file and to document how it was generated. \n",
    "\n",
    "Ensure that the following Python packages are installed on your system:\n",
    "- requests\n",
    "- pandas\n",
    "\n",
    "Then, in a terminal, run the following commands. The commands can be found alongside this notebook.   \n",
    "Both commands will take a potentially long time to run (depending on your system and network) and the first command will download a large amount of data over the internet. If you are working in a HPC/cloud environment, you may want to split the data into chunks and parallelise these tasks. This is beyond the scope of this recipe.  \n",
    "\n",
    "__Step 1: Download FASTQ files__    \n",
    "Run the following command in a terminal (relative paths from this folder):\n",
    "```bash\n",
    "python ../../../real-world_gold-standard_data/Pf8-GenReMekong/scripts/ENA_data_helper.py  \\\n",
    "    download \\\n",
    "    --data ../../../real-world_gold-standard_data/Pf8-GenReMekong/additional_output_files/ENA_download_manifest_concordant_phenotype_high_quality_samples.csv  \\\n",
    "    --skip_errors \\\n",
    "    --download_attempts 10 \\\n",
    "    --out ENA_download_concordant_phenotype_high_quality_sample \\\n",
    "```\n",
    "It will create a folder \"ENA_download_concordant_phenotype_high_quality_sample\" in this directory. The folder contains the 3628 FASTQ (gzipped) files belonging to the [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv), created [here](../../../real-world_gold-standard_data/Pf8-GenReMekong/scripts/pf8genre.ipynb#Save-results:-concordant-phenotype-data).  \n",
    "\n",
    "__Step 2: Obtain read counts__  \n",
    "Once the FASTQ download has completed, use the script \"run_counts.sh\", also in this folder, like this:  \n",
    "(or change the path to the downloaded FASTQ files if different)\n",
    "\n",
    "```bash\n",
    "./run_counts.sh ENA_download_concordant_phenotype_high_quality_sample\n",
    "```  \n",
    "This will create a file \"fastq_readcounts.csv\" with file names and read counts for all mate1 files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4671e11d-e3c2-4644-aab0-0220ac9035ad",
   "metadata": {},
   "source": [
    "### Calculate AmpSeq panel stats\n",
    "Using the dataset [Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv](../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv) and the pre-computed (see above) csv file of read counts, create read count stats for the three SpotMalaria amplicon panels.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5619e6ee-c39b-4cdf-b0cf-b1cfc32b0880",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8dc8d3bc-d7ea-4954-a9f2-09f92f564cbb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR14388603_1.fastq.gz</td>\n",
       "      <td>32819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR14388604_1.fastq.gz</td>\n",
       "      <td>29409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR14388605_1.fastq.gz</td>\n",
       "      <td>1094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR14388606_1.fastq.gz</td>\n",
       "      <td>38189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR14388607_1.fastq.gz</td>\n",
       "      <td>35804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3623</th>\n",
       "      <td>ERR15628752_1.fastq.gz</td>\n",
       "      <td>18788821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3624</th>\n",
       "      <td>ERR15628753_1.fastq.gz</td>\n",
       "      <td>13952241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3625</th>\n",
       "      <td>ERR15628764_1.fastq.gz</td>\n",
       "      <td>14159323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3626</th>\n",
       "      <td>ERR15628798_1.fastq.gz</td>\n",
       "      <td>7910579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3627</th>\n",
       "      <td>ERR15628802_1.fastq.gz</td>\n",
       "      <td>9683996</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3628 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        file     count\n",
       "0     ERR14388603_1.fastq.gz     32819\n",
       "1     ERR14388604_1.fastq.gz     29409\n",
       "2     ERR14388605_1.fastq.gz      1094\n",
       "3     ERR14388606_1.fastq.gz     38189\n",
       "4     ERR14388607_1.fastq.gz     35804\n",
       "...                      ...       ...\n",
       "3623  ERR15628752_1.fastq.gz  18788821\n",
       "3624  ERR15628753_1.fastq.gz  13952241\n",
       "3625  ERR15628764_1.fastq.gz  14159323\n",
       "3626  ERR15628798_1.fastq.gz   7910579\n",
       "3627  ERR15628802_1.fastq.gz   9683996\n",
       "\n",
       "[3628 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df = pd.read_csv('fastq_readcounts.csv')\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d6160fab-2fc3-4700-bda3-8bac7c28f01b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>Artemisinin</th>\n",
       "      <th>Piperaquine</th>\n",
       "      <th>Mefloquine</th>\n",
       "      <th>Chloroquine</th>\n",
       "      <th>Pyrimethamine</th>\n",
       "      <th>Sulfadoxine</th>\n",
       "      <th>DHA-PPQ</th>\n",
       "      <th>AS-MQ</th>\n",
       "      <th>Species</th>\n",
       "      <th>ENA</th>\n",
       "      <th>% callable</th>\n",
       "      <th>ENA_acc_Pf8</th>\n",
       "      <th>ENA_acc_GenRe_GRC1</th>\n",
       "      <th>ENA_acc_GenRe_GRC2</th>\n",
       "      <th>ENA_acc_GenRe_SPEC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RCN12025</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409305</td>\n",
       "      <td>89.14</td>\n",
       "      <td>ERR15625306</td>\n",
       "      <td>ERR14388603</td>\n",
       "      <td>ERR14388604</td>\n",
       "      <td>ERR14388605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RCN12026</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409312</td>\n",
       "      <td>85.83</td>\n",
       "      <td>ERR15625307</td>\n",
       "      <td>ERR14388606</td>\n",
       "      <td>ERR14388607</td>\n",
       "      <td>ERR14388608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RCN12028</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409373</td>\n",
       "      <td>87.23</td>\n",
       "      <td>ERR15625309</td>\n",
       "      <td>ERR14388612</td>\n",
       "      <td>ERR14388613</td>\n",
       "      <td>ERR14388614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RCN12031</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409314</td>\n",
       "      <td>87.04</td>\n",
       "      <td>ERR15625310</td>\n",
       "      <td>ERR14388621</td>\n",
       "      <td>ERR14388622</td>\n",
       "      <td>ERR14388623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RCN12032</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR3409311</td>\n",
       "      <td>88.65</td>\n",
       "      <td>ERR15625311</td>\n",
       "      <td>ERR14388624</td>\n",
       "      <td>ERR14388625</td>\n",
       "      <td>ERR14388626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>902</th>\n",
       "      <td>RCN26767</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221085</td>\n",
       "      <td>87.92</td>\n",
       "      <td>ERR15628752</td>\n",
       "      <td>ERR14397546</td>\n",
       "      <td>ERR14397547</td>\n",
       "      <td>ERR14397548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>RCN26775</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221086</td>\n",
       "      <td>87.30</td>\n",
       "      <td>ERR15628753</td>\n",
       "      <td>ERR14397570</td>\n",
       "      <td>ERR14397571</td>\n",
       "      <td>ERR14397572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>RCN26820</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221102</td>\n",
       "      <td>89.35</td>\n",
       "      <td>ERR15628764</td>\n",
       "      <td>ERR14397663</td>\n",
       "      <td>ERR14397664</td>\n",
       "      <td>ERR14397665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>RCN26932</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221141</td>\n",
       "      <td>86.03</td>\n",
       "      <td>ERR15628798</td>\n",
       "      <td>ERR14397975</td>\n",
       "      <td>ERR14397976</td>\n",
       "      <td>ERR14397977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906</th>\n",
       "      <td>RCN26938</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Resistant</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Sensitive</td>\n",
       "      <td>Pf</td>\n",
       "      <td>ERR7221132</td>\n",
       "      <td>89.21</td>\n",
       "      <td>ERR15628802</td>\n",
       "      <td>ERR14397990</td>\n",
       "      <td>ERR14397991</td>\n",
       "      <td>ERR14397992</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>907 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample Artemisinin Piperaquine Mefloquine Chloroquine Pyrimethamine  \\\n",
       "0    RCN12025   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "1    RCN12026   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "2    RCN12028   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "3    RCN12031   Resistant   Resistant  Sensitive   Resistant     Resistant   \n",
       "4    RCN12032   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "..        ...         ...         ...        ...         ...           ...   \n",
       "902  RCN26767   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "903  RCN26775   Sensitive   Sensitive  Sensitive   Resistant     Resistant   \n",
       "904  RCN26820   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "905  RCN26932   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "906  RCN26938   Resistant   Sensitive  Sensitive   Resistant     Resistant   \n",
       "\n",
       "    Sulfadoxine    DHA-PPQ      AS-MQ Species         ENA  % callable  \\\n",
       "0     Resistant  Sensitive  Sensitive      Pf  ERR3409305       89.14   \n",
       "1     Resistant  Resistant  Sensitive      Pf  ERR3409312       85.83   \n",
       "2     Resistant  Resistant  Sensitive      Pf  ERR3409373       87.23   \n",
       "3     Resistant  Resistant  Sensitive      Pf  ERR3409314       87.04   \n",
       "4     Resistant  Sensitive  Sensitive      Pf  ERR3409311       88.65   \n",
       "..          ...        ...        ...     ...         ...         ...   \n",
       "902   Resistant  Sensitive  Sensitive      Pf  ERR7221085       87.92   \n",
       "903   Sensitive  Sensitive  Sensitive      Pf  ERR7221086       87.30   \n",
       "904   Resistant  Sensitive  Sensitive      Pf  ERR7221102       89.35   \n",
       "905   Resistant  Sensitive  Sensitive      Pf  ERR7221141       86.03   \n",
       "906   Resistant  Sensitive  Sensitive      Pf  ERR7221132       89.21   \n",
       "\n",
       "     ENA_acc_Pf8 ENA_acc_GenRe_GRC1 ENA_acc_GenRe_GRC2 ENA_acc_GenRe_SPEC  \n",
       "0    ERR15625306        ERR14388603        ERR14388604        ERR14388605  \n",
       "1    ERR15625307        ERR14388606        ERR14388607        ERR14388608  \n",
       "2    ERR15625309        ERR14388612        ERR14388613        ERR14388614  \n",
       "3    ERR15625310        ERR14388621        ERR14388622        ERR14388623  \n",
       "4    ERR15625311        ERR14388624        ERR14388625        ERR14388626  \n",
       "..           ...                ...                ...                ...  \n",
       "902  ERR15628752        ERR14397546        ERR14397547        ERR14397548  \n",
       "903  ERR15628753        ERR14397570        ERR14397571        ERR14397572  \n",
       "904  ERR15628764        ERR14397663        ERR14397664        ERR14397665  \n",
       "905  ERR15628798        ERR14397975        ERR14397976        ERR14397977  \n",
       "906  ERR15628802        ERR14397990        ERR14397991        ERR14397992  \n",
       "\n",
       "[907 rows x 16 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_df = pd.read_csv('../real-world_gold-standard_data/Pf8-GenReMekong/Pf8-GenReMekong_concordant_phenotype_high_quality_samples.csv')\n",
    "samples_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c8d0e88-7d68-4a23-bd31-92ed22f872b7",
   "metadata": {},
   "source": [
    "Read counts in ```counts_df``` are assigned to file names that do not contain the name of the primer panel to which the file belongs. Use the ```samples_df``` data to assign the panel name (SPEC, GRC1 or GRC2) to each read count by looking up the file name.  \n",
    "Drop Pf8 (WGS) data, which is not relevant here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c876cc3-0795-40e6-8d9e-f11372cab460",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'GenRe_SPEC_ENA_FASTQ_FTP_1'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_3960673/3769241651.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m counts_df['panel']=counts_df.apply(\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;34m'SPEC'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0msamples_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGenRe_SPEC_ENA_FASTQ_FTP_1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;34m'GRC1'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0msamples_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGenRe_GRC1_ENA_FASTQ_FTP_1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontains\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, func, axis, raw, result_type, args, by_row, engine, engine_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m  10397\u001b[0m             \u001b[0mengine_kwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mengine_kwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10398\u001b[0m             \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10399\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m  10400\u001b[0m         )\n\u001b[0;32m> 10401\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__finalize__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"apply\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    912\u001b[0m         \u001b[0;31m# raw\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    913\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraw\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    914\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine_kwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    915\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 916\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1061\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1062\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1063\u001b[0;31m             \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1064\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1065\u001b[0m             \u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_series_numba\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1066\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1077\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1078\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0moption_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"mode.chained_assignment\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1079\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseries_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1080\u001b[0m                 \u001b[0;31m# ignore SettingWithCopy here in case the user mutates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1081\u001b[0;31m                 \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1082\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1083\u001b[0m                     \u001b[0;31m# If we have a view on v, we need to make a copy because\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1084\u001b[0m                     \u001b[0;31m#  series_generator will swap out the underlying data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_3960673/3769241651.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.11/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6317\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6318\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6319\u001b[0m         ):\n\u001b[1;32m   6320\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6321\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'GenRe_SPEC_ENA_FASTQ_FTP_1'"
     ]
    }
   ],
   "source": [
    "counts_df['panel']=counts_df.apply(\n",
    "    lambda row: \n",
    "    'SPEC' if samples_df.GenRe_SPEC_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'GRC1' if samples_df.GenRe_GRC1_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'GRC2' if samples_df.GenRe_GRC2_ENA_FASTQ_FTP_1.str.contains(row.file).any() else \n",
    "    'Pf8' if samples_df.Pf8_ENA_FASTQ_FTP_1.str.contains(row.file).any() else None, axis=1)\n",
    "counts_df = counts_df[ counts_df['panel'].isin(['SPEC','GRC1','GRC2'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "731cdf31-bd8a-421c-8aee-f04f4eeb4ba1",
   "metadata": {},
   "source": [
    "Sanity check: we should have 907 samples for each GenRe primer panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51e781cd-c5c9-4637-a499-36e5651ba811",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "counts_df.groupby('panel').size().to_frame('count').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea3afc4d-9f6e-497c-ad9c-abfe7bf14b50",
   "metadata": {},
   "source": [
    "Calculate stats for read counts per primer panel: minimum, maximum, mean and percentiles at 25%, 50% and 75%.  \n",
    "These will be used to guide the number of reads to simulate for each panel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d717a1-8d71-437f-b92a-49f61f00f6cd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def q25(x):\n",
    "    return x.quantile(0.25)\n",
    "def q50(x):\n",
    "    return x.quantile(0.5)\n",
    "def q75(x):\n",
    "    return x.quantile(0.75)\n",
    "\n",
    "counts_df.groupby('panel').agg({'count': ['mean', 'min','max', q25,q50,q75]})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "300d0a58-b5b1-44f9-8320-fd97d78a8c9f",
   "metadata": {},
   "source": [
    "### Conclusion\n",
    "The read count stats shown above can be used as guidelines for simulation runs. Specifically, the 25, 50 and 75 quantile are useful guides for three datasets to be created.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0f5b215-754a-490a-89cf-53a47de1d642",
   "metadata": {},
   "source": [
    "## Part 2: Configuring the read simulation pipeline\n",
    "The [read simulation pipeline](https://github.com/ISOinaBox/pop_var_sim) requires a set of input files to configure the simulation run. This part is about creating these files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2af2109-7db4-4074-8267-88e08e244347",
   "metadata": {},
   "source": [
    "### Create a bed-like file of primer positions\n",
    "A file of primer positions in a bed-like format is rquired to configure start/end positions of amplicons to simulate.  \n",
    "As we are simulating a SpotMalaria AmpSeq run, we are using primer positions for the GenRe-04b-SpotMalaria AmpSeq panel, available from www.malariagen.net in form of an Excel file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19cf1db4-fe4b-44a4-99c1-4400d2d47f2a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df = pd.read_excel(\n",
    "    \"https://www.malariagen.net/wp-content/uploads/2023/11/20200705-GenRe-04b-SpotMalaria-SupplementaryFile1.xlsx\",\n",
    "    sheet_name=\"P. falciparum amplicon primers\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09944b0f-fc25-4ebc-a24a-4da09dc2c206",
   "metadata": {},
   "source": [
    "Change column names in line with expected headers for bed-like config file for the simulation pipeline.  \n",
    "Add a column DESC, expected by the simulation pipeline. Use it to record the primer panel ('GRC1','GRC2' or 'SPEC') and drop rows (if any) that do not have a primer panel assigned.  \n",
    "Only keep the columns relevant to the simulation pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19353f64-97d0-480a-8ebb-9f1327488aa2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df.rename(\n",
    "    columns={\n",
    "        'Chromosome':'#CHROM', \n",
    "        'Amplicon Start Position':'START',\n",
    "        'Amplicon Stop Position':'END'\n",
    "    },inplace=True)\n",
    "\n",
    "panel_df['DESC'] = panel_df.apply(\n",
    "    lambda row:\n",
    "    'SPEC' if 'SPEC' in row['Multiplex'] else \n",
    "    'GRC1' if 'GRC1' in row['Multiplex'] else \n",
    "    'GRC2' if 'GRC2' in row['Multiplex'] else\n",
    "    np.NaN, \n",
    "    axis=1\n",
    ")\n",
    "panel_df=panel_df[panel_df['DESC'].notna()]\n",
    "keep_cols=['#CHROM','START','END','DESC']\n",
    "panel_df=panel_df[keep_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b5e61af-2b1c-41cb-a060-4d6cfe38bc03",
   "metadata": {},
   "source": [
    "Coordinates are 1-based in this file but need to be converted to 0-based for the simulation pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d27fe78-082c-4f89-928a-358206b71e51",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df['START']=pd.to_numeric(panel_df['START'],errors='coerce')-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b784ca2-82b6-4f23-9a42-8abb263b63bc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df['END']=pd.to_numeric(panel_df['END'],errors='coerce')-1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27b8dae3-645a-4da4-b342-9ddecfa152aa",
   "metadata": {},
   "source": [
    "Keep only rows where primer start and end is provided"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76312dd6-8979-42e4-af23-78ea17ea3aaf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df = panel_df[(panel_df['START'].notna()) & (panel_df['END'].notna())]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09b344a8-0302-4800-a57c-3c542df14f8d",
   "metadata": {},
   "source": [
    "convert positions to integer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb2d703e-f356-4e53-bf86-207b09c497de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df['START']=panel_df['START'].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d44be3a-997c-4966-946f-70cccb6b4ac4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df['END']=panel_df['END'].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41dff90d-2c24-4064-a629-b13eb30a037c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "panel_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f389c749-4dd8-4c42-a7dd-b260156c6d13",
   "metadata": {},
   "source": [
    "Create versions of the panel file for each of the three sub-panels and save files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6955b22d-26d9-406d-a2bb-6af200127279",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "spec_panel_df = panel_df[panel_df['DESC']=='SPEC']\n",
    "grc1_panel_df = panel_df[panel_df['DESC']=='GRC1']\n",
    "grc2_panel_df = panel_df[panel_df['DESC']=='GRC2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "692521b7-b326-4c09-abd4-d93d1e8b726a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "spec_panel_df.to_csv('SpotMalaria-SPEC_Pf_amplicon_primers.bed', index=False, sep=\"\\t\")\n",
    "grc1_panel_df.to_csv('SpotMalaria-GRC1_Pf_amplicon_primers.bed', index=False, sep=\"\\t\")\n",
    "grc2_panel_df.to_csv('SpotMalaria-GRC2_Pf_amplicon_primers.bed', index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8c339af-40a3-4e4a-90cd-ad41c956541e",
   "metadata": {},
   "source": [
    "### Reference genome and annotation\n",
    "The read simulation pipeline requires a reference genome. The _P. falciparum_ 3D7 reference genome was obtained from \n",
    "https://plasmodb.org/common/downloads/Current_Release/Pfalciparum3D7/fasta/data/PlasmoDB-68_Pfalciparum3D7_Genome.fasta\n",
    "and is provided in this folder.  \n",
    "In addition, a GFF genome annotation file is required for the next step, creating a variant file. It was obtained from \n",
    "https://plasmodb.org/a/service/raw-files/release-68/Pfalciparum3D7/gff/data/PlasmoDB-68_Pfalciparum3D7.gff and is also provided in this folder.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b493f0-22b4-4ad8-94e6-818650906621",
   "metadata": {},
   "source": [
    "### Variant file\n",
    "This section creates an example variants file for the read simulation pipeline. It can be used to simulate a run of the SpotMalaria panel with a genome that carries a resistance genotype. The simulation pipeline requires a vcf-like file with SNP data in order to create a mutant genome from the reference WT.\n",
    "\n",
    "As pointed out in the introduction, this is intended as a template for creating tailor-made simulated read datasets. In this example, reads will be created to simulate a chloroquine resistance encoded in the crt gene (PF3D7_0709000). This could be used to verify that a pipeline correctly calls this mutation and the resistance phenotype from the simulated read data.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70fc3489-7395-480f-8570-788ec787792e",
   "metadata": {},
   "source": [
    "Drug-resistance genotypes are recorded in the Pf8_resistance_classification.pdf file available [here](https://pf8-release.cog.sanger.ac.uk/Pf8_resistance_classification.pdf).  \n",
    "\n",
    "According to page 3 of this document, the mutation PF3D7_0709000:K76T leads to chloroquine resistance.  \n",
    "\n",
    "To create the variants (vcf-like) file for the simualtion pipeline, we need to convert this aminoacid position into a genome position and apply a SNP at DNA level.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a92d3eef-813f-4693-8dd9-67106b060741",
   "metadata": {},
   "source": [
    "A python module is provided in this folder, which can be used stand-alone but will be used via imported functions here in order to convert from gene/AA to genomic NA coordinates via a GFF annotation file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb55aaba-c2d5-4a26-a75f-af4620127c46",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from snp2mutant_coords import find_cds_for_gene, aa_to_genomic_from_exons, sequence_for_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb9f56f1-511a-4001-a893-c7553c26489a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene_id='PF3D7_0709000'\n",
    "aa_pos=76\n",
    "gff_file='PlasmoDB-68_Pfalciparum3D7.gff'\n",
    "\n",
    "chrom, strand, exons, diagnostics = find_cds_for_gene(gff_file, gene_id)\n",
    "start, end = aa_to_genomic_from_exons(exons, aa_pos, strand)\n",
    "\n",
    "print(f'{gene_id} AA pos {str(aa_pos)} = genomic position (1-based) {chrom}:{str(start)}-{str(end)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "322d20a0-a5b6-46f4-b1b8-e96227242bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq=sequence_for_pos(\n",
    "    'PlasmoDB-68_Pfalciparum3D7_Genome.fasta',\n",
    "    chrom, \n",
    "    strand,\n",
    "    start,\n",
    "    end)\n",
    "print(f'sequence at {chrom}:{str(start)}-{str(end)}:{seq}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fc66e66-f616-4aa4-a0d9-afa23c3d7df6",
   "metadata": {},
   "source": [
    "__RESULT:__ \n",
    "To create the chloroquine resistance mutaion PF3D7_0709000:K76T, codon 'AAA' at genomic position f3D7_07_v3:403624-403626 (encoding Lys/K) needs to be changed to 'ACA' to encode Thr/T. For reference see https://en.wikipedia.org/wiki/DNA_and_RNA_codon_tables.  \n",
    "\n",
    "The vcf-like file for this mutation looks like this:\n",
    "```\n",
    "#CHROM\tPOS\tREF\tALT\n",
    "Pf3D7_07_v3\t403625\tA\tC\n",
    "```\n",
    "This has been saved in this folder as file ```Pf3D7_07_v3_403625_a2c```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67417a6a-fac2-415c-bdcb-ab84aef7fcbe",
   "metadata": {},
   "source": [
    "### Simulation pipeline final input files\n",
    "With the file created so far in this section, we are now able to create the final input files for a run of the [read simulation pipeline](https://github.com/ISOinaBox/pop_var_sim).  \n",
    "#### Haplotype manifest\n",
    "According to the [pipeline manual](https://github.com/ISOinaBox/iib_readsim/blob/main/README.md), a \"haplotype manifest\" csv file is required to configure the reference genome(s) on which to create simulated reads. As we have three primer panels and two genotypes (WT and crt mutant), we need one line per genotype and primer panel in the manifest as shown here and saved as file ```hap_manifest_example.csv``` in this folder:  \n",
    "```\n",
    "haplotype,base_fasta,variants_file,bed_file\n",
    "wt_spec,PlasmoDB-68_Pfalciparum3D7_Genome.fasta,null,SpotMalaria-SPEC_Pf_amplicon_primers.bed\n",
    "wt_grc1,PlasmoDB-68_Pfalciparum3D7_Genome.fasta,null,SpotMalaria-GRC1_Pf_amplicon_primers.bed\n",
    "wt_grc2,PlasmoDB-68_Pfalciparum3D7_Genome.fasta,null,SpotMalaria-GRC2_Pf_amplicon_primers.bed\n",
    "crtmut_spec,PlasmoDB-68_Pfalciparum3D7_Genome.fasta,Pf3D7_07_v3_403625_a2c,SpotMalaria-SPEC_Pf_amplicon_primers.bed\n",
    "crtmut_grc1,PlasmoDB-68_Pfalciparum3D7_Genome.fasta,Pf3D7_07_v3_403625_a2c,SpotMalaria-GRC1_Pf_amplicon_primers.bed\n",
    "crtmut_grc2,PlasmoDB-68_Pfalciparum3D7_Genome.fasta,Pf3D7_07_v3_403625_a2c,SpotMalaria-GRC2_Pf_amplicon_primers.bed\n",
    "```\n",
    "\n",
    "__Explanation of the file__:  \n",
    "Line 1 configures a WT genotype, using the 3D7 genome provided in this folder. The value in the 'variants_file' column is 'none' because we use the 3D7 genome as-is. The 'bed_file' is the file created earlier, providing positions of the primers in the SpotMalaria SPEC sub-panel.  Lines 2 and 3 do the same but refer to the GRC1 and GRC2 bed files, respectively.  \n",
    "Lines 4-6 configure the mutant genotype for the crt mutation as discussed in detail in the previous section. Here, we use the same 3D7 reference genome, but provide the vcf-like file for a single SNP that re-creates the resistant mutant, which was produced in the previous section. The bed (primer position) files are the same as for lines 1-3. \n",
    "\n",
    "#### Sample manifest\n",
    "The sample manifest file defines the samples that are to be simulated. Each sample can contain one or more haplotypes, as defined in the above haplotype manifest file. This would enable the simulation of mixed WT/mutant samples, which could be used to test the robustness of a pipeline for heterozygous results.  \n",
    "In the example provided here, we are not creating such mixed samples. Instead, we configure the pipeline to create one sample for each genotype (WT/mut) and for each primer sub-panel.  \n",
    "The read numbers are taken from the earlier analysis of real-world read counts in this notebook. For each genotype/primer-panel combination, we simulate one run each with the 25, 50 and 75 quantile of read counts from part 1 of this notebook.  \n",
    "The final file, which is available in this folder as ```sample_manifest_example.json``` has the following content:  \n",
    "```\n",
    "[\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_spec_q25\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 654\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"wt_spec_q50\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 988\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"wt_spec_q75\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 1445\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"wt_grc1_q25\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 25568\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"wt_grc1_q50\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 32424\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"wt_grc1_q75\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 40281\n",
    "    },\n",
    "        {\n",
    "        \"sample_id\": \"wt_grc2_q25\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 26732\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"wt_grc2_q50\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 32519\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"wt_grc2_q75\", \n",
    "        \"genotypes\": [\"wt_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 39864\n",
    "    },\n",
    "\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_spec_q25\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 654\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_spec_q50\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 988\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_spec_q75\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 1445\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_grc1_q25\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 25568\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_grc1_q50\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 32424\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_grc1_q75\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 40281\n",
    "    },\n",
    "        {\n",
    "        \"sample_id\": \"crtmut_grc2_q25\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 26732\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_grc2_q50\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 32519\n",
    "    },\n",
    "    {\n",
    "        \"sample_id\": \"crtmut_grc2_q75\", \n",
    "        \"genotypes\": [\"crtmut_spec\"], \n",
    "        \"proportions\": [1], \n",
    "        \"num_reads\": 39864\n",
    "    }\n",
    "\n",
    "]\n",
    "\n",
    "```\n",
    "As an example, sample \"crtmut_grc2_q75\" uses the crt mutant genotype to simulate a run of the GRC2 panel with 39864 reads, which is the 75 quantile read count for the GRC2 panel as calculated from real-world data in part 1 of this notebook."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
